{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Gene Sets and Expression for Tumor vs. Normal and Tissue Classification\n",
    "\n",
    "Use gene sets from [MSigDB](http://software.broadinstitute.org/gsea/msigdb) to both prune the number of genes/features as well as a source of pathway information to encorporate into layer design."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tensorflow as tf\n",
    "import keras\n",
    "import matplotlib.pyplot as pyplot\n",
    "\n",
    "# fix random seed for reproducibility\n",
    "np.random.seed(42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load TCGA+TARGET+GTEX Expression and Convert To Hugo\n",
    "\n",
    "Expression in TPM is only available with Ensembl gene identifiers from Xena. Load this up and convert into Hugo gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expression: samples=60498 genes=19260\n",
      "CPU times: user 2.58 s, sys: 3.98 s, total: 6.56 s\n",
      "Wall time: 6.64 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "expression = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").sort_index(axis=1)\n",
    "print(\"expression: samples={} genes={}\".format(*expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Feature labels are ensemble ids, convert to hugo gene names for use in interpreting\n",
    "hidden layers in any trained models as they are better known to most bioinformaticians \n",
    "and clinicians. We're using an assembled table from John Vivian @ UCSC here. Another\n",
    "option would be ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\n",
    "\"\"\"\n",
    "ensemble_to_hugo = pd.read_table(\n",
    "    \"https://github.com/jvivian/docker_tools/blob/master/gencode_hugo_mapping/attrs.tsv?raw=true\",\n",
    "    index_col=0)\n",
    "ensemble_to_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Ensembl ids may map to multiple hugo, pick the first and then generate an index matching our expression\n",
    "hugo = ensemble_to_hugo[~ensemble_to_hugo.index.duplicated(keep='first')].reindex(expression.index.values)[\"geneName\"].fillna(\"\")\n",
    "# Make sure we end up with the order of features being identical as some ensemble id's\n",
    "# map to the same hugo gene id\n",
    "assert(expression.index.equals(hugo.index))\n",
    "# # Change index from Ensembl to Hugo\n",
    "expression_hugo = expression\n",
    "expression_hugo.index = hugo.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2226-SM-5N9CH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-3026-SM-5GZYU</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ZR-A9CJ-01</th>\n",
       "      <th>TCGA-ZS-A9CD-01</th>\n",
       "      <th>TCGA-ZS-A9CE-01</th>\n",
       "      <th>TCGA-ZS-A9CF-01</th>\n",
       "      <th>TCGA-ZS-A9CF-02</th>\n",
       "      <th>TCGA-ZS-A9CG-01</th>\n",
       "      <th>TCGA-ZT-A8OM-01</th>\n",
       "      <th>TCGA-ZU-A8S4-01</th>\n",
       "      <th>TCGA-ZU-A8S4-11</th>\n",
       "      <th>TCGA-ZX-AA5X-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RP11-368I23.2</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-1.7809</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-167B3.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-742D12.2</th>\n",
       "      <td>-4.2934</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>0.3573</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB4B</th>\n",
       "      <td>5.1190</td>\n",
       "      <td>4.1277</td>\n",
       "      <td>4.4067</td>\n",
       "      <td>5.6860</td>\n",
       "      <td>4.0357</td>\n",
       "      <td>4.6849</td>\n",
       "      <td>4.5009</td>\n",
       "      <td>5.3954</td>\n",
       "      <td>4.9402</td>\n",
       "      <td>5.4683</td>\n",
       "      <td>...</td>\n",
       "      <td>4.1780</td>\n",
       "      <td>4.5547</td>\n",
       "      <td>3.6737</td>\n",
       "      <td>4.9331</td>\n",
       "      <td>3.6254</td>\n",
       "      <td>3.7646</td>\n",
       "      <td>5.5201</td>\n",
       "      <td>5.4216</td>\n",
       "      <td>3.3647</td>\n",
       "      <td>4.7991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC104071.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "RP11-368I23.2                   -9.9658                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -4.2934                    0.0014   \n",
       "RAB4B                            5.1190                    4.1277   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "RP11-368I23.2                   -9.9658                   -1.2481   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -9.9658                   -5.5735   \n",
       "RAB4B                            4.4067                    5.6860   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "RP11-368I23.2                   -3.8160                   -1.7809   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                    0.3573                   -9.9658   \n",
       "RAB4B                            4.0357                    4.6849   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-2226-SM-5N9CH  GTEX-1117F-2426-SM-5EGGH  \\\n",
       "RP11-368I23.2                   -9.9658                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -6.5064                   -5.0116   \n",
       "RAB4B                            4.5009                    5.3954   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-3026-SM-5GZYU  \\\n",
       "RP11-368I23.2                   -3.6259                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -9.9658                   -5.0116   \n",
       "RAB4B                            4.9402                    5.4683   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "                    ...         TCGA-ZR-A9CJ-01  TCGA-ZS-A9CD-01  \\\n",
       "RP11-368I23.2       ...                 -9.9658          -4.6082   \n",
       "RP11-167B3.1        ...                 -9.9658          -9.9658   \n",
       "RP11-742D12.2       ...                 -9.9658          -9.9658   \n",
       "RAB4B               ...                  4.1780           4.5547   \n",
       "AC104071.1          ...                 -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZS-A9CE-01  TCGA-ZS-A9CF-01  TCGA-ZS-A9CF-02  \\\n",
       "RP11-368I23.2          -9.9658          -9.9658          -4.6082   \n",
       "RP11-167B3.1           -9.9658          -9.9658          -9.9658   \n",
       "RP11-742D12.2          -9.9658          -9.9658          -9.9658   \n",
       "RAB4B                   3.6737           4.9331           3.6254   \n",
       "AC104071.1             -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZS-A9CG-01  TCGA-ZT-A8OM-01  TCGA-ZU-A8S4-01  \\\n",
       "RP11-368I23.2          -9.9658          -3.6259          -9.9658   \n",
       "RP11-167B3.1           -9.9658          -9.9658          -9.9658   \n",
       "RP11-742D12.2          -9.9658          -9.9658          -9.9658   \n",
       "RAB4B                   3.7646           5.5201           5.4216   \n",
       "AC104071.1             -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZU-A8S4-11  TCGA-ZX-AA5X-01  \n",
       "RP11-368I23.2          -9.9658          -9.9658  \n",
       "RP11-167B3.1           -9.9658          -9.9658  \n",
       "RP11-742D12.2          -9.9658          -6.5064  \n",
       "RAB4B                   3.3647           4.7991  \n",
       "AC104071.1             -9.9658          -9.9658  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 41s, sys: 26.1 s, total: 2min 7s\n",
      "Wall time: 1min 33s\n"
     ]
    }
   ],
   "source": [
    "# %%time\n",
    "# exp = expression_hugo\n",
    "# exp.apply(np.exp2)\n",
    "# exp.apply(lambda x: x - 0.001)\n",
    "# exp.groupby(level=0).aggregate(np.mean).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3min, sys: 35 s, total: 3min 35s\n",
      "Wall time: 2min 41s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Multiple Ensemble genes map to the same Hugo name. Each of these values has been normalized via log2(TPM+0.001)\n",
    "# so we convert back into TPM, compute the mean, and re-normalize\n",
    "expression_hugo_mean = expression_hugo.apply(np.exp2).apply(lambda x: x - 0.001).groupby(level=0).aggregate(np.mean).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2226-SM-5N9CH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-3026-SM-5GZYU</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ZR-A9CJ-01</th>\n",
       "      <th>TCGA-ZS-A9CD-01</th>\n",
       "      <th>TCGA-ZS-A9CE-01</th>\n",
       "      <th>TCGA-ZS-A9CF-01</th>\n",
       "      <th>TCGA-ZS-A9CF-02</th>\n",
       "      <th>TCGA-ZS-A9CG-01</th>\n",
       "      <th>TCGA-ZT-A8OM-01</th>\n",
       "      <th>TCGA-ZU-A8S4-01</th>\n",
       "      <th>TCGA-ZU-A8S4-11</th>\n",
       "      <th>TCGA-ZX-AA5X-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.96580</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.96580</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.96580</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.406064</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.003874</td>\n",
       "      <td>-2.244654</td>\n",
       "      <td>-3.45799</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>4.4595</td>\n",
       "      <td>1.1512</td>\n",
       "      <td>5.2411</td>\n",
       "      <td>5.4758</td>\n",
       "      <td>4.553400</td>\n",
       "      <td>4.6224</td>\n",
       "      <td>4.5028</td>\n",
       "      <td>5.613000</td>\n",
       "      <td>4.447600</td>\n",
       "      <td>4.73020</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6463</td>\n",
       "      <td>11.1518</td>\n",
       "      <td>11.8930</td>\n",
       "      <td>8.9466</td>\n",
       "      <td>7.6347</td>\n",
       "      <td>9.1454</td>\n",
       "      <td>7.6426</td>\n",
       "      <td>3.5994</td>\n",
       "      <td>10.9593</td>\n",
       "      <td>3.5935</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "                            -9.9658                   -9.9658   \n",
       "5S_rRNA                     -9.9658                   -9.9658   \n",
       "5_8S_rRNA                   -9.9658                   -9.9658   \n",
       "7SK                         -9.9658                   -9.9658   \n",
       "A1BG                         4.4595                    1.1512   \n",
       "\n",
       "           GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "                            -9.9658                   -9.9658   \n",
       "5S_rRNA                     -9.9658                   -9.9658   \n",
       "5_8S_rRNA                   -9.9658                   -9.9658   \n",
       "7SK                         -9.9658                   -9.9658   \n",
       "A1BG                         5.2411                    5.4758   \n",
       "\n",
       "           GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "                          -9.965800                   -9.9658   \n",
       "5S_rRNA                   -9.965800                   -9.9658   \n",
       "5_8S_rRNA                 -9.965800                   -9.9658   \n",
       "7SK                       -3.406064                   -9.9658   \n",
       "A1BG                       4.553400                    4.6224   \n",
       "\n",
       "           GTEX-1117F-2226-SM-5N9CH  GTEX-1117F-2426-SM-5EGGH  \\\n",
       "                            -9.9658                 -9.965800   \n",
       "5S_rRNA                     -9.9658                 -9.965800   \n",
       "5_8S_rRNA                   -9.9658                 -9.965800   \n",
       "7SK                         -9.9658                 -2.003874   \n",
       "A1BG                         4.5028                  5.613000   \n",
       "\n",
       "           GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-3026-SM-5GZYU  \\\n",
       "                          -9.965800                  -9.96580   \n",
       "5S_rRNA                   -9.965800                  -9.96580   \n",
       "5_8S_rRNA                 -9.965800                  -9.96580   \n",
       "7SK                       -2.244654                  -3.45799   \n",
       "A1BG                       4.447600                   4.73020   \n",
       "\n",
       "                ...         TCGA-ZR-A9CJ-01  TCGA-ZS-A9CD-01  TCGA-ZS-A9CE-01  \\\n",
       "                ...                 -9.9658          -9.9658          -9.9658   \n",
       "5S_rRNA         ...                 -9.9658          -9.9658          -9.9658   \n",
       "5_8S_rRNA       ...                 -9.9658          -9.9658          -9.9658   \n",
       "7SK             ...                 -9.9658          -9.9658          -9.9658   \n",
       "A1BG            ...                  3.6463          11.1518          11.8930   \n",
       "\n",
       "           TCGA-ZS-A9CF-01  TCGA-ZS-A9CF-02  TCGA-ZS-A9CG-01  TCGA-ZT-A8OM-01  \\\n",
       "                   -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "5S_rRNA            -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "5_8S_rRNA          -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "7SK                -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "A1BG                8.9466           7.6347           9.1454           7.6426   \n",
       "\n",
       "           TCGA-ZU-A8S4-01  TCGA-ZU-A8S4-11  TCGA-ZX-AA5X-01  \n",
       "                   -9.9658          -9.9658          -9.9658  \n",
       "5S_rRNA            -9.9658          -9.9658          -9.9658  \n",
       "5_8S_rRNA          -9.9658          -9.9658          -9.9658  \n",
       "7SK                -9.9658          -9.9658          -9.9658  \n",
       "A1BG                3.5994          10.9593           3.5935  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_hugo_mean.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoU2-30</th>\n",
       "      <th>snoU2_19</th>\n",
       "      <th>snoU83B</th>\n",
       "      <th>snoZ196</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>uc_338</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.4595</td>\n",
       "      <td>0.9343</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>7.5126</td>\n",
       "      <td>0.8164</td>\n",
       "      <td>-2.1140</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.158328</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>1.1512</td>\n",
       "      <td>-1.2828</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>6.0777</td>\n",
       "      <td>-2.3147</td>\n",
       "      <td>0.5568</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.126224</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>5.2411</td>\n",
       "      <td>0.8488</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>10.0319</td>\n",
       "      <td>0.1257</td>\n",
       "      <td>-1.1172</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.864951</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>5.4758</td>\n",
       "      <td>2.6325</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>9.7572</td>\n",
       "      <td>1.7702</td>\n",
       "      <td>-1.8836</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.918501</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.406064</td>\n",
       "      <td>4.5534</td>\n",
       "      <td>1.3051</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>7.7931</td>\n",
       "      <td>-0.0725</td>\n",
       "      <td>-2.2447</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.018394</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 58581 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          5S_rRNA  5_8S_rRNA       7SK    A1BG  A1BG-AS1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658    -9.9658 -9.965800  4.4595    0.9343   \n",
       "GTEX-1117F-0426-SM-5EGHI  -9.9658    -9.9658 -9.965800  1.1512   -1.2828   \n",
       "GTEX-1117F-0526-SM-5EGHJ  -9.9658    -9.9658 -9.965800  5.2411    0.8488   \n",
       "GTEX-1117F-0626-SM-5N9CS  -9.9658    -9.9658 -9.965800  5.4758    2.6325   \n",
       "GTEX-1117F-0726-SM-5GIEN  -9.9658    -9.9658 -3.406064  4.5534    1.3051   \n",
       "\n",
       "                            A1CF      A2M  A2M-AS1   A2ML1  A2ML1-AS1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7 -5.0116   7.5126   0.8164 -2.1140    -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI -6.5064   6.0777  -2.3147  0.5568    -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ -6.5064  10.0319   0.1257 -1.1172    -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS -9.9658   9.7572   1.7702 -1.8836    -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN -9.9658   7.7931  -0.0725 -2.2447    -9.9658   \n",
       "\n",
       "                             ...      snoU2-30  snoU2_19  snoU83B  snoZ196  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "\n",
       "                          snoZ278  snoZ40   snoZ6  snosnR66    uc_338  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658 -9.9658 -9.9658   -9.9658  0.158328   \n",
       "GTEX-1117F-0426-SM-5EGHI  -9.9658 -9.9658 -9.9658   -9.9658 -2.126224   \n",
       "GTEX-1117F-0526-SM-5EGHJ  -9.9658 -9.9658 -9.9658   -9.9658 -0.864951   \n",
       "GTEX-1117F-0626-SM-5N9CS  -9.9658 -9.9658 -9.9658   -9.9658 -0.918501   \n",
       "GTEX-1117F-0726-SM-5GIEN  -9.9658 -9.9658 -9.9658   -9.9658 -2.018394   \n",
       "\n",
       "                          yR211F11.2  \n",
       "GTEX-1117F-0226-SM-5GZZ7     -9.9658  \n",
       "GTEX-1117F-0426-SM-5EGHI     -9.9658  \n",
       "GTEX-1117F-0526-SM-5EGHJ     -9.9658  \n",
       "GTEX-1117F-0626-SM-5N9CS     -9.9658  \n",
       "GTEX-1117F-0726-SM-5GIEN     -9.9658  \n",
       "\n",
       "[5 rows x 58581 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Drop one column that has empty gene name, artificat of Ensembl to Hugo and this will\n",
    "# ensure the Treehouse compendium used below as a verification matches index wise\n",
    "X = expression_hugo_mean.drop(labels=\"\").T\n",
    "X.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Y: samples=19131 labels=6\n",
      "CPU times: user 54.1 ms, sys: 6.91 ms, total: 61 ms\n",
      "Wall time: 60.4 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Load samples and labels\n",
    "# X = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\")\n",
    "# print(\"X: samples={} genes={}\".format(*X.shape))\n",
    "\n",
    "Y = pd.read_table(\"data/TcgaTargetGTEX_phenotype.txt.gz\", compression=\"gzip\", header=0,\n",
    "                      names=[\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "                      sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\")\n",
    "print(\"Y: samples={} labels={}\".format(*Y.shape))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ingest Clinical and Pathways\n",
    "\n",
    "Load the expression and clinical as wrangled in ingest.ipynb as well as gene sets downloaded from msigdb. Wrangle so that only genes present in each are present and ensure the order of features is consistent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19126 samples with non-null labels\n"
     ]
    }
   ],
   "source": [
    "# Sort so indexes match\n",
    "X.sort_index(axis=0, inplace=True)\n",
    "X.sort_index(axis=1, inplace=True)\n",
    "Y.sort_index(axis=0, inplace=True)\n",
    "\n",
    "# Remove rows where the class is null or the sample is missing\n",
    "intersection = X.index.intersection(Y[pd.notnull(Y[\"primary_site\"])].index)\n",
    "X = X[X.index.isin(intersection)]\n",
    "Y = Y[Y.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(X.index.equals(Y.index))\n",
    "\n",
    "print(intersection.shape[0], \"samples with non-null labels\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "      <th>tumor_normal_value</th>\n",
       "      <th>primary_site_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>18968</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126.000000</td>\n",
       "      <td>19126.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>46</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1212</td>\n",
       "      <td>1212</td>\n",
       "      <td>1846</td>\n",
       "      <td>9185</td>\n",
       "      <td>10453</td>\n",
       "      <td>10534</td>\n",
       "      <td>10530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.550559</td>\n",
       "      <td>20.651992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.497450</td>\n",
       "      <td>12.419634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>19.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>45.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         category                    disease primary_site  \\\n",
       "count                       19126                      19126        19126   \n",
       "unique                         93                         93           46   \n",
       "top     Breast Invasive Carcinoma  Breast Invasive Carcinoma        Brain   \n",
       "freq                         1212                       1212         1846   \n",
       "mean                          NaN                        NaN          NaN   \n",
       "std                           NaN                        NaN          NaN   \n",
       "min                           NaN                        NaN          NaN   \n",
       "50%                           NaN                        NaN          NaN   \n",
       "max                           NaN                        NaN          NaN   \n",
       "\n",
       "          sample_type gender  study tumor_normal  tumor_normal_value  \\\n",
       "count           19126  18968  19126        19126        19126.000000   \n",
       "unique             16      2      3            2                 NaN   \n",
       "top     Primary Tumor   Male   TCGA        Tumor                 NaN   \n",
       "freq             9185  10453  10534        10530                 NaN   \n",
       "mean              NaN    NaN    NaN          NaN            0.550559   \n",
       "std               NaN    NaN    NaN          NaN            0.497450   \n",
       "min               NaN    NaN    NaN          NaN            0.000000   \n",
       "50%               NaN    NaN    NaN          NaN            1.000000   \n",
       "max               NaN    NaN    NaN          NaN            1.000000   \n",
       "\n",
       "        primary_site_value  \n",
       "count         19126.000000  \n",
       "unique                 NaN  \n",
       "top                    NaN  \n",
       "freq                   NaN  \n",
       "mean             20.651992  \n",
       "std              12.419634  \n",
       "min               0.000000  \n",
       "50%              19.000000  \n",
       "max              45.000000  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "Y.loc[:, \"tumor_normal\"] = Y.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)\n",
    "\n",
    "# Convert tumor_normal and primary_site into numerical values for two-hot multi-class training\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "Y[\"tumor_normal_value\"] = pd.Series(encoder.fit_transform(Y[\"tumor_normal\"]), index=Y.index)\n",
    "encoder = LabelEncoder()\n",
    "Y[\"primary_site_value\"] = pd.Series(encoder.fit_transform(Y[\"primary_site\"]), index=Y.index)\n",
    "Y.describe(include=\"all\", percentiles=[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoU2-30</th>\n",
       "      <th>snoU2_19</th>\n",
       "      <th>snoU83B</th>\n",
       "      <th>snoZ196</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>uc_338</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.4595</td>\n",
       "      <td>0.9343</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>7.5126</td>\n",
       "      <td>0.8164</td>\n",
       "      <td>-2.1140</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>0.158328</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-145LS-1226-SM-5Q5D9</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.341285</td>\n",
       "      <td>3.3350</td>\n",
       "      <td>1.6281</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>11.3547</td>\n",
       "      <td>1.0847</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-0.727251</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-X638-0126-SM-47JZ8</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.2609</td>\n",
       "      <td>1.6875</td>\n",
       "      <td>-4.0350</td>\n",
       "      <td>6.7694</td>\n",
       "      <td>0.7146</td>\n",
       "      <td>5.0845</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.740712</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-24-1422-01</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.442188</td>\n",
       "      <td>4.1962</td>\n",
       "      <td>2.3788</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>6.4925</td>\n",
       "      <td>0.6425</td>\n",
       "      <td>-0.6643</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.577273</td>\n",
       "      <td>-4.0350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BF-AAP0-06</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>6.6121</td>\n",
       "      <td>3.7919</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>7.9917</td>\n",
       "      <td>0.1648</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.873583</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 58581 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          5S_rRNA  5_8S_rRNA       7SK    A1BG  A1BG-AS1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658    -9.9658 -9.965800  4.4595    0.9343   \n",
       "GTEX-145LS-1226-SM-5Q5D9  -9.9658    -9.9658 -4.341285  3.3350    1.6281   \n",
       "GTEX-X638-0126-SM-47JZ8   -9.9658    -9.9658 -9.965800  4.2609    1.6875   \n",
       "TCGA-24-1422-01           -9.9658    -9.9658 -4.442188  4.1962    2.3788   \n",
       "TCGA-BF-AAP0-06           -9.9658    -9.9658 -9.965800  6.6121    3.7919   \n",
       "\n",
       "                            A1CF      A2M  A2M-AS1   A2ML1  A2ML1-AS1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7 -5.0116   7.5126   0.8164 -2.1140    -9.9658   \n",
       "GTEX-145LS-1226-SM-5Q5D9 -6.5064  11.3547   1.0847 -4.6082    -9.9658   \n",
       "GTEX-X638-0126-SM-47JZ8  -4.0350   6.7694   0.7146  5.0845    -9.9658   \n",
       "TCGA-24-1422-01          -5.5735   6.4925   0.6425 -0.6643    -9.9658   \n",
       "TCGA-BF-AAP0-06          -9.9658   7.9917   0.1648 -4.6082    -9.9658   \n",
       "\n",
       "                             ...      snoU2-30  snoU2_19  snoU83B  snoZ196  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-145LS-1226-SM-5Q5D9     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-X638-0126-SM-47JZ8      ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "TCGA-24-1422-01              ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "TCGA-BF-AAP0-06              ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "\n",
       "                          snoZ278  snoZ40   snoZ6  snosnR66    uc_338  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658 -9.9658 -9.9658   -9.9658  0.158328   \n",
       "GTEX-145LS-1226-SM-5Q5D9  -9.9658 -9.9658 -9.9658   -9.9658 -0.727251   \n",
       "GTEX-X638-0126-SM-47JZ8   -9.9658 -9.9658 -9.9658   -9.9658 -2.740712   \n",
       "TCGA-24-1422-01           -9.9658 -9.9658 -9.9658   -9.9658 -3.577273   \n",
       "TCGA-BF-AAP0-06           -9.9658 -9.9658 -9.9658   -9.9658 -3.873583   \n",
       "\n",
       "                          yR211F11.2  \n",
       "GTEX-1117F-0226-SM-5GZZ7     -9.9658  \n",
       "GTEX-145LS-1226-SM-5Q5D9     -9.9658  \n",
       "GTEX-X638-0126-SM-47JZ8      -9.9658  \n",
       "TCGA-24-1422-01              -4.0350  \n",
       "TCGA-BF-AAP0-06              -9.9658  \n",
       "\n",
       "[5 rows x 58581 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X[0:20000:3000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "      <th>tumor_normal_value</th>\n",
       "      <th>primary_site_value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-145LS-1226-SM-5Q5D9</th>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-X638-0126-SM-47JZ8</th>\n",
       "      <td>Skin - Sun Exposed (Lower Leg)</td>\n",
       "      <td>Skin - Sun Exposed (Lower Leg)</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-24-1422-01</th>\n",
       "      <td>Ovarian Serous Cystadenocarcinoma</td>\n",
       "      <td>Ovarian Serous Cystadenocarcinoma</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BF-AAP0-06</th>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   category  \\\n",
       "id                                                            \n",
       "GTEX-1117F-0226-SM-5GZZ7             Adipose - Subcutaneous   \n",
       "GTEX-145LS-1226-SM-5Q5D9                               Lung   \n",
       "GTEX-X638-0126-SM-47JZ8      Skin - Sun Exposed (Lower Leg)   \n",
       "TCGA-24-1422-01           Ovarian Serous Cystadenocarcinoma   \n",
       "TCGA-BF-AAP0-06                     Skin Cutaneous Melanoma   \n",
       "\n",
       "                                                    disease    primary_site  \\\n",
       "id                                                                            \n",
       "GTEX-1117F-0226-SM-5GZZ7             Adipose - Subcutaneous  Adipose Tissue   \n",
       "GTEX-145LS-1226-SM-5Q5D9                               Lung            Lung   \n",
       "GTEX-X638-0126-SM-47JZ8      Skin - Sun Exposed (Lower Leg)            Skin   \n",
       "TCGA-24-1422-01           Ovarian Serous Cystadenocarcinoma           Ovary   \n",
       "TCGA-BF-AAP0-06                     Skin Cutaneous Melanoma            Skin   \n",
       "\n",
       "                            sample_type  gender study tumor_normal  \\\n",
       "id                                                                   \n",
       "GTEX-1117F-0226-SM-5GZZ7  Normal Tissue  Female  GTEX       Normal   \n",
       "GTEX-145LS-1226-SM-5Q5D9  Normal Tissue  Female  GTEX       Normal   \n",
       "GTEX-X638-0126-SM-47JZ8   Normal Tissue  Female  GTEX       Normal   \n",
       "TCGA-24-1422-01           Primary Tumor  Female  TCGA        Tumor   \n",
       "TCGA-BF-AAP0-06              Metastatic  Female  TCGA        Tumor   \n",
       "\n",
       "                          tumor_normal_value  primary_site_value  \n",
       "id                                                                \n",
       "GTEX-1117F-0226-SM-5GZZ7                   0                   0  \n",
       "GTEX-145LS-1226-SM-5Q5D9                   0                  22  \n",
       "GTEX-X638-0126-SM-47JZ8                    0                  33  \n",
       "TCGA-24-1422-01                            1                  26  \n",
       "TCGA-BF-AAP0-06                            1                  33  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y[0:20000:3000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 186 gene sets\n",
      "Subsetting to 5172 genes\n",
      "X_pruned shape (19126, 5172)\n"
     ]
    }
   ],
   "source": [
    "# Load gene sets from downloaded MSigDB gmt file (KEGG to for now as its experimental vs. computational)\n",
    "with open(\"data/c2.cp.kegg.v6.1.symbols.gmt\") as f:\n",
    "    gene_sets = { line.strip().split(\"\\t\")[0]: line.strip().split(\"\\t\")[2:] for line in f.readlines()}\n",
    "print(\"Loaded {} gene sets\".format(len(gene_sets)))\n",
    "\n",
    "# Drop any genes that are not in X - sort so that the order is the same as X_pruned.columns\n",
    "gene_sets = {name: sorted([gene for gene in genes if gene in X.columns.values]) \n",
    "             for name, genes in gene_sets.items()}\n",
    "\n",
    "# Find the union of all gene's in the gene sets in order to filter our input rows\n",
    "all_gene_set_genes = sorted(list(set().union(*[gene_set for gene_set in gene_sets.values()])))\n",
    "print(\"Subsetting to {} genes\".format(len(all_gene_set_genes)))\n",
    "\n",
    "# Prune X to only include genes in the gene sets\n",
    "X_pruned = X.drop(labels=(set(X.columns) - set(all_gene_set_genes)), axis=1, errors=\"ignore\")\n",
    "assert X_pruned[\"TP53\"][\"TCGA-ZP-A9D4-01\"] == X[\"TP53\"][\"TCGA-ZP-A9D4-01\"]\n",
    "print(\"X_pruned shape\", X_pruned.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure the genes are the same and in the same order\n",
    "assert len(all_gene_set_genes) == len(X_pruned.columns.values)\n",
    "assert list(X_pruned.columns.values) == all_gene_set_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wrangle Labels, Train and Test Sets\n",
    "\n",
    "Add one hot labels for tumor/normal and primary site and then build training and test sets stratifying on primary site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19126, 47)\n",
      "[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]\n"
     ]
    }
   ],
   "source": [
    "# Create a multi-label output with tumor/normal + primary site one-hots\n",
    "# from keras.utils import np_utils\n",
    "# primary_site_one_hot = np_utils.to_categorical(Y[\"primary_site_value\"])\n",
    "# tumor_normal_one_hot = Y[\"tumor_normal_value\"]\n",
    "# Y_multiclass_onehot = np.append(\n",
    "#     tumor_normal_one_hot.values.reshape(tumor_normal_one_hot.shape[0],-1), \n",
    "#     primary_site_one_hot, axis=1)\n",
    "# print(Y_multiclass_onehot.shape)\n",
    "# print(Y_multiclass_onehot[0:20000:5000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19126, 46)\n",
      "[[1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]\n"
     ]
    }
   ],
   "source": [
    "# Create a primary site one-hot\n",
    "from keras.utils import np_utils\n",
    "primary_site_one_hot = np_utils.to_categorical(Y[\"primary_site_value\"])\n",
    "tumor_normal_one_hot = Y[\"tumor_normal_value\"]\n",
    "Y_multiclass_onehot = primary_site_one_hot\n",
    "print(Y_multiclass_onehot.shape)\n",
    "print(Y_multiclass_onehot[0:20000:5000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15300, 5172) (3826, 5172)\n"
     ]
    }
   ],
   "source": [
    "# Split into stratified training and test sets based on classes (i.e. tissue type) so that we have equal\n",
    "# proportions of each tissue type in the train and test sets\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "for train_index, test_index in split.split(X_pruned.values, Y[\"tumor_normal_value\"]):\n",
    "    X_train, X_test = X_pruned.values[train_index], X_pruned.values[test_index]\n",
    "    y_train, y_test = Y_multiclass_onehot[train_index], Y_multiclass_onehot[test_index]\n",
    "    classes_train, classes_test = Y[\"primary_site_value\"].values[train_index], Y[\"primary_site_value\"].values[test_index]\n",
    "\n",
    "print(X_train.shape, X_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEICAYAAABWJCMKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xu8VVW99/HPV8QgxVBEQ0BBQ4+XjoTbW9rR8s7jSeto6skb6UF79JWWVmh1vGXpOaVpmj6UJFpKHM3kePQxNC9ZeUHEC5KJSrJlCxsUwfKG/s4fY2yZLNbee+077Pl9v17rtdYcc8w5xxxrzvmbY8y55lJEYGZm5bNOTxfAzMx6hgOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSfWKACDpPEm/6OZl3iTpsHZO+ylJz3Z2mbqapC0kvSGpTzunv1TSKW2cZp6k/fLncyT9rD3Lbmbeb0jaKn++TtJ3O3He10j6TmfNrzDfbt/W1zaSvijpt2tAOUZICknr9nRZmrPWBABJ/yppRt5pGyTdKWmvHirLPwI7Abfl4RMkvZfLtkzSLEmHNDd9RPw+IrbtrvK2haRhkm6RtFjS65KeknQCQES8FBEbRMR7Oe99kk5qw+z/E/iWpPXaU7aI+F5EtLq8WsuV1+WF9pSlYnknSHqwYt6nRMSFHZ13Z1obgkdnBOKI+GVEHNBZZeoqxZObDs5nte2vVmtFAJD0NeBHwPeAzYAtgJ8Ah/ZQkU4Gfhmr/oruTxGxATAQuBaYKmnjygm7+2ygHWfrNwDzgS2BQcBxwMLOKEtENAB/Bj7bGfNrrzX5jMxa5u+uk0XEGv0CPgK8ARzRQp7zgF8Uhv8LeAV4HXgA2KEwbizwDLAceBk4K6dvAtwOLAVeBX4PrNPM8l4A9ioMnwA8WBheHwigDtgHqAe+mct0Q1NaIf884OvAk8DfSAFkM+DOXM67gY1qXL/rgKuBO/K8vk46gK9byPMvwKxm1u0NYHQz40bk9VoXuAh4D3grT3NlzvMPwPRch88CX6iYx7eAn7fwXR4L/BVYkvPOA/ar/J6BfsAvcr6lwKO5zporVwCnAs8BLxbSPlaot2ty2ZcD9wNbVq53oZz3AScB2+VlvZeXt7Qwv+8W8v8bMDfXyzRg88K4AE7JZXsNuApQC9v6zcCvcjlnAjsVxm8O3AI0Ai8CX8npBwHvAO/mcj4BfBp4qjDt3cAjheEHgcNamm8etw4wAXg+fx9TgY0r6u544CVgMfCtZtZtfC7fO7mM/13YP75J2j/eJm1/TctbTtqfP9fC/tiW+t0V+BNpm2oArgTWq2VeQB/gB3kdXyBtb6tsN4X53AC8D7yZ1/UbOX134I95+U8A+1Ss1wt5nV8Evkgz21/Nx9e2HpC7+5U33BXVKrFipygGgC8BA4APkVoOswrjGoBP5c8bAWPy5++TDgB98+tT1TYSVh7cB1fb4PLGeXr+kj5COtivAC7J5elP9QDwEOkANhRYRNqxP5Gn+R1wbo3rdx0pMOxJ2jH7kXaQgwt5bgXObKYu7wb+ABwFbFExbkRxgyYfBCvqZj4wLtfDmLwzFAPU54GZzSx7+7wR/1Net0tz3VULACcD/w18mLTj7QxsWK1chR13OrAx0L+QVgwAywvLvrzwna6y3pXLoOKAU5jfd/Pnz+R6GJPn/WPggYqy3U5qPW5BOsge1MK2/i5wOGk7PYt0MOibv+/HgH8H1gO2Ih0wDmxmP+lHOgBtkr+vV4AFpG2rfx43qIb5nkHafofl9ft/wE0VdffTPM+dSAfx7ZpZvw/qrWL/mAUML3x3R5CC0jrAkaSTnSHVvo821u/OpIPwurnsc4AzapkXKTD8OZdzY+BemgkAhfXarzA8lBRAx+b12j8PDybtW8uAbXPeIeT9qnJ92/JaG7qABgGLI2JFrRNExKSIWB4Rb5M2+p0kfSSPfhfYXtKGEfFaRMwspA8hnfW9G6mfvtqDkgbm9+UV6btLWkraiY4mnZG8nse9TzqAvx0RbzZT7B9HxMKIeJnU+ng4Ih7P63ArKRjUsn4At0XEHyLi/Yh4C5gMHAOQu6UOBG5sphxH5OV/B3gxX8/YpZm8lQ4B5kXEzyNiRa7bW0gHqybLWVmHlQ4Hbo+IB/K6fYdUd9W8S9o2PhYR70XEYxGxrJXyfT8iXm3hO/ifwrK/BewhaXgr86zFF4FJETEzz/vsPO8RhTwXR8TSiHiJdOAY3cL8HouImyPiXVKQ7Ec6aO1COjG5ICLeiXR946ekYL6avG3MIAW9OtIZ9oOkk4fdgeciYkkN8z2ZdFZfX9gmD6/orjk/It6MiCdIZ7Y7tVZpFa6IiPlN311E/FdELMjb+K9IZ+S7tjB9TfWbt6OH8vY7jxTM9q5xXl8AfpTL+SrppLItjgHuiIg78npNJ30/Y/P494EdJfWPiIaImN3G+a9mbQgAS4BNau37k9RH0sWSnpe0jBRlIZ3lQOr+GAv8VdL9kvbI6f9JaqL/VtILkiY0s4il+X1ARfpDETEwIjaJiN0j4u7CuMa8s7Wk2M/+ZpXhDWpcP0hn4UW/AP5Z0gakjfT3kfrjV5OD4oSI2IHUIpkF/EaSWik/pOsGu0la2vQiHfw+WsgzgJV1WGnzYtkj4m+k77+aG4C7gCmSFkj6D0l9WylfZb00Oz4i3iB112zeyjS12JzUrVWc9xLSGV+TVwqf/07+vmso5/ukLsbNSfW/eUX9n0P6HptzP6lF+k/5832kA97eeZga5rslcGth3BxSl0RxuW1ZvxbXGUDScfnkpGmZO7LqPlCppuVL2kbS7ZJeyfvX96rMt7l5rbL9UvjOa7QlcERFPe9Fatn8jdTSOQVokPQ/kv6hjfNfzdoQAP5E6uOq9ZbLfyVdHN6P1AUzIqcLICIejYhDgU2B35D6K8ln1GdGxFbAPwNfk7Rv5czzF/E8sE0b1qFaS6K9Wly/asvLrYo/AZ8j9bHfUMuCImIxqU9zc1KTdrUsFcPzgftzIGx6bRARXy7k2Y50BlhNA6n5DICkD5PO8quV7d2IOD8itgc+SWp9HNdMuZorb6XisjcgrfMCUvcCpO6mJsWg1tp8F5B27qZ5r09ar5dbma6Wcq5D6npZQKr/Fyvqf0BENJ1BVitnZQC4n9UDQGvznU/qYiyO75e3u7Zq9buTtCWpBXIaMCgiBgJPs+o+0F5Xk7pxRkXEhqRAV+t8V9l+SV1ELam2/9xQUY/rR8TFABFxV0TsT+qp+DOpDqrNp2ZrfADI3Sj/Dlwl6TBJH5bUV9LBkv6jyiQDSH2MS0g77PeaRkhaL98j/JHcfF5GOlNB0iGSPpbPdJvS32umWHewerOwuzS7fq24HvgG8HFSl1JVki6RtKOkdSUNAL4MzM1dAZUWkvqDm9wObCPp2Pwd9ZW0i6TtCnn2Jl3cruZm4BBJe+VbRS+gmW1U0qclfTzf5bSM1CXU9H1VlqtWYwvLvpDUDTc/IhpJB+tjcgvsS8DWhekWAsNauL31RmCcpNGSPkT6zh7OXQztsbOkz+dW8Rmk7eEh4BFgmaRvSuqfy7pjoQtvITAiB40mfwS2JXWfPJK7FbYEdiPdYEAN870GuCgfmJE0WFJ779Cr5btrug7XmJc3jtQC6AwDSNvTG/kM+8ut5C+aCnxF6VbqjUgXqltSua5NLfUDcx33k7RPnt9mkj6bTx7eJl0rK27vLW1/zVrjAwBARFwKfA34NulLn0+K/r+pkv16UtPrZdLFz4cqxh8LzMvNu1PIfePAKNIF0DdIZ8s/iYj7minSROCLNXaLdLbW1q85t5Kb6rkV05wP57xLSRf6tqT52zYvJ/X1vibpiohYDhxA6hteQGoqN138RtIQ0oXeat8b+eBzKumA2UC6y6K+mWV/lBQwlpG6HO4n7UCrlauFda10I3AuqetnZ1L3VZN/I91RtQTYgXTgbPI7YDbwiqTFVdbrHtL1jFvyem1NM/3yNbqN1B3wGml7/nxuEb1Har2OJl0YXgz8jNRShHT3GMASSTNz2f5GuuFgdkS8k8f/CfhrRCzKeVqb7+WkO5t+K2k5aZvcrZ3rdi3pGt1SSc1tJ88AP8zlXEg6qflDO5dX6SxSK3s56Qz7V22Y9qekbsknSHX661byfx/4dl7XsyJiPql1fw4rj3NfJx2n1wHOJO1Xr5JOpP5vnk+L219Lmm5fsjaSdCMwNSKqbqRrIknPAydXXJ/ozuX/EHg+In7SE8s3s1U5AJSEpH8hnY1vky8cmlnJ+Vd1JSDpPlLXy7E++JtZE7cAzMxKaq24CGxmZp2v1S4gSf1It4N9KOe/OSLOlTQSmEK6V3omqXvhnXyb2/WkuyiWAEc23e4m6WzgRNLtS1+JiLtaWvYmm2wSI0aMaOeqmZmV02OPPbY4Iga3lq+WawBvA5+JiDfyLy0flHQn6bbMyyJiiqRrSAf2q/P7axHxMUlHkS48Hilpe9KtbzuQflh0t6Rt8i1mVY0YMYIZM2bUUEQzM2siqaZfIbfaBRTJG3mw6UFpQXrA1c05fTIrf6l7aB4mj9833y9/KDAl0vNwXiQ9dqGlZ3eYmVkXqukaQP5V2izSUyqnkx6FsDRWPqCtnpXPNRlKfh5GHv866WfvH6RXmcbMzLpZTQEg0tMWR5OeObIr6Xkuq2XL79V+HRstpK9C0nilf/6a0djYWEvxzMysHdr0O4CIWJrvKd8dGChp3XyW3/QwKkhn9sOB+vysko+QfrrclN6kOE1xGRNJj1qgrq7O96iaWc3effdd6uvreeut1h6+2zv069ePYcOG0bdvaw/Cra6Wu4AGA+/mg39/0lMoLyE9B/tw0p1Ax5P/H5f0TJDjSc/pOBz4XUSEpGnAjZIuJV0EHkV6yJSZWaeor69nwIABjBgxgp55VFf3iQiWLFlCfX09I0eObNc8amkBDAEm56curkN6/s3tkp4hPYv9u8DjpIc4kd9vkNT093dH5cLOljSV9ACzFcCpLd0BZGbWVm+99VYpDv4Akhg0aBAd6SpvNQBExJMU/o2qkP4CVe7iyX98ckQz87qI9J+tZmZdogwH/yYdXVf/EtjMrKT8MDgz67Uum/6XTp3fV/dv+Y8AlyxZwr77pj8SfOWVV+jTpw+DB6cf5D7yyCOst17r/9kybtw4JkyYwLbbbtvxArfCAcDWap29g7dFawcDK59BgwYxa9YsAM477zw22GADzjrrrFXyRAQRwTrrVO+A+fnPf97l5WziLiAzsy42d+5cdtxxR0455RTGjBlDQ0MD48ePp66ujh122IELLrjgg7x77bUXs2bNYsWKFQwcOJAJEyaw0047sccee7Bo0aJOLZcDgJlZN3jmmWc48cQTefzxxxk6dCgXX3wxM2bM4IknnmD69Ok888wzq03z+uuvs/fee/PEE0+wxx57MGnSpE4tkwOAmVk32Hrrrdlll10+GL7pppsYM2YMY8aMYc6cOVUDQP/+/Tn44IMB2HnnnZk3b16nlsnXAMzMusH666//wefnnnuOyy+/nEceeYSBAwdyzDHHVP31cvGicZ8+fVixYsVqeTrCLQAzs262bNkyBgwYwIYbbkhDQwN33dXiX6N0GbcAzKzXWlPv1BozZgzbb789O+64I1tttRV77rlnj5Rjjf5P4Lq6uvAfwlhLfBuoFc2ZM4fttqv2sOLeq9o6S3osIupam9ZdQGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlL+HYCZ9V73fr9z5/fps1sc3RmPgwaYNGkSY8eO5aMf/WjHytsKBwAzs05Sy+OgazFp0iTGjBnjAGBm1htMnjyZq666infeeYdPfvKTXHnllbz//vuMGzeOWbNmERGMHz+ezTbbjFmzZnHkkUfSv3//NrUc2soBwMysiz399NPceuut/PGPf2Tddddl/PjxTJkyha233prFixfz1FNPAbB06VIGDhzIj3/8Y6688kpGjx7dpeVyADAz62J33303jz76KHV16ekMb775JsOHD+fAAw/k2Wef5fTTT2fs2LEccMAB3VouBwAzsy4WEXzpS1/iwgsvXG3ck08+yZ133skVV1zBLbfcwsSJE7utXL4N1Mysi+23335MnTqVxYsXA+luoZdeeonGxkYigiOOOILzzz+fmTNnAjBgwACWL1/e5eVyC8DMeq9WbtvsLh//+Mc599xz2W+//Xj//ffp27cv11xzDX369OHEE08kIpDEJZdcAsC4ceM46aSTuvwisB8HbWs1Pw7aivw46MSPgzYzsxY5AJiZlVSrAUDScEn3Spojabak03P6eZJeljQrv8YWpjlb0lxJz0o6sJB+UE6bK2lC16ySmZXZmtyt3dk6uq61XAReAZwZETMlDQAekzQ9j7ssIn5QzCxpe+AoYAdgc+BuSU2dpVcB+wP1wKOSpkXEMx1aAzOzrF+/fixZsoRBgwYhqaeL06UigiVLltCvX792z6PVABARDUBD/rxc0hxgaAuTHApMiYi3gRclzQV2zePmRsQLAJKm5LwOAGbWKYYNG0Z9fT2NjY09XZRu0a9fP4YNG9bu6dt0G6ikEcAngIeBPYHTJB0HzCC1El4jBYeHCpPVszJgzK9I363KMsYD4wG22GKLthTPzEqub9++jBw5sqeLsdao+SKwpA2AW4AzImIZcDWwNTCa1EL4YVPWKpNHC+mrJkRMjIi6iKhreoyqmZl1vppaAJL6kg7+v4yIXwNExMLC+J8Ct+fBemB4YfJhwIL8ubl0MzPrZrXcBSTgWmBORFxaSB9SyPY54On8eRpwlKQPSRoJjAIeAR4FRkkaKWk90oXiaZ2zGmZm1la1tAD2BI4FnpI0K6edAxwtaTSpG2cecDJARMyWNJV0cXcFcGpEvAcg6TTgLqAPMCkiZnfiupiZWRvUchfQg1Tvv7+jhWkuAi6qkn5HS9OZmVn38S+BzcxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspPyXkF3A/1JlZmsDtwDMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKRaDQCShku6V9IcSbMlnZ7TN5Y0XdJz+X2jnC5JV0iaK+lJSWMK8zo+539O0vFdt1pmZtaadWvIswI4MyJmShoAPCZpOnACcE9EXCxpAjAB+CZwMDAqv3YDrgZ2k7QxcC5QB0Sez7SIeK2zV6rMLpv+lx5Z7lf336ZHlmtm7ddqCyAiGiJiZv68HJgDDAUOBSbnbJOBw/LnQ4HrI3kIGChpCHAgMD0iXs0H/enAQZ26NmZmVrM2XQOQNAL4BPAwsFlENEAKEsCmOdtQYH5hsvqc1lx65TLGS5ohaUZjY2NbimdmZm1QcwCQtAFwC3BGRCxrKWuVtGghfdWEiIkRURcRdYMHD661eGZm1kY1BQBJfUkH/19GxK9z8sLctUN+X5TT64HhhcmHAQtaSDczsx5Qy11AAq4F5kTEpYVR04CmO3mOB24rpB+X7wbaHXg9dxHdBRwgaaN8x9ABOc3MzHpALXcB7QkcCzwlaVZOOwe4GJgq6UTgJeCIPO4OYCwwF/g7MA4gIl6VdCHwaM53QUS82ilrYWZmbdZqAIiIB6nefw+wb5X8AZzazLwmAZPaUkAzM+sa/iWwmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUm1GgAkTZK0SNLThbTzJL0saVZ+jS2MO1vSXEnPSjqwkH5QTpsraULnr4qZmbVFLS2A64CDqqRfFhGj8+sOAEnbA0cBO+RpfiKpj6Q+wFXAwcD2wNE5r5mZ9ZB1W8sQEQ9IGlHj/A4FpkTE28CLkuYCu+ZxcyPiBQBJU3LeZ9pcYjMz6xQduQZwmqQncxfRRjltKDC/kKc+pzWXvhpJ4yXNkDSjsbGxA8UzM7OWtDcAXA1sDYwGGoAf5nRVyRstpK+eGDExIuoiom7w4MHtLJ6ZmbWm1S6gaiJiYdNnST8Fbs+D9cDwQtZhwIL8ubl0MzPrAe1qAUgaUhj8HNB0h9A04ChJH5I0EhgFPAI8CoySNFLSeqQLxdPaX2wzM+uoVlsAkm4C9gE2kVQPnAvsI2k0qRtnHnAyQETMljSVdHF3BXBqRLyX53MacBfQB5gUEbM7fW3MzKxmtdwFdHSV5GtbyH8RcFGV9DuAO9pUOjMz6zL+JbCZWUk5AJiZlZQDgJlZSbXrNlAzK6fLpv+lR5b71f236ZHl9nYOAGZmLejNQc9dQGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5V8CW6foqV9Lmln7uQVgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZS/h2A2VrGv7mwzuIWgJlZSTkAmJmVlAOAmVlJtXoNQNIk4BBgUUTsmNM2Bn4FjADmAV+IiNckCbgcGAv8HTghImbmaY4Hvp1n+92ImNy5q2JmvZWve3SNWloA1wEHVaRNAO6JiFHAPXkY4GBgVH6NB66GDwLGucBuwK7AuZI26mjhzcys/VoNABHxAPBqRfKhQNMZ/GTgsEL69ZE8BAyUNAQ4EJgeEa9GxGvAdFYPKmZm1o3aew1gs4hoAMjvm+b0ocD8Qr76nNZc+mokjZc0Q9KMxsbGdhbPzMxa09kXgVUlLVpIXz0xYmJE1EVE3eDBgzu1cGZmtlJ7fwi2UNKQiGjIXTyLcno9MLyQbxiwIKfvU5F+XzuXbbZG8IVJW9u1NwBMA44HLs7vtxXST5M0hXTB9/UcJO4Cvle48HsAcHb7i10b76BmZs2r5TbQm0hn75tIqifdzXMxMFXSicBLwBE5+x2kW0Dnkm4DHQcQEa9KuhB4NOe7ICIqLyybmVk3ajUARMTRzYzat0reAE5tZj6TgEltKp2ZmXUZ/xLYzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSqq9fwlptkbY/aWJPbbsh7YY32PLNusMbgGYmZWUA4CZWUk5AJiZlZQDgJlZSfkicBfwhUkzWxu4BWBmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZSHQoAkuZJekrSLEkzctrGkqZLei6/b5TTJekKSXMlPSlpTGesgJmZtU9n3Ab66YhYXBieANwTERdLmpCHvwkcDIzKr92Aq/O7daKeugXVt5+arX26ogvoUGBy/jwZOKyQfn0kDwEDJQ3pguWbmVkNOhoAAvitpMckNZ0CbhYRDQD5fdOcPhSYX5i2PqetQtJ4STMkzWhsbOxg8czMrDkd7QLaMyIWSNoUmC7pzy3kVZW0WC0hYiIwEaCurm618WZm1jk61AKIiAX5fRFwK7ArsLCpaye/L8rZ64HhhcmHAQs6snwzM2u/dgcASetLGtD0GTgAeBqYBhyfsx0P3JY/TwOOy3cD7Q683tRVZGZm3a8jXUCbAbdKaprPjRHx/yU9CkyVdCLwEnBEzn8HMBaYC/wdGNeBZZuZWQe1OwBExAvATlXSlwD7VkkP4NT2Ls/MzDqXfwlsZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5T+FN7Oa+WmzvYtbAGZmJeUWgJlZC3qq1QM/6PIlOACYrWV67oBkvY0DgHUKH5SsK3n76hq+BmBmVlJuAZi1k89KbW3XqwOAd1Azs+a5C8jMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKS6PQBIOkjSs5LmSprQ3cs3M7OkWwOApD7AVcDBwPbA0ZK2784ymJlZ0t0tgF2BuRHxQkS8A0wBDu3mMpiZGd3/hzBDgfmF4Xpgt2IGSeOB8XnwDUnPdmB5mwCLOzB9b+K6WJXrY1Wuj5XWjLo46YcdmXrLWjJ1dwBQlbRYZSBiItApf+UlaUZE1HXGvNZ2rotVuT5W5fpYqUx10d1dQPXA8MLwMGBBN5fBzMzo/gDwKDBK0khJ6wFHAdO6uQxmZkY3dwFFxApJpwF3AX2ASRExuwsX6X+FX8l1sSrXx6pcHyuVpi4UEa3nMjOzXse/BDYzKykHADOzkuqVAaDsj5uQNEnSIklPF9I2ljRd0nP5faOeLGN3kTRc0r2S5kiaLen0nF7W+ugn6RFJT+T6OD+nj5T0cK6PX+WbNEpBUh9Jj0u6PQ+Xpi56XQDw4yYAuA44qCJtAnBPRIwC7snDZbACODMitgN2B07N20NZ6+Nt4DMRsRMwGjhI0u7AJcBluT5eA07swTJ2t9OBOYXh0tRFrwsA+HETRMQDwKsVyYcCk/PnycBh3VqoHhIRDRExM39eTtrRh1Le+oiIeCMP9s2vAD4D3JzTS1MfkoYB/wf4WR4WJaqL3hgAqj1uYmgPlWVNsllENEA6KAKb9nB5up2kEcAngIcpcX3kLo9ZwCJgOvA8sDQiVuQsZdpnfgR8A3g/Dw+iRHXRGwNAq4+bsPKRtAFwC3BGRCzr6fL0pIh4LyJGk36JvyuwXbVs3Vuq7ifpEGBRRDxWTK6StdfWRXc/C6g7+HET1S2UNCQiGiQNIZ39lYKkvqSD/y8j4tc5ubT10SQilkq6j3RtZKCkdfOZb1n2mT2Bz0oaC/QDNiS1CEpTF72xBeDHTVQ3DTg+fz4euK0Hy9Jtcp/utcCciLi0MKqs9TFY0sD8uT+wH+m6yL3A4TlbKeojIs6OiGERMYJ0nPhdRHyREtVFr/wlcI7oP2Ll4yYu6uEidStJNwH7kB5ruxBlGl/tAAAAgUlEQVQ4F/gNMBXYAngJOCIiKi8U9zqS9gJ+DzzFyn7ec0jXAcpYH/9IurDZh3QCODUiLpC0FemGiY2Bx4FjIuLtnitp95K0D3BWRBxSprrolQHAzMxa1xu7gMzMrAYOAGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlL/C1Dl1pwt/VgfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f92f4452e48>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pyplot.hist(classes_train, alpha=0.5, label='Train')\n",
    "pyplot.hist(classes_test, alpha=0.5, label='Test')\n",
    "pyplot.legend(loc='upper right')\n",
    "pyplot.title(\"Class (Primary Site) distribution between train and test\")\n",
    "pyplot.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYcAAAEICAYAAAC0+DhzAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3X+cVXW97/HXW0BBRUHAHzGcQCUTKQsnxeyUhSFSRzz3auEjj0R4uXasU/ZT857wR3q0X5RpejmKoqdUskxuVyPyR/ZLZFREgYjJDCZRBhC0FBX9nD/Wd3Q7a8/sPXvvmQ3M+/l47Mes9V3f9V2f79pr9mev71qzRhGBmZlZoV3qHYCZmW1/nBzMzCzHycHMzHKcHMzMLMfJwczMcpwczMwsx8nBakbSf0k6v8y6B0uKgvlfSPpYjeI4VtLygvkWScfWou3U3ipJ/1ir9gra/Y2kj9e63Z2JpGskfWU7iOMMSffWO47u1KuTg6S/FbxelfRCwXxNPqhqRVJfSSFpqSQVlF8q6Zp6xlYLETExIn7QWZ2CfTCyRFv3RsRhtYirWMKLiEMi4te1aL9WdoTEUoskHRFnRMQlNQqpW7T/4lNlW2V/4aq1Xp0cImLPthewBvingrJOP6i6k6S+nSweAZzSzdvYYe2s/eoN/N5tX3p1ciilfdaWdJykJwrmWyR9QdJj6WxjjqT9JC2U9GwaKhlUUP8kScslbZZ0t6RD2rX1RUmPAs93EtbXgQsl9ekg5i5toyt9kLSLpFslPZXav1fSoWXuyz6SZkvaKOlPwKR2y1/75ivpLZLuk7RF0gZJP0zV7ks/l6dY/2fbeyLpK5KeAv6z/fuUHCVppaRnJF0rabe0rTcMDxSenUj6V+CjwFfS9m4r2GfHpun+ki6XtE7SXyV9W9KuaVlbbF+S1CrpSUmnl9hVoyU1pb7fJmlwQWzHSLo/7fulkt6byi8DjgauTnF+R9LFkman5btJ2irpkjS/Z5rfu7N207JBkq5L/WuRdKGkXQr23a/S+7pZ0uOSJnbw/t8EvAm4M8X4OaVv2JKmS1oD/KLUMaaC38mu7t8U70pJz0n6k6QzCpZ12pakYZJ+ln4n7gdGdfIe3pfWaRuFeFfB9v+QjsE7JY1I5bukY2h9et+XSRrT0fHXYyLCr+wRIk8Ax7Ur+y/g/IL544AnCuZbgN8B+wINwEagCTgc6A/8Cjgv1T0U+BvwAaAf8BXgj0C/grYeTO0MKBJfXyDIDsqlwMdT+aXANZVuo4t92AX4ODAwLbsCaOpof7WL/1PA8rSNIWS/QFGw/DcFffoR8OW0vf7AMe32wch278k24BJgV2BAB+/TsrTtocD9bXECZwD3FtnPIzvqU2rv2DR9Sdp/w9I+XAzMahfbrPR+nAj8Hdirg330G2AtMAbYA/gpcH1aNiK9N8en/TIJ2AAMab//0vxE4OE0/V7gT8BvC5Y9WGa7PwO+D+wO7E92/Mwo2HcvA58A+gCfBtZ28jv22n5L8wenfX1dan8AXTjGKti//wQcCIjsd+QF4O3ltAXcCtyU4nw7sK7wuGm3nYMpOLZT2cnAKuAQsmPsfODXadmHgAeAvVP/xwD7l/qd6u6Xzxyq992IWB8RLWS/oL+PiEciYivZL/c7U72pwIKIuDsiXib7UN8LOKpdWy0R8UIn2wvgq8AsSf3aLat0G2X1ISJejYjrI+K5tOx84AhJe5Sxnz4CzE7b3phi68jLwEjggIjYGhG/LdH2NrJfoJc62XeXp21vIPtAP7WMmMvxsbTt1ohYD1wI/EvB8q3A1yLi5YhYALwIvKWT9uZFxIqI+DvZ+zxVkoDTyd7bhel9+DnwCO3OwAr8Fhij7KzvvcAcYJSk3YH3kSV9OmtX0nBgAnB2RDwfEU8B3yE7ztr8KSLmRsQrwDygQdLQ0rvtDWal9l+o4Bgre/9GxP+LiMcjczdwF/CPpdpKv2cnAf+e4lwG3NjFPv5v4JKIWBUR24CvAUemffwy2e/pW1OcK9K+risnh+o9XTD9QpH5PdP0m4C/tC2IiFfJvkkNL6i/tpwNpgP3abJvboUq3UZZfVA2NPT1NHzwLNCc6pTzYfCmdtv+S0cVgc+TfXtrkvSopGkl2n46Il4qUaf9tt9Uon65DuCNffkLb9zfG9IHZ5vnef2YKKZ9nLsB+wBvBk5NQy2bJW0GxtNBP1JyeZgsMbwXuJfsjOlo3pgcOmv3zWn7TxcsuxLYr2BThR9ibcOhnfWv0z5XcIyVvX8lfVjSYkmbUl8mtmu3o7b2IzszKvf4LebNwJUF+3ED8CrQEBG/AK4GriLb11dLGtjF9mvOyaFzfyc7jWyzfxVtPUl2gADZOCPZMMdfC+p05Q6H/5NeA7pxG+2dDkwmOyXfm+z0GbLT9FLWkQ1htPmHjipGxLrI7ko5ADgLmCNpFB3HXk6f2m/7yTRd6j0u1fY6CvZ5avuvHdQtR/s4XwQ2kX0wXRcRgwpee0TENzqJ81dkwyVvIxsO+hVwAnAE0Ha3VWftriX7gNynYNleEfH2CvtWdF9GGj9JqjnGOiRpANnQ0H8A+0XEIOAXZbb7NNkHeVnHL8X7uZZsOK5wPw+IiMUAEfGdiBgHjCUbVvpcJ231CCeHzi0FPiRpsKQDgH+roq35wInK7sHvB3wReI5sjLrLIuKXZNcTTuuubRQxkOzDaiPZB+rFXVh3PvBZScMlDSG7plCUpI+k022AzWS/IK+kb3UbycaNu+pTBds+F7gllT8CvF3S29IHyKx26z1dYns3AV+VNFTSMODfycaJK3W6pLemYZQLgPnpw/NG4J8lfTB9u+4v6f2S2s4cisX5K7Lx+2VpKONeYCbwx4jYlOp02G5ErE1tfFPSXunC6cEquGDdRaX2JVR3jHVmN7JrUq3AK5I+TDZkVlIaov0pcIGkAZLG8sahw/bWAyGpsK9XA+e1XVxXdqH/5DR9ZHr1Jfuy8hLQdgZTzj7rFk4OnbseWEl2Cvlz4OZKG4qI5cA0slPHVrKx4hPTgVep88iGHLpzG4WuI/vG/STZxeXfdWHdq8jGeB8FlpB9i+vIUcASSX8HfgKcFRFr0rJZwA/T6fn/6ML2bwJ+SXZhdhXZdQciYkWavjeV39duvWuAw9MdJsVivoAswTxKdtF7Mdm300rdSJZc1pENZXw2xfkE8M9kyaeV7Nbrz/P67/B3eH146Nup7DdkH7BtfVpGNr79Wh/LaPc0sovjK4BnyG4WqPQM+hKyD9jNkj7bQZ1qjrEORcRm4GzgNrIzsZPJLraX65PAYLIP62tTnB1t6zmyY2Bx6mtjRPwI+DbwozRctozsJgCAQanNzWQ3xqwDZqdlpY6/bqM3ntGZmZn5zMHMzIpwcjAzsxwnBzMzy3FyMDOznB32QVdDhw6NkSNH1jsMM7MdyoMPPrghIoaVqrfDJoeRI0fS1NRU7zDMzHYoksr6624PK5mZWY6Tg5mZ5Tg5mJlZzg57zcHMrFwvv/wyLS0tbN26td6h9Jj+/fvT0NBAv37tn+xfHicHM9vptbS0MHDgQEaOHIlU1QNedwgRwcaNG2lpaWHUqM7+aV3HPKxkZju9rVu3MmTIkF6RGAAkMWTIkKrOlJwczKxX6C2JoU21/XVyMDOzHF9zMLNeZ/aiP9a0vbM/2Nm/BYeNGzcyYUL2v4Weeuop+vTpw7Bh2R8pP/DAA+y6664ltzF9+nTOOeccDjnkkOoDLkOvTA61PjDKVeoAMrOd05AhQ1i6dCkA559/PnvuuSdf+MIX3lAnIogIdtml+IDOddd1+P+FuoWHlczM6qS5uZmxY8dy5plnMm7cONatW8fMmTNpbGzksMMO48ILL3yt7nve8x6WLl3Ktm3bGDRoEOeccw6HH344Rx99NOvXr695bE4OZmZ1tGLFCmbMmMHDDz/M8OHDufTSS2lqauKRRx5h0aJFrFixIrfOli1beN/73scjjzzC0Ucfzdy5c2sel5ODmVkdHXTQQbzrXe96bf6mm25i3LhxjBs3jpUrVxZNDgMGDOCEE04A4IgjjuCJJ56oeVy98pqDmdn2Yo899nhtevXq1Xz3u9/lgQceYNCgQZx22mlF/1ah8AJ2nz592LZtW83j8pmDmdl24tlnn2XgwIHstdderFu3joULF9YtFp85mFmvs73eOThu3DjGjBnD2LFjOfDAAznmmGPqFosiom4br0ZjY2NU+s9+fCurWe+ycuVKDj300HqH0eOK9VvSgxHRWGpdDyuZmVlOyeQgaa6k9ZIeK7LsC5JC0tA0L0mXS2qWtEzSuIK60yStTq9pBeVHSHo0rXO5etsDUMzMtkPlnDlcD0xqXyhpBPBBYE1B8QnA6PSaCVyV6u4DzAKOAo4EZkkanNa5KtVtWy+3LTMz61klk0NE3AdsKrJoNvAloPCixRTghsjcDwySdABwPLAoIjZFxDPAImBSWrZXRPw+sosfNwAnVdclMzOrVkXXHCSdCPw1Ih5pt2g4sLZgviWVdVbeUqS8o+3OlNQkqam1tbWS0M3MrAxdTg6SdgfOA75abHGRsqigvKiImBMRjRHR2PZEQzMzq71K/s7hIGAU8Ei6dtwAPCTpSLJv/iMK6jYAT6byY9uV35vKG4rUNzPrPvf8R23be/+5nS6uxSO7AebOncvkyZPZf//9q4u3DF0+c4iIRyNi34gYGREjyT7gx0XEU8AC4PR019J4YEtErAMWAhMlDU4XoicCC9Oy5ySNT3cpnQ7cXqO+mZltF9oe2b106VLOPPNMzj777Nfmy00MkCWHp556qhsjfV3JMwdJN5F96x8qqQWYFRHXdlD9DmAy0Aw8D0wHiIhNki4ClqR6F0ZE20XuT5LdETUAuDO9zMx6hXnz5nHllVfy0ksv8e53v5srrriCV199lenTp7N06VIigpkzZ7LffvuxdOlSPvrRjzJgwIAunXFUomRyiIhTSywfWTAdwFkd1JsL5J4rGxFNwNhScZiZ7Wwee+wxbrvtNn73u9/Rt29fZs6cyc0338xBBx3Ehg0bePTRRwHYvHkzgwYN4nvf+x5XXHEF73jHO7o9Nj9bycysTn75y1+yZMkSGhuzp1m88MILjBgxguOPP55Vq1bxmc98hsmTJzNx4sQej83JwcysTiKCT3ziE1x00UW5ZcuWLePOO+/k8ssv58c//jFz5szp0dj8bCUzszo57rjjmD9/Phs2bACyu5rWrFlDa2srEcEpp5zCBRdcwEMPPQTAwIEDee6553okNp85mFnvU+LW057ytre9jVmzZnHcccfx6quv0q9fP66++mr69OnDjBkziAgkcdlllwEwffp0zjjjjB65IO1HdvcgP7LbrD78yO7X+ZHdZmZWMScHMzPLcXIws15hRx1Cr1S1/XVyMLOdXv/+/dm4cWOvSRARwcaNG+nfv3/FbfhuJTPb6TU0NNDS0kJvetR///79aWhoKF2xA04OZrbT69evH6NGjap3GDsUDyuZmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeWUTA6S5kpaL+mxgrJvSPqDpGWSbpM0qGDZuZKaJa2SdHxB+aRU1izpnILyUZIWS1ot6RZJ3feAcjMzK0s5Zw7XA5PalS0CxkbE24E/AucCSBoDTAUOS+t8X1IfSX2AK4ETgDHAqakuwGXA7IgYDTwDzKiqR2ZmVrWSySEi7gM2tSv7RURsS7P3A20P8JgC3BwRL0bEn4Fm4Mj0ao6IxyPiJeBmYIokAR8Abk3rzwNOqrJPZmZWpVpcc/gEcGeaHg6sLVjWkso6Kh8CbC5ING3lRUmaKalJUlNveoCWmVlPqyo5SDoP2Ab8oK2oSLWooLyoiJgTEY0R0Ths2LCuhmtmZmWq+KmskqYBHwYmxOsPSW8BRhRUawCeTNPFyjcAgyT1TWcPhfXNzKxOKjpzkDQJ+DJwYkQ8X7BoATBV0m6SRgGjgQeAJcDodGfSrmQXrRekpHIPcHJafxpwe2VdMTOzWinnVtabgN8Dh0hqkTQDuAIYCCyStFTS1QARsRyYD6wAfg6cFRGvpLOCTwELgZXA/FQXsiTzOUnNZNcgrq1pD83MrMtKDitFxKlFijv8AI+Ii4GLi5TfAdxRpPxxsruZzMxsO+G/kDYzsxwnBzMzy3FyMDOzHCcHMzPLcXIwM7McJwczM8txcjAzsxwnBzMzy3FyMDOzHCcHMzPLcXIwM7McJwczM8txcjAzsxwnBzMzy3FyMDOzHCcHMzPLcXIwM7McJwczM8txcjAzs5ySyUHSXEnrJT1WULaPpEWSVqefg1O5JF0uqVnSMknjCtaZluqvljStoPwISY+mdS6XpFp30szMuqacM4frgUntys4B7oqI0cBdaR7gBGB0es0EroIsmQCzgKOAI4FZbQkl1ZlZsF77bZmZWQ8rmRwi4j5gU7viKcC8ND0POKmg/IbI3A8MknQAcDywKCI2RcQzwCJgUlq2V0T8PiICuKGgLTMzq5NKrznsFxHrANLPfVP5cGBtQb2WVNZZeUuR8qIkzZTUJKmptbW1wtDNzKyUWl+QLna9ICooLyoi5kREY0Q0Dhs2rMIQzcyslEqTw9NpSIj0c30qbwFGFNRrAJ4sUd5QpNzMzOqo0uSwAGi742gacHtB+enprqXxwJY07LQQmChpcLoQPRFYmJY9J2l8ukvp9IK2zMysTvqWqiDpJuBYYKikFrK7ji4F5kuaAawBTknV7wAmA83A88B0gIjYJOkiYEmqd2FEtF3k/iTZHVEDgDvTy8zM6qhkcoiIUztYNKFI3QDO6qCducDcIuVNwNhScZiZWc/xX0ibmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlVJUcJJ0tabmkxyTdJKm/pFGSFktaLekWSbumurul+ea0fGRBO+em8lWSjq+uS2ZmVq2Kk4Ok4cC/AY0RMRboA0wFLgNmR8Ro4BlgRlplBvBMRBwMzE71kDQmrXcYMAn4vqQ+lcZlZmbVq3ZYqS8wQFJfYHdgHfAB4Na0fB5wUpqekuZJyydIUiq/OSJejIg/A83AkVXGZWZmVag4OUTEX4FvAmvIksIW4EFgc0RsS9VagOFpejiwNq27LdUfUlheZB0zM6uDaoaVBpN96x8FvAnYAzihSNVoW6WDZR2VF9vmTElNkppaW1u7HrSZmZWlmmGl44A/R0RrRLwM/AR4NzAoDTMBNABPpukWYARAWr43sKmwvMg6bxARcyKiMSIahw0bVkXoZmbWmWqSwxpgvKTd07WDCcAK4B7g5FRnGnB7ml6Q5knL746ISOVT091Mo4DRwANVxGVmZlXqW7pKcRGxWNKtwEPANuBhYA7w/4GbJX0tlV2bVrkWuFFSM9kZw9TUznJJ88kSyzbgrIh4pdK4zMysehUnB4CImAXMalf8OEXuNoqIrcApHbRzMXBxNbGYmVnt+C+kzcwsx8nBzMxynBzMzCzHycHMzHKcHMzMLMfJwczMcpwczMwsx8nBzMxynBzMzCzHycHMzHKcHMzMLMfJwczMcpwczMwsx8nBzMxynBzMzCzHycHMzHKcHMzMLMfJwczMcpwczMwsp6rkIGmQpFsl/UHSSklHS9pH0iJJq9PPwamuJF0uqVnSMknjCtqZluqvljSt2k6ZmVl1qj1z+C7w84h4K3A4sBI4B7grIkYDd6V5gBOA0ek1E7gKQNI+wCzgKOBIYFZbQjEzs/qoODlI2gt4L3AtQES8FBGbgSnAvFRtHnBSmp4C3BCZ+4FBkg4AjgcWRcSmiHgGWARMqjQuMzOrXjVnDgcCrcB1kh6WdI2kPYD9ImIdQPq5b6o/HFhbsH5LKuuoPEfSTElNkppaW1urCN3MzDpTTXLoC4wDroqIdwJ/5/UhpGJUpCw6Kc8XRsyJiMaIaBw2bFhX4zUzszJVkxxagJaIWJzmbyVLFk+n4SLSz/UF9UcUrN8APNlJuZmZ1UnFySEingLWSjokFU0AVgALgLY7jqYBt6fpBcDp6a6l8cCWNOy0EJgoaXC6ED0xlZmZWZ30rXL9TwM/kLQr8DgwnSzhzJc0A1gDnJLq3gFMBpqB51NdImKTpIuAJanehRGxqcq4zMysClUlh4hYCjQWWTShSN0AzuqgnbnA3GpiMTOz2vFfSJuZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeU4OZiZWY6Tg5mZ5Tg5mJlZjpODmZnlODmYmVmOk4OZmeU4OZiZWU7VyUFSH0kPS/pZmh8labGk1ZJukbRrKt8tzTen5SML2jg3la+SdHy1MZmZWXVqcebwGWBlwfxlwOyIGA08A8xI5TOAZyLiYGB2qoekMcBU4DBgEvB9SX1qEJeZmVWoquQgqQH4EHBNmhfwAeDWVGUecFKanpLmScsnpPpTgJsj4sWI+DPQDBxZTVxmZladas8cvgN8CXg1zQ8BNkfEtjTfAgxP08OBtQBp+ZZU/7XyIuu8gaSZkpokNbW2tlYZupmZdaTi5CDpw8D6iHiwsLhI1SixrLN13lgYMSciGiOicdiwYV2K18zMyte3inWPAU6UNBnoD+xFdiYxSFLfdHbQADyZ6rcAI4AWSX2BvYFNBeVtCtcxM7M6qPjMISLOjYiGiBhJdkH57oj4GHAPcHKqNg24PU0vSPOk5XdHRKTyqeluplHAaOCBSuMyM7PqVXPm0JEvAzdL+hrwMHBtKr8WuFFSM9kZw1SAiFguaT6wAtgGnBURr3RDXK8Zv2ZOdzbfiW/WabtmZl1Tk+QQEfcC96bpxylyt1FEbAVO6WD9i4GLaxGLmZlVz38hbWZmOU4OZmaW4+RgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmOU4OZmaW4+RgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmOU4OZmaW4+RgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmORUnB0kjJN0jaaWk5ZI+k8r3kbRI0ur0c3Aql6TLJTVLWiZpXEFb01L91ZKmVd8tMzOrRjVnDtuAz0fEocB44CxJY4BzgLsiYjRwV5oHOAEYnV4zgasgSybALOAo4EhgVltCMTOz+qg4OUTEuoh4KE0/B6wEhgNTgHmp2jzgpDQ9BbghMvcDgyQdABwPLIqITRHxDLAImFRpXGZmVr2aXHOQNBJ4J7AY2C8i1kGWQIB9U7XhwNqC1VpSWUflxbYzU1KTpKbW1tZahG5mZkVUnRwk7Qn8GPhsRDzbWdUiZdFJeb4wYk5ENEZE47Bhw7oerJmZlaWq5CCpH1li+EFE/CQVP52Gi0g/16fyFmBEweoNwJOdlJuZWZ1Uc7eSgGuBlRHx7YJFC4C2O46mAbcXlJ+e7loaD2xJw04LgYmSBqcL0RNTmZmZ1UnfKtY9BvgX4FFJS1PZV4BLgfmSZgBrgFPSsjuAyUAz8DwwHSAiNkm6CFiS6l0YEZuqiMvMzKpUcXKIiN9Q/HoBwIQi9QM4q4O25gJzK43FzMxqy38hbWZmOU4OZmaW4+RgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmOU4OZmaW4+RgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmOU4OZmaW4+RgZmY5Tg5mZpZTzX+CMzPrtWYv+mNdtnv2B9/SI9vxmYOZmeU4OZiZWc52kxwkTZK0SlKzpHPqHY+ZWW+2XVxzkNQHuBL4INACLJG0ICJW1DcyM7Pixq+ZU6ctf7NHtrK9nDkcCTRHxOMR8RJwMzClzjGZmfVa28WZAzAcWFsw3wIc1b6SpJnAzDT7N0mrKtzeUGBDhetW7oxv9fgmC9Snz/XlPu/8elt/4YxvVdvnN5dTaXtJDipSFrmCiDlA1edykpoiorHadnYk7nPv0Nv63Nv6Cz3X5+1lWKkFGFEw3wA8WadYzMx6ve0lOSwBRksaJWlXYCqwoM4xmZn1WtvFsFJEbJP0KWAh0AeYGxHLu3GT9brNoJ7c596ht/W5t/UXeqjPisgN7ZuZWS+3vQwrmZnZdsTJwczMcnbq5FDqkRySdpN0S1q+WNLIno+ydsro7+ckrZC0TNJdksq633l7Vu5jVySdLCkk7fC3PZbTZ0kfSe/1ckk/7OkYa62MY/sfJN0j6eF0fE+uR5y1ImmupPWSHutguSRdnvbHMknjah5EROyUL7IL238CDgR2BR4BxrSr86/A1Wl6KnBLvePu5v6+H9g9TX9yR+5vuX1O9QYC9wH3A431jrsH3ufRwMPA4DS/b73j7oE+zwE+mabHAE/UO+4q+/xeYBzwWAfLJwN3kv2N2Hhgca1j2JnPHMp5JMcUYF6avhWYIKnYH+TtCEr2NyLuiYjn0+z9ZH9PsiMr97ErFwFfB7b2ZHDdpJw+/y/gyoh4BiAi1vdwjLVWTp8D2CtN780O/ndSEXEfsKmTKlOAGyJzPzBI0gG1jGFnTg7FHskxvKM6EbEN2AIM6ZHoaq+c/haaQfbNY0dWss+S3gmMiIif9WRg3aic9/ktwFsk/VbS/ZIm9Vh03aOcPp8PnCapBbgD+HTPhFY3Xf1977Lt4u8cukk5j+Qo67EdO4iy+yLpNKAReF+3RtT9Ou2zpF2A2cDHeyqgHlDO+9yXbGjpWLKzw19LGhsRm7s5tu5STp9PBa6PiG9JOhq4MfX51e4Pry66/bNrZz5zKOeRHK/VkdSX7HS0s1O57VlZjyCRdBxwHnBiRLzYQ7F1l1J9HgiMBe6V9ATZ2OyCHfyidLnH9e0R8XJE/BlYRZYsdlTl9HkGMB8gIn4P9Cd7KN/OqtsfObQzJ4dyHsmxAJiWpk8G7o50tWcHVLK/aYjl/5Ilhh19HBpK9DkitkTE0IgYGREjya6znBgRTfUJtybKOa5/SnbzAZKGkg0zPd6jUdZWOX1eA0wAkHQoWXJo7dEoe9YC4PR019J4YEtErKvlBnbaYaXo4JEcki4EmiJiAXAT49KVAAAAmklEQVQt2elnM9kZw9T6RVydMvv7DWBP4EfpuvuaiDixbkFXqcw+71TK7PNCYKKkFcArwBcjYmP9oq5OmX3+PPCfks4mG175+A78RQ9JN5ENCw5N11FmAf0AIuJqsusqk4Fm4Hlges1j2IH3n5mZdZOdeVjJzMwq5ORgZmY5Tg5mZpbj5GBmZjlODmZmluPkYGZmOU4OZmaW89/bToloBV9SXgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f92f4c85630>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pyplot.hist(y_train[:, 0], alpha=0.5, label='Train')\n",
    "pyplot.hist(y_test[:, 0], alpha=0.5, label='Test')\n",
    "pyplot.legend(loc='upper right')\n",
    "pyplot.title(\"Tumor Normal distribution between train and test\")\n",
    "pyplot.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tumor Normal Binary Classification\n",
    "First train a basic binary tumor vs. normal classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "input_4 (InputLayer)         (None, 5172)              0         \n",
      "_________________________________________________________________\n",
      "batch_normalization_4 (Batch (None, 5172)              20688     \n",
      "_________________________________________________________________\n",
      "dense_13 (Dense)             (None, 32)                165536    \n",
      "_________________________________________________________________\n",
      "dropout_195 (Dropout)        (None, 32)                0         \n",
      "_________________________________________________________________\n",
      "dense_14 (Dense)             (None, 8)                 264       \n",
      "_________________________________________________________________\n",
      "dropout_196 (Dropout)        (None, 8)                 0         \n",
      "_________________________________________________________________\n",
      "dense_15 (Dense)             (None, 1)                 9         \n",
      "=================================================================\n",
      "Total params: 186,497\n",
      "Trainable params: 176,153\n",
      "Non-trainable params: 10,344\n",
      "_________________________________________________________________\n",
      "None\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 17s 1ms/step - loss: 0.5994 - acc: 0.8104\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 3s 204us/step - loss: 0.2569 - acc: 0.9484\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 3s 204us/step - loss: 0.1785 - acc: 0.9731\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 3s 199us/step - loss: 0.1527 - acc: 0.9780\n",
      "Epoch 00004: early stopping\n",
      "3826/3826 [==============================] - 6s 2ms/step\n",
      "['loss', 'acc'] [0.032549997350571985, 0.9835337165507483]\n"
     ]
    }
   ],
   "source": [
    "from keras.models import Model, Sequential\n",
    "from keras.layers import InputLayer, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "classify = [\n",
    "    InputLayer(input_shape=(X_train.shape[1],)),\n",
    "    BatchNormalization(),\n",
    "\n",
    "    Dense(32, activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(8, activity_regularizer=regularizers.l1(1e-5), activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(1, activation='sigmoid')\n",
    "]\n",
    "\n",
    "model = Sequential(classify)\n",
    "print(model.summary())\n",
    "\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\")]\n",
    "model.fit(X_train, y_train[:,0], epochs=epochs, batch_size=batch_size, shuffle=\"batch\", callbacks=callbacks)\n",
    "\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test[:,0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multi-label Tumor vs. Normal and Primary Site Classification\n",
    "Multi-label classification of tumor/normal + primary site with union of genes from all gene sets as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "input_6 (InputLayer)         (None, 5172)              0         \n",
      "_________________________________________________________________\n",
      "batch_normalization_6 (Batch (None, 5172)              20688     \n",
      "_________________________________________________________________\n",
      "dense_22 (Dense)             (None, 32)                165536    \n",
      "_________________________________________________________________\n",
      "dropout_387 (Dropout)        (None, 32)                0         \n",
      "_________________________________________________________________\n",
      "dense_23 (Dense)             (None, 16)                528       \n",
      "_________________________________________________________________\n",
      "dropout_388 (Dropout)        (None, 16)                0         \n",
      "_________________________________________________________________\n",
      "dense_24 (Dense)             (None, 46)                782       \n",
      "=================================================================\n",
      "Total params: 187,534\n",
      "Trainable params: 177,190\n",
      "Non-trainable params: 10,344\n",
      "_________________________________________________________________\n",
      "None\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 27s 2ms/step - loss: 0.5418 - acc: 0.7386\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 3s 210us/step - loss: 0.2197 - acc: 0.9492\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 3s 196us/step - loss: 0.1687 - acc: 0.9723\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 3s 196us/step - loss: 0.1543 - acc: 0.9758\n",
      "Epoch 00004: early stopping\n",
      "3826/3826 [==============================] - 12s 3ms/step\n",
      "['loss', 'acc'] [0.07185998140706233, 0.9805223416484468]\n",
      "CPU times: user 2min 56s, sys: 6min 35s, total: 9min 31s\n",
      "Wall time: 49.2 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from keras.models import Model, Sequential\n",
    "from keras.layers import InputLayer, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "classify = [\n",
    "    InputLayer(input_shape=(X_train.shape[1],)),\n",
    "    BatchNormalization(),\n",
    "\n",
    "    Dense(32, activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(16, activity_regularizer=regularizers.l1(1e-5), activation='relu'),\n",
    "    Dropout(0.5),\n",
    "    \n",
    "    Dense(y_train.shape[1], activation='sigmoid')\n",
    "]\n",
    "\n",
    "model = Sequential(classify)\n",
    "print(model.summary())\n",
    "\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\")]\n",
    "model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, shuffle=\"batch\", callbacks=callbacks)\n",
    "\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "X: GTEX-1117F-0226-SM-5GZZ7 0  vs  0.855168 0  vs  4\n",
      "Prediction Normal 32\n",
      "X: GTEX-1269C-2826-SM-5EQ5O 0  vs  0.04318888 6  vs  32\n",
      "Prediction Normal 7\n",
      "X: GTEX-13FTX-0526-SM-5IFGS 0  vs  0.013360138 14  vs  7\n",
      "Prediction Normal 13\n",
      "X: GTEX-145LS-1226-SM-5Q5D9 0  vs  0.018882453 22  vs  13\n",
      "Prediction Normal 21\n",
      "X: GTEX-POYW-1226-SM-5LZWQ 0  vs  0.00898424 22  vs  21\n",
      "Prediction Normal 21\n",
      "X: GTEX-T6MN-1226-SM-3NMA5 0  vs  0.03294729 21  vs  21\n",
      "Prediction Normal 20\n",
      "X: GTEX-X638-0126-SM-47JZ8 0  vs  0.028094813 33  vs  20\n",
      "Prediction Normal 32\n",
      "X: GTEX-ZDYS-1126-SM-5K7UB 0  vs  0.037653007 9  vs  32\n",
      "Prediction Normal 8\n",
      "X: TARGET-10-PARFTR-04 1  vs  0.038637772 45  vs  8\n",
      "Prediction Normal 44\n",
      "X: TCGA-24-1422-01 1  vs  0.0049079712 26  vs  44\n",
      "Prediction Normal 44\n",
      "X: TCGA-58-8387-01 1  vs  0.02075427 22  vs  44\n",
      "Prediction Normal 21\n",
      "X: TCGA-AA-3527-11 0  vs  0.030726403 12  vs  21\n",
      "Prediction Normal 11\n",
      "X: TCGA-BF-AAP0-06 1  vs  0.029990522 33  vs  11\n",
      "Prediction Normal 32\n",
      "X: TCGA-CJ-4639-01 1  vs  0.03260799 19  vs  32\n",
      "Prediction Normal 18\n",
      "X: TCGA-DD-A4NS-01 1  vs  0.011015654 21  vs  18\n",
      "Prediction Normal 20\n",
      "X: TCGA-EE-A2GS-06 1  vs  0.028263526 33  vs  20\n",
      "Prediction Normal 32\n",
      "X: TCGA-G2-A2EK-01 1  vs  0.034545165 4  vs  32\n",
      "Prediction Normal 32\n",
      "X: TCGA-KL-8332-11 0  vs  0.045479633 19  vs  32\n",
      "Prediction Normal 18\n",
      "X: TCGA-S9-A6WE-01 1  vs  0.02583432 8  vs  18\n",
      "Prediction Normal 7\n",
      "X: TCGA-ZB-A96E-01 1  vs  0.0021801456 40  vs  7\n",
      "Prediction Normal 18\n"
     ]
    }
   ],
   "source": [
    "ids = Y.index.values[0:20000:1000]\n",
    "\n",
    "for id in ids:\n",
    "    print(\"X:\", id, Y.loc[id].tumor_normal_value, \" vs \", prediction[0][0],\n",
    "          Y.loc[id].primary_site_value, \" vs \", np.argmax(prediction[0][1:]))\n",
    "    prediction = model.predict(np.array([X_pruned.loc[id].values]))\n",
    "    print(\"Prediction\", \"Tumor\" if prediction[0][0] >= 0.5 else \"Normal\", \n",
    "          np.argmax(prediction[0][1:]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mulit-label Classification w/Per Pathway Sub-Network Input Layer\n",
    "For each pathway build a custom input layer that extracts the expression levels for the genes in the pathway from the full input vector and feeds this into a dense single output hidden neuron. These are then aggregated and fed into a standard set of stacked layers and trained to classify tumor/norml and primary site. The hidden per pathway neurons are named after the pathway an 'may' indicate which pathway lead to a given classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pathway KEGG_ABC_TRANSPORTERS Gene Indexes: [  11   12   13   14   15   16   17   18   19   20   21   22   23   24\n",
      "   25   26   27   28   29   30   31   32   33   34   35   36   37   38\n",
      "   39   40   41   42   43   44   45   46   47   48   49   50   51  848\n",
      " 4692 4693]\n"
     ]
    }
   ],
   "source": [
    "# Convert list of gene names into list of indexes into the features of X_pruned\n",
    "gene_set_indexes = {name: np.searchsorted(X_pruned.columns.values, genes) for name, genes in gene_sets.items()}\n",
    "print(\"Pathway KEGG_ABC_TRANSPORTERS Gene Indexes:\", gene_set_indexes[\"KEGG_ABC_TRANSPORTERS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trainable params:: 16343\n",
      "CPU times: user 1min 33s, sys: 824 ms, total: 1min 34s\n",
      "Wall time: 1min 34s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from keras.models import Model, Sequential\n",
    "from keras.layers import Input, Lambda, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "from keras.layers.merge import concatenate\n",
    "from keras import backend as K\n",
    "import itertools\n",
    "\n",
    "main_input = Input(shape=(X_train.shape[1],), name=\"main_input\")\n",
    "\n",
    "\"\"\"\n",
    "Build per pathway sub-networks\n",
    "\"\"\"\n",
    "# Set to a small number (~4) for debugging, set to None to build all pathways\n",
    "# max_num_pathways = 4\n",
    "max_num_pathways = None \n",
    "\n",
    "# Extract the features/gene's for each pathway from the aggregate x input vector\n",
    "pathways = [Lambda(lambda x: K.concatenate([x[:, i:i+1] for i in indexes]))(main_input)\n",
    "            for name, indexes in itertools.islice(gene_set_indexes.items(), max_num_pathways)]\n",
    "\n",
    "# Add a dense hidden layer per pathway with width proportional to the number of genes in the pathway\n",
    "# pathways = [Dense(max(2, len(i)//4), activation='relu')(p) for p, i in zip(pathways, gene_set_indexes.values())]\n",
    "\n",
    "# Add a named binary output for each pathway \n",
    "# pathways = [Dense(1, activity_regularizer=regularizers.l1(1e-5), activation='relu', name=name)(p) \n",
    "#             for p, name in zip(pathways, gene_set_indexes.keys())]\n",
    "\n",
    "pathways = [Dropout(0.5)(Dense(1, activity_regularizer=regularizers.l1(1e-5), activation='relu', name=name)(p))\n",
    "            for p, name in zip(pathways, gene_set_indexes.keys())]\n",
    "\n",
    "# Concatenate the binary outputs of each of the sub-networks back into a single vector\n",
    "x = keras.layers.concatenate(pathways)\n",
    "           \n",
    "\"\"\"\n",
    "Add traditional stacked network for final multi-label classification\n",
    "\"\"\"\n",
    "x = Dense(16, activation='relu')(x)\n",
    "x = Dropout(0.5)(x)\n",
    "    \n",
    "x = Dense(8, activation='relu')(x)\n",
    "x = Dropout(0.5)(x)\n",
    "\n",
    "x = Dense(y_train.shape[1], activation='sigmoid')(x)\n",
    "\n",
    "model = Model(inputs=[main_input], outputs=[x])\n",
    "# print(model.summary())  # Too detailed when building full set of pathways\n",
    "print(\"Trainable params::\", np.sum([np.product(v.shape) for v in model.trainable_weights]).value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compiling and training...\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 33s 2ms/step - loss: 1.0994 - acc: 0.7331\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 13s 861us/step - loss: 0.5745 - acc: 0.8309\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 13s 870us/step - loss: 0.3620 - acc: 0.9177\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 14s 886us/step - loss: 0.2490 - acc: 0.9612\n",
      "Epoch 5/10\n",
      "15300/15300 [==============================] - 14s 935us/step - loss: 0.2063 - acc: 0.9719\n",
      "Epoch 6/10\n",
      "15300/15300 [==============================] - 14s 922us/step - loss: 0.1808 - acc: 0.9755\n",
      "Epoch 7/10\n",
      "15300/15300 [==============================] - 12s 813us/step - loss: 0.1628 - acc: 0.9771\n",
      "Epoch 00007: early stopping\n",
      "3826/3826 [==============================] - 30s 8ms/step\n",
      "['loss', 'acc'] [0.09926590386931958, 0.9782609340724556]\n",
      "CPU times: user 7min 39s, sys: 2min 18s, total: 9min 58s\n",
      "Wall time: 4min 14s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from time import time\n",
    "from keras.callbacks import TensorBoard\n",
    "tensorboard = TensorBoard(log_dir=\"data/logs/{}\".format(time()))\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "print(\"Compiling and training...\")\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, shuffle=\"batch\", \n",
    "          callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\"), tensorboard])\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Treehouse\n",
    "Evaluate the model on samples from the [Treehouse public compendium](https://treehousegenomics.soe.ucsc.edu/explore-our-data/)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expression: samples=60498 genes=19260\n",
      "CPU times: user 932 ms, sys: 2 s, total: 2.93 s\n",
      "Wall time: 2.92 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if not os.path.exists(\"data\"):\n",
    "    os.makedirs(\"data\")\n",
    "    \n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\"):\n",
    "    print(\"Downloading Treehouse Public Compendium\")\n",
    "    r = requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\")\n",
    "    r.raise_for_status()\n",
    "    with open(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    expression = pd.read_csv(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \n",
    "                             sep=\"\\t\", index_col=0).dropna().astype(np.float32)\n",
    "    expression.to_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "treehouse_expression = pd.read_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\").sort_index(axis=1)\n",
    "print(\"expression: samples={} genes={}\".format(*expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Gene</th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2ML1-AS1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoU2-30</th>\n",
       "      <th>snoU2_19</th>\n",
       "      <th>snoU83B</th>\n",
       "      <th>snoZ196</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>uc_338</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAKSWW-03</th>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-3.675801</td>\n",
       "      <td>5.631403</td>\n",
       "      <td>3.270715</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-0.886033</td>\n",
       "      <td>-2.634004</td>\n",
       "      <td>-2.931300</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.093503</td>\n",
       "      <td>-4.604685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXHJ-09</th>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>5.203204</td>\n",
       "      <td>3.016318</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-0.996812</td>\n",
       "      <td>0.537100</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.430414</td>\n",
       "      <td>-9.965784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXSP-09</th>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>5.406404</td>\n",
       "      <td>2.163832</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-0.284324</td>\n",
       "      <td>-2.931300</td>\n",
       "      <td>-3.456416</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.112369</td>\n",
       "      <td>-9.965784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-03</th>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>6.067202</td>\n",
       "      <td>1.628147</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.883068</td>\n",
       "      <td>-6.493342</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.724956</td>\n",
       "      <td>-9.965784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-04</th>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-4.365655</td>\n",
       "      <td>5.172804</td>\n",
       "      <td>1.856440</td>\n",
       "      <td>-5.006953</td>\n",
       "      <td>-1.148480</td>\n",
       "      <td>-3.306172</td>\n",
       "      <td>-1.684536</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-9.965784</td>\n",
       "      <td>-1.367746</td>\n",
       "      <td>-5.006953</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 58581 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Gene                  5S_rRNA  5_8S_rRNA       7SK      A1BG  A1BG-AS1  \\\n",
       "TARGET-10-PAKSWW-03 -9.965784  -9.965784 -3.675801  5.631403  3.270715   \n",
       "TARGET-10-PAMXHJ-09 -9.965784  -9.965784 -9.965784  5.203204  3.016318   \n",
       "TARGET-10-PAMXSP-09 -9.965784  -9.965784 -9.965784  5.406404  2.163832   \n",
       "TARGET-10-PANCVR-03 -9.965784  -9.965784 -9.965784  6.067202  1.628147   \n",
       "TARGET-10-PANCVR-04 -9.965784  -9.965784 -4.365655  5.172804  1.856440   \n",
       "\n",
       "Gene                     A1CF       A2M   A2M-AS1     A2ML1  A2ML1-AS1  \\\n",
       "TARGET-10-PAKSWW-03 -9.965784 -0.886033 -2.634004 -2.931300  -9.965784   \n",
       "TARGET-10-PAMXHJ-09 -9.965784 -0.996812  0.537100 -9.965784  -9.965784   \n",
       "TARGET-10-PAMXSP-09 -9.965784 -0.284324 -2.931300 -3.456416  -9.965784   \n",
       "TARGET-10-PANCVR-03 -9.965784 -1.883068 -6.493342 -9.965784  -9.965784   \n",
       "TARGET-10-PANCVR-04 -5.006953 -1.148480 -3.306172 -1.684536  -9.965784   \n",
       "\n",
       "Gene                    ...      snoU2-30  snoU2_19   snoU83B   snoZ196  \\\n",
       "TARGET-10-PAKSWW-03     ...     -9.965784 -9.965784 -9.965784 -9.965784   \n",
       "TARGET-10-PAMXHJ-09     ...     -9.965784 -9.965784 -9.965784 -9.965784   \n",
       "TARGET-10-PAMXSP-09     ...     -9.965784 -9.965784 -9.965784 -9.965784   \n",
       "TARGET-10-PANCVR-03     ...     -9.965784 -9.965784 -9.965784 -9.965784   \n",
       "TARGET-10-PANCVR-04     ...     -9.965784 -9.965784 -9.965784 -9.965784   \n",
       "\n",
       "Gene                  snoZ278    snoZ40     snoZ6  snosnR66    uc_338  \\\n",
       "TARGET-10-PAKSWW-03 -9.965784 -9.965784 -9.965784 -9.965784 -1.093503   \n",
       "TARGET-10-PAMXHJ-09 -9.965784 -9.965784 -9.965784 -9.965784 -1.430414   \n",
       "TARGET-10-PAMXSP-09 -9.965784 -9.965784 -9.965784 -9.965784 -1.112369   \n",
       "TARGET-10-PANCVR-03 -9.965784 -9.965784 -9.965784 -9.965784 -1.724956   \n",
       "TARGET-10-PANCVR-04 -9.965784 -9.965784 -9.965784 -9.965784 -1.367746   \n",
       "\n",
       "Gene                 yR211F11.2  \n",
       "TARGET-10-PAKSWW-03   -4.604685  \n",
       "TARGET-10-PAMXHJ-09   -9.965784  \n",
       "TARGET-10-PAMXSP-09   -9.965784  \n",
       "TARGET-10-PANCVR-03   -9.965784  \n",
       "TARGET-10-PANCVR-04   -5.006953  \n",
       "\n",
       "[5 rows x 58581 columns]"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Convert into same normalized TPM: X log2(TPM+0.001)  XTreehouse log2(TPM+1) and invert to match X\n",
    "XTreehouse = treehouse_expression.T.apply(lambda x: np.log2((np.power(2, x) - 1.0) + 0.001))\n",
    "\n",
    "# Check that the features/genes are identical and in the same order\n",
    "assert np.array_equal(X.columns.values, XTreehouse.columns.values)\n",
    "\n",
    "XTreehouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.542399833268772 3.5424123\n",
      "5.848000021348486 5.848003\n",
      "2.8819000795903658 2.8819194\n",
      "3.802300007434838 3.8023105\n",
      "5.156300072229567 5.1563044\n",
      "3.2495998806503223 3.249615\n",
      "2.411700016129224 2.411727\n",
      "3.9346000910563292 3.9346094\n",
      "1.438799972912041 1.4388531\n"
     ]
    }
   ],
   "source": [
    "# Check to see if the common samples at this point actually d\n",
    "# X log2(TPM+0.001)  XTreehouse log2(TPM+1)\n",
    "for sample in [\"TARGET-10-PAMXHJ-09\", \"TARGET-10-PANCVR-04\", \"TARGET-10-PAMXSP-09\"]:\n",
    "    for gene in [\"AAGAB\", \"TP53\", \"BRCA1\"]:\n",
    "        print(X.loc[sample][gene], XTreehouse.loc[sample][gene])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_in_years</th>\n",
       "      <th>gender</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAKSWW-03</th>\n",
       "      <td>15.11</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXHJ-09</th>\n",
       "      <td>6.08</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXSP-09</th>\n",
       "      <td>3.35</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-03</th>\n",
       "      <td>6.4</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-04</th>\n",
       "      <td>6.4</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    age_in_years gender                       disease\n",
       "id                                                                   \n",
       "TARGET-10-PAKSWW-03        15.11   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PAMXHJ-09         6.08   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PAMXSP-09         3.35   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PANCVR-03          6.4   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PANCVR-04          6.4   male  acute lymphoblastic leukemia"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the clinical data for Treehouse\n",
    "import requests\n",
    "if not os.path.exists(\"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\"):\n",
    "    with open(\"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\").content)\n",
    "\n",
    "YTreehouse = pd.read_table(\"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\", compression=\"gzip\",\n",
    "                  header=0, names=[\"id\", \"age_in_years\", \"gender\", \"disease\"],\n",
    "                  sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=0)\n",
    "\n",
    "YTreehouse.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_in_years</th>\n",
       "      <th>gender</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TARGET-20-PANTNA-09</th>\n",
       "      <td>17.58</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-21-PAMYMA-09</th>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-21-PANVPB-09</th>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-21-PANZLR-09</th>\n",
       "      <td>NaN</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-50-PAJNAA-01</th>\n",
       "      <td>9</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-50-PAJNAV-01</th>\n",
       "      <td>4</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-50-PAKNAL-01</th>\n",
       "      <td>6</td>\n",
       "      <td>unknown</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    age_in_years   gender disease\n",
       "id                                               \n",
       "TARGET-20-PANTNA-09        17.58  unknown     NaN\n",
       "TARGET-21-PAMYMA-09          NaN  unknown     NaN\n",
       "TARGET-21-PANVPB-09          NaN  unknown     NaN\n",
       "TARGET-21-PANZLR-09          NaN  unknown     NaN\n",
       "TARGET-50-PAJNAA-01            9  unknown     NaN\n",
       "TARGET-50-PAJNAV-01            4  unknown     NaN\n",
       "TARGET-50-PAKNAL-01            6  unknown     NaN"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# YTreehouse[YTreehouse.disease.dtype != np.float]\n",
    "YTreehouse[YTreehouse.disease.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "XTreehouse_pruned = XTreehouse.drop(labels=(set(XTreehouse.columns) - set(all_gene_set_genes)), axis=1, errors=\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "A2M        -0.9971\n",
       "A4GALT     -4.6082\n",
       "AACS        0.7916\n",
       "AADAT      -9.9658\n",
       "AANAT       1.6695\n",
       "AARS        3.3421\n",
       "AARS2       1.0293\n",
       "AASDH       2.3337\n",
       "AASDHPPT    2.8681\n",
       "AASS        2.2633\n",
       "Name: TARGET-10-PAMXHJ-09, dtype: float64"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_pruned.loc[\"TARGET-10-PAMXHJ-09\"][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Gene\n",
       "A2M        -0.996812\n",
       "A4GALT     -4.604685\n",
       "AACS        0.791683\n",
       "AADAT      -9.965784\n",
       "AANAT       1.669545\n",
       "AARS        3.342114\n",
       "AARS2       1.029371\n",
       "AASDH       2.333729\n",
       "AASDHPPT    2.868120\n",
       "AASS        2.263330\n",
       "Name: TARGET-10-PAMXHJ-09, dtype: float32"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ABCA13'"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"]- X_pruned.loc[\"TARGET-10-PAMXHJ-09\"]).idxmax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-6.493342 -6.506400067567295\n"
     ]
    }
   ],
   "source": [
    "print(XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"][\"ABCA13\"], X_pruned.loc[\"TARGET-10-PAMXHJ-09\"][\"ABCA13\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(np.array_equal(XTreehouse_pruned.columns, X_pruned.columns))\n",
    "np.allclose(XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"], X_pruned.loc[\"TARGET-10-PAMXHJ-09\"], .1, .1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.013058144807285288"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "argmax = (XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"] - X_pruned.loc[\"TARGET-10-PAMXHJ-09\"]).values.argmax()\n",
    "gene = XTreehouse_pruned.columns.values[argmax]\n",
    "XTreehouse_pruned.loc[\"TARGET-10-PAMXHJ-09\"][gene] - X_pruned.loc[\"TARGET-10-PAMXHJ-09\"][gene]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "slideshow": {
     "slide_type": "-"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0.03602051, 0.00082846, 0.009581  , 0.00217423, 0.00652128,\n",
       "        0.02527082, 0.0255355 , 0.00190764, 0.0217655 , 0.06135037,\n",
       "        0.00501902, 0.00243543, 0.00993029, 0.00218832, 0.02461774,\n",
       "        0.00682906, 0.00103806, 0.03076606, 0.01329999, 0.03639663,\n",
       "        0.01020063, 0.01095279, 0.02787906, 0.00408726, 0.02223892,\n",
       "        0.00985814, 0.02439659, 0.00953044, 0.01006105, 0.00517068,\n",
       "        0.00609144, 0.00484305, 0.00704081, 0.14349389, 0.00178283,\n",
       "        0.00867913, 0.0084342 , 0.01555362, 0.00505872, 0.00600773,\n",
       "        0.01408515, 0.00815202, 0.02530229, 0.0070094 , 0.00233284,\n",
       "        0.01902671]], dtype=float32)"
      ]
     },
     "execution_count": 169,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict(np.array([X_pruned.loc[\"GTEX-1117F-0226-SM-5GZZ7\"].values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "metadata": {},
   "outputs": [],
   "source": [
    "predictions = {id: model.predict(np.array([XTreehouse_pruned.loc[id].values])) \n",
    "               for id in XTreehouse.index if id.startswith(\"TH\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for id, prediction in predictions.items():\n",
    "    print(id, prediction[0][0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TH01_0051_S01 21\n",
      "TH01_0053_S01 45\n",
      "TH01_0054_S01 5\n",
      "TH01_0055_S01 8\n",
      "TH01_0061_S01 33\n",
      "TH01_0062_S01 5\n",
      "TH01_0063_S01 9\n",
      "TH01_0064_S01 5\n",
      "TH01_0069_S01 9\n",
      "TH01_0119_S01 21\n",
      "TH01_0120_S01 45\n",
      "TH01_0121_S01 33\n",
      "TH01_0122_S01 45\n",
      "TH01_0123_S01 45\n",
      "TH01_0124_S01 45\n",
      "TH01_0125_S01 45\n",
      "TH01_0126_S01 45\n",
      "TH01_0127_S01 45\n",
      "TH01_0128_S01 33\n",
      "TH01_0129_S01 45\n",
      "TH01_0130_S01 45\n",
      "TH01_0131_S01 45\n",
      "TH01_0132_S01 45\n",
      "TH01_0133_S01 33\n",
      "TH01_0134_S01 45\n",
      "TH01_0135_S01 33\n",
      "TH03_0003_S01 42\n",
      "TH03_0003_S03 17\n",
      "TH03_0003_S04 33\n",
      "TH03_0004_S02 21\n",
      "TH03_0004_S03 21\n",
      "TH03_0004_S04 22\n",
      "TH03_0005_S01 33\n",
      "TH03_0006_S01 33\n",
      "TH03_0006_S03 33\n",
      "TH03_0008_S01 19\n",
      "TH03_0010_S01 45\n",
      "TH03_0010_S02 45\n",
      "TH03_0011_S01 45\n",
      "TH03_0012_S01 19\n",
      "TH03_0012_S02 19\n",
      "TH03_0013_S01 45\n",
      "TH03_0013_S02 45\n",
      "TH03_0014_S01 22\n",
      "TH03_0015_S01 19\n",
      "TH03_0016_S01 8\n",
      "TH03_0017_S01 21\n",
      "TH03_0018_S01 19\n",
      "TH03_0018_S02 33\n",
      "TH03_0019_S01 21\n",
      "TH03_0020_S01 8\n",
      "TH03_0022_S01 33\n",
      "TH03_0024_S01 17\n",
      "TH03_0025_S02 22\n",
      "TH03_0027_S01 22\n",
      "TH03_0029_S01 33\n",
      "TH03_0038_S01 33\n",
      "TH03_0103_S01 22\n",
      "TH03_0104_S01 21\n",
      "TH03_0105_S01 33\n",
      "TH03_0106_S01 19\n",
      "TH03_0107_S01 21\n",
      "TH03_0112_S01 9\n",
      "TH03_0112_S02 9\n",
      "TH03_0113_S01 9\n",
      "TH03_0113_S02 9\n",
      "TH03_0114_S01 21\n",
      "TH03_0115_S01 8\n",
      "TH03_0116_S01 21\n",
      "TH03_0117_S01 8\n",
      "TH03_0118_S01 45\n",
      "TH03_0139_S01 33\n",
      "TH03_0140_S01 33\n",
      "TH03_0141_S01 33\n",
      "TH03_0143_S01 33\n",
      "TH03_0143_S02 22\n",
      "TH03_0144_S01 19\n",
      "TH03_0146_S01 22\n",
      "TH03_0148_S01 22\n",
      "TH03_0149_S01 45\n",
      "TH03_0156_S01 9\n",
      "TH03_0157_S01 19\n",
      "TH03_0159_S01 19\n",
      "TH03_0159_S02 19\n",
      "TH03_0160_S01 33\n",
      "TH03_0285_S01 33\n",
      "TH03_0286_S01 21\n",
      "TH03_0287_S01 19\n",
      "TH03_0288_S01 19\n",
      "TH03_0289_S01 45\n",
      "TH03_0290_S01 8\n",
      "TH03_0291_S01 8\n",
      "TH03_0296_S01 33\n",
      "TH03_0296_S02 22\n",
      "TH04_0021_S01 8\n",
      "TH06_0284_S01 19\n",
      "TH06_0610_S01 9\n",
      "TH06_0611_S01 33\n",
      "TH06_0612_S01 9\n",
      "TH06_0613_S01 8\n",
      "TH06_0613_S02 22\n",
      "TH06_0615_S01 33\n",
      "TH06_0616_S01 8\n",
      "TH06_0617_S01 8\n",
      "TH06_0618_S01 9\n",
      "TH06_0619_S01 22\n",
      "TH06_0620_S01 8\n",
      "TH06_0621_S01 22\n",
      "TH06_0622_S01 33\n",
      "TH06_0623_S01 19\n",
      "TH06_0624_S01 22\n",
      "TH06_0625_S01 9\n",
      "TH06_0626_S01 33\n",
      "TH06_0627_S01 8\n",
      "TH06_0628_S01 9\n",
      "TH06_0629_S01 33\n",
      "TH06_0630_S01 22\n",
      "TH06_0631_S01 9\n",
      "TH06_0632_S01 33\n",
      "TH06_0633_S01 33\n",
      "TH06_0634_S01 33\n",
      "TH06_0645_S01 33\n",
      "TH06_0646_S01 9\n",
      "TH06_0647_S01 8\n",
      "TH06_0648_S01 45\n",
      "TH06_0649_S01 19\n",
      "TH26_0657_S01 33\n",
      "TH26_0657_S02 33\n",
      "TH26_0657_S03 33\n",
      "TH26_0657_S04 33\n",
      "TH26_0657_S05 33\n",
      "TH26_0657_S06 33\n",
      "THR08_0161_S01 45\n",
      "THR08_0161_S02 14\n",
      "THR08_0162_S01 45\n",
      "THR08_0162_S02 14\n",
      "THR08_0163_S01 45\n",
      "THR08_0164_S01 14\n",
      "THR08_0165_S01 14\n",
      "THR08_0166_S01 14\n",
      "THR08_0167_S01 14\n",
      "THR08_0168_S01 45\n",
      "THR08_0169_S01 45\n",
      "THR08_0170_S01 45\n",
      "THR08_0171_S01 45\n",
      "THR08_0172_S01 14\n",
      "THR08_0173_S01 14\n",
      "THR08_0174_S01 14\n",
      "THR08_0175_S01 14\n",
      "THR08_0176_S01 14\n",
      "THR08_0177_S01 45\n",
      "THR08_0178_S01 14\n",
      "THR08_0179_S01 45\n",
      "THR08_0180_S01 14\n",
      "THR08_0181_S01 45\n",
      "THR08_0182_S01 45\n",
      "THR08_0183_S01 45\n",
      "THR08_0184_S01 45\n",
      "THR08_0185_S01 45\n",
      "THR08_0186_S01 14\n",
      "THR08_0186_S02 14\n",
      "THR08_0187_S01 45\n",
      "THR08_0187_S02 14\n",
      "THR08_0188_S01 45\n",
      "THR08_0189_S01 45\n",
      "THR08_0190_S01 45\n",
      "THR08_0191_S01 14\n",
      "THR08_0192_S01 45\n",
      "THR08_0193_S01 45\n",
      "THR08_0194_S01 14\n",
      "THR08_0195_S01 45\n",
      "THR08_0196_S01 14\n",
      "THR08_0197_S01 14\n",
      "THR08_0198_S01 14\n",
      "THR08_0198_S02 14\n",
      "THR08_0199_S01 45\n",
      "THR08_0200_S01 14\n",
      "THR08_0201_S01 14\n",
      "THR08_0202_S01 14\n",
      "THR08_0203_S01 14\n",
      "THR08_0204_S01 14\n",
      "THR08_0205_S01 14\n",
      "THR08_0206_S01 14\n",
      "THR08_0207_S01 8\n",
      "THR08_0208_S01 45\n",
      "THR08_0209_S01 45\n",
      "THR08_0210_S01 45\n",
      "THR08_0211_S01 45\n",
      "THR08_0212_S01 45\n",
      "THR08_0213_S01 45\n",
      "THR08_0214_S01 45\n",
      "THR08_0215_S01 45\n",
      "THR08_0216_S01 45\n",
      "THR08_0217_S01 45\n",
      "THR08_0218_S01 45\n",
      "THR09_0219_S01 33\n",
      "THR09_0220_S01 33\n",
      "THR09_0221_S01 33\n",
      "THR09_0222_S01 33\n",
      "THR09_0223_S01 14\n",
      "THR09_0224_S01 14\n",
      "THR09_0561_S01 14\n",
      "THR09_0562_S01 33\n",
      "THR09_0563_S01 33\n",
      "THR09_0564_S01 14\n",
      "THR09_0565_S01 33\n",
      "THR09_0566_S01 14\n",
      "THR09_0567_S01 14\n",
      "THR09_0568_S01 14\n",
      "THR09_0569_S01 9\n",
      "THR09_0570_S01 14\n",
      "THR09_0571_S01 45\n",
      "THR09_0572_S01 33\n",
      "THR09_0573_S01 33\n",
      "THR09_0574_S01 14\n",
      "THR09_0575_S01 33\n",
      "THR09_0576_S01 33\n",
      "THR09_0577_S01 33\n",
      "THR10_0225_S01 8\n",
      "THR10_0226_S01 9\n",
      "THR10_0227_S01 9\n",
      "THR10_0228_S01 45\n",
      "THR10_0229_S01 8\n",
      "THR10_0230_S01 19\n",
      "THR10_0231_S01 22\n",
      "THR10_0232_S01 9\n",
      "THR10_0233_S01 17\n",
      "THR10_0234_S01 19\n",
      "THR10_0235_S01 19\n",
      "THR10_0236_S01 17\n",
      "THR10_0237_S01 19\n",
      "THR10_0238_S01 9\n",
      "THR10_0243_S01 8\n",
      "THR11_0244_S01 8\n",
      "THR11_0245_S01 8\n",
      "THR11_0246_S01 8\n",
      "THR11_0247_S01 8\n",
      "THR11_0248_S01 8\n",
      "THR11_0249_S01 8\n",
      "THR11_0250_S01 8\n",
      "THR11_0251_S01 8\n",
      "THR11_0252_S01 8\n",
      "THR11_0253_S01 8\n",
      "THR11_0254_S01 8\n",
      "THR11_0255_S01 8\n",
      "THR11_0256_S01 8\n",
      "THR11_0257_S01 8\n",
      "THR11_0258_S01 8\n",
      "THR11_0259_S01 8\n",
      "THR11_0260_S01 8\n",
      "THR11_0261_S01 8\n",
      "THR11_0262_S01 8\n",
      "THR11_0263_S01 8\n",
      "THR11_0264_S01 8\n",
      "THR11_0265_S01 8\n",
      "THR11_0266_S01 8\n",
      "THR11_0267_S01 8\n",
      "THR11_0268_S01 8\n",
      "THR11_0269_S01 8\n",
      "THR11_0270_S01 8\n",
      "THR11_0271_S01 8\n",
      "THR11_0273_S01 8\n",
      "THR11_0274_S01 8\n",
      "THR11_0275_S01 8\n",
      "THR11_0276_S01 8\n",
      "THR12_0277_S01 17\n",
      "THR12_0278_S01 19\n",
      "THR12_0278_S03 19\n",
      "THR12_0279_S01 19\n",
      "THR12_0280_S01 19\n",
      "THR12_0280_S03 17\n",
      "THR12_0280_S04 19\n",
      "THR12_0281_S01 17\n",
      "THR12_0282_S01 19\n",
      "THR12_0365_S01 19\n",
      "THR12_0365_S02 19\n",
      "THR12_0365_S03 19\n",
      "THR12_0365_S04 19\n",
      "THR12_0365_S06 19\n",
      "THR12_0366_S02 21\n",
      "THR12_0366_S03 21\n",
      "THR12_0366_S04 21\n",
      "THR12_0368_S01 19\n",
      "THR12_0369_S01 17\n",
      "THR12_0370_S01 17\n",
      "THR13_0375_S01 33\n",
      "THR13_0376_S01 8\n",
      "THR13_0377_S02 8\n",
      "THR13_0378_S02 8\n",
      "THR13_0379_S09 33\n",
      "THR13_0380_S02 22\n",
      "THR13_0382_S02 9\n",
      "THR13_0383_S01 9\n",
      "THR13_0384_S02 9\n",
      "THR13_0385_S01 22\n",
      "THR14_0297_S01 8\n",
      "THR14_0298_S01 19\n",
      "THR14_0299_S01 8\n",
      "THR14_0300_S01 8\n",
      "THR14_0301_S01 8\n",
      "THR14_0302_S01 8\n",
      "THR14_0303_S01 8\n",
      "THR14_0304_S01 8\n",
      "THR14_0305_S01 8\n",
      "THR14_0306_S01 8\n",
      "THR14_0307_S01 8\n",
      "THR14_0308_S01 8\n",
      "THR14_0309_S01 8\n",
      "THR14_0310_S01 8\n",
      "THR14_0311_S01 8\n",
      "THR14_0312_S01 8\n",
      "THR14_0313_S01 8\n",
      "THR14_0314_S01 8\n",
      "THR14_0315_S01 8\n",
      "THR14_0316_S01 8\n",
      "THR14_0317_S01 8\n",
      "THR14_0318_S01 8\n",
      "THR14_0319_S01 8\n",
      "THR14_0320_S01 8\n",
      "THR14_0321_S01 8\n",
      "THR14_0324_S01 8\n",
      "THR14_0325_S01 8\n",
      "THR14_0326_S01 8\n",
      "THR14_0327_S01 8\n",
      "THR15_0328_S01 33\n",
      "THR15_0329_S01 19\n",
      "THR15_0330_S01 33\n",
      "THR15_0334_S01 33\n",
      "THR15_0336_S01 33\n",
      "THR15_0337_S01 22\n",
      "THR15_0338_S01 33\n",
      "THR15_0339_S01 33\n",
      "THR15_0341_S01 33\n",
      "THR15_0342_S01 33\n",
      "THR15_0343_S01 33\n",
      "THR15_0344_S01 33\n",
      "THR15_0345_S01 33\n",
      "THR15_0346_S01 33\n",
      "THR15_0347_S01 33\n",
      "THR15_0348_S01 33\n",
      "THR15_0349_S01 19\n",
      "THR15_0350_S01 33\n",
      "THR15_0352_S01 33\n",
      "THR15_0353_S01 33\n",
      "THR15_0354_S01 17\n",
      "THR15_0355_S01 33\n",
      "THR15_0356_S01 19\n",
      "THR15_0357_S01 33\n",
      "THR15_0358_S01 33\n",
      "THR15_0359_S01 33\n",
      "THR15_0360_S01 9\n",
      "THR15_0361_S01 33\n",
      "THR15_0362_S01 33\n",
      "THR17_0387_S01 8\n",
      "THR17_0388_S01 8\n",
      "THR17_0389_S01 8\n",
      "THR17_0390_S01 8\n",
      "THR17_0391_S01 8\n",
      "THR17_0392_S01 8\n",
      "THR17_0393_S01 8\n",
      "THR17_0394_S01 8\n",
      "THR17_0395_S01 8\n",
      "THR17_0396_S01 8\n",
      "THR17_0397_S01 8\n",
      "THR17_0398_S01 8\n",
      "THR17_0399_S01 8\n",
      "THR17_0400_S01 8\n",
      "THR17_0401_S01 8\n",
      "THR17_0402_S01 8\n",
      "THR17_0403_S01 8\n",
      "THR18_0404_S01 8\n",
      "THR18_0405_S01 22\n",
      "THR18_0406_S01 8\n",
      "THR18_0407_S01 8\n",
      "THR18_0408_S01 8\n",
      "THR18_0409_S01 8\n",
      "THR19_0410_S01 8\n",
      "THR19_0411_S01 8\n",
      "THR19_0412_S01 8\n",
      "THR19_0413_S01 8\n",
      "THR19_0414_S01 8\n",
      "THR19_0415_S01 8\n",
      "THR19_0416_S01 8\n",
      "THR19_0417_S01 8\n",
      "THR19_0418_S01 8\n",
      "THR19_0419_S01 8\n",
      "THR19_0420_S01 8\n",
      "THR19_0421_S01 8\n",
      "THR19_0422_S01 8\n",
      "THR19_0423_S01 8\n",
      "THR19_0424_S01 8\n",
      "THR19_0425_S01 8\n",
      "THR19_0426_S01 8\n",
      "THR19_0427_S01 8\n",
      "THR19_0428_S01 8\n",
      "THR19_0429_S01 8\n",
      "THR19_0430_S01 8\n",
      "THR19_0431_S01 8\n",
      "THR19_0432_S01 8\n",
      "THR19_0433_S01 8\n",
      "THR19_0434_S01 8\n",
      "THR19_0435_S01 8\n",
      "THR19_0436_S01 8\n",
      "THR19_0437_S01 8\n",
      "THR19_0438_S01 8\n",
      "THR19_0439_S01 8\n",
      "THR19_0440_S01 8\n",
      "THR19_0441_S01 8\n",
      "THR19_0442_S01 8\n",
      "THR19_0443_S01 8\n",
      "THR19_0444_S01 8\n",
      "THR19_0445_S01 8\n",
      "THR19_0446_S01 8\n",
      "THR19_0447_S01 8\n",
      "THR19_0448_S01 8\n",
      "THR19_0449_S01 8\n",
      "THR19_0450_S01 8\n",
      "THR19_0451_S01 8\n",
      "THR19_0452_S01 8\n",
      "THR19_0453_S01 8\n",
      "THR19_0454_S01 8\n",
      "THR19_0455_S01 8\n",
      "THR19_0456_S01 8\n",
      "THR19_0457_S01 8\n",
      "THR19_0458_S01 8\n",
      "THR19_0459_S01 8\n",
      "THR19_0460_S01 8\n",
      "THR19_0461_S01 8\n",
      "THR19_0462_S01 8\n",
      "THR19_0463_S01 8\n",
      "THR19_0464_S01 8\n",
      "THR19_0465_S01 8\n",
      "THR19_0468_S01 8\n",
      "THR19_0469_S01 8\n",
      "THR19_0470_S01 8\n",
      "THR19_0471_S01 8\n",
      "THR19_0472_S01 8\n",
      "THR19_0473_S01 8\n",
      "THR19_0474_S01 8\n",
      "THR19_0475_S01 8\n",
      "THR19_0476_S01 8\n",
      "THR19_0477_S01 8\n",
      "THR19_0478_S01 8\n",
      "THR19_0479_S01 8\n",
      "THR19_0480_S01 8\n",
      "THR20_0481_S01 8\n",
      "THR20_0482_S01 26\n",
      "THR20_0483_S01 8\n",
      "THR20_0484_S01 8\n",
      "THR20_0485_S01 8\n",
      "THR20_0486_S01 8\n",
      "THR20_0487_S01 8\n",
      "THR20_0488_S01 8\n",
      "THR20_0489_S01 8\n",
      "THR20_0490_S01 8\n",
      "THR20_0491_S01 8\n",
      "THR20_0492_S01 8\n",
      "THR20_0493_S01 8\n",
      "THR20_0494_S01 8\n",
      "THR20_0495_S01 8\n",
      "THR20_0496_S01 8\n",
      "THR20_0497_S01 8\n",
      "THR20_0498_S01 8\n",
      "THR20_0499_S01 8\n",
      "THR20_0500_S01 9\n",
      "THR20_0501_S01 9\n",
      "THR20_0502_S01 9\n",
      "THR20_0503_S01 9\n",
      "THR20_0504_S01 9\n",
      "THR20_0505_S01 9\n",
      "THR20_0506_S01 9\n",
      "THR20_0507_S01 8\n",
      "THR20_0508_S01 8\n",
      "THR20_0509_S01 8\n",
      "THR20_0510_S01 9\n",
      "THR20_0511_S01 33\n",
      "THR20_0512_S01 8\n",
      "THR20_0513_S01 8\n",
      "THR20_0514_S01 8\n",
      "THR20_0515_S01 8\n",
      "THR20_0516_S01 19\n",
      "THR20_0517_S01 9\n",
      "THR20_0518_S01 19\n",
      "THR20_0519_S01 9\n",
      "THR21_0520_S01 8\n",
      "THR21_0521_S01 8\n",
      "THR21_0522_S01 8\n",
      "THR21_0523_S01 8\n",
      "THR21_0524_S01 8\n",
      "THR21_0525_S01 9\n",
      "THR21_0526_S01 8\n",
      "THR21_0527_S01 8\n",
      "THR21_0528_S01 8\n",
      "THR21_0530_S01 8\n",
      "THR21_0531_S01 8\n",
      "THR21_0532_S01 8\n",
      "THR21_0532_S02 8\n",
      "THR21_0532_S03 8\n",
      "THR21_0533_S01 17\n",
      "THR21_0534_S01 8\n",
      "THR21_0537_S01 8\n",
      "THR21_0538_S01 8\n",
      "THR21_0539_S01 8\n",
      "THR21_0540_S01 8\n",
      "THR21_0541_S01 8\n",
      "THR21_0542_S01 8\n",
      "THR21_0543_S01 8\n",
      "THR21_0544_S01 8\n",
      "THR21_0545_S01 22\n",
      "THR21_0546_S01 8\n",
      "THR21_0547_S01 8\n",
      "THR21_0548_S01 8\n",
      "THR21_0549_S01 8\n",
      "THR21_0550_S01 8\n",
      "THR21_0551_S01 8\n",
      "THR21_0552_S01 8\n",
      "THR21_0553_S01 8\n",
      "THR21_0554_S01 8\n",
      "THR21_0554_S02 8\n",
      "THR21_0555_S01 8\n",
      "THR21_0556_S01 8\n",
      "THR21_0557_S01 8\n",
      "THR21_0558_S01 8\n",
      "THR21_0559_S01 8\n",
      "THR21_0560_S01 33\n",
      "THR22_0578_S01 8\n",
      "THR22_0579_S01 33\n",
      "THR22_0580_S01 33\n",
      "THR22_0581_S01 33\n",
      "THR22_0582_S01 14\n",
      "THR22_0583_S01 33\n",
      "THR22_0584_S01 14\n",
      "THR22_0585_S01 14\n",
      "THR22_0586_S01 33\n",
      "THR22_0587_S01 33\n",
      "THR22_0588_S01 14\n",
      "THR22_0589_S01 33\n",
      "THR22_0590_S01 33\n",
      "THR22_0591_S01 33\n",
      "THR22_0592_S01 33\n",
      "THR22_0593_S01 33\n",
      "THR22_0594_S01 33\n",
      "THR22_0595_S01 33\n",
      "THR22_0599_S01 14\n",
      "THR22_0600_S01 33\n",
      "THR22_0601_S01 33\n",
      "THR22_0602_S01 33\n",
      "THR22_0603_S01 33\n",
      "THR22_0604_S01 33\n"
     ]
    }
   ],
   "source": [
    "for id, prediction in predictions.items():\n",
    "    print(id, np.argmax(prediction[0][0:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [],
   "source": [
    "# sample_id = \"TH01_0717_S01\"\n",
    "\n",
    "# x = pd.read_csv(\"/treehouse/archive/downstream/{}/secondary/ucsc_cgl-rnaseq-cgl-pipeline-3.3.4-785eee9/RSEM/Hugo/rsem_genes.hugo.results\"\n",
    "#             .format(sample_id), \n",
    "#             sep=\"\\t\").set_index(\"gene_name\")\n",
    "# print(x.shape)\n",
    "# x.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'Tensor' object has no attribute 'loc'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-167-176ff3dded17>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mx_pruned\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mX_pruned\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mx_pruned\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msort_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mx_pruned\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mx_pruned\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0mx_pruned\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mduplicated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkeep\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'first'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mx_pruned\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'Tensor' object has no attribute 'loc'"
     ]
    }
   ],
   "source": [
    "x_pruned = x.loc[X_pruned.columns.values]\n",
    "x_pruned.sort_index(axis=0, inplace=True)\n",
    "x_pruned = x_pruned[~x_pruned.index.duplicated(keep='first')]\n",
    "x_pruned.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x_pruned[\"expected_count\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "model.predict(np.array([x_pruned[\"FPKM\"].values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
