{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#  Gene Sets and Expression for Tumor vs. Normal and Tissue Classification\n",
    "\n",
    "Use gene sets from [MSigDB](http://software.broadinstitute.org/gsea/msigdb) to both prune the number of genes/features as well as a source of pathway information to encorporate into layer design."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n",
      "Using TensorFlow backend.\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import tensorflow as tf\n",
    "import keras\n",
    "import matplotlib.pyplot as pyplot\n",
    "\n",
    "# fix random seed for reproducibility\n",
    "np.random.seed(42)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Expression and Convert To Hugo\n",
    "\n",
    "Expression in TPM is only available with Ensembl gene identifiers from Xena. Load this up and convert into Hugo gene names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "expression: samples=60498 genes=19260\n",
      "CPU times: user 2.43 s, sys: 3.48 s, total: 5.91 s\n",
      "Wall time: 5.89 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "expression = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").sort_index(axis=1)\n",
    "print(\"expression: samples={} genes={}\".format(*expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneType</th>\n",
       "      <th>geneStatus</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>transcriptName</th>\n",
       "      <th>transcriptType</th>\n",
       "      <th>transcriptStatus</th>\n",
       "      <th>havanaGeneId</th>\n",
       "      <th>havanaTranscriptId</th>\n",
       "      <th>ccdsId</th>\n",
       "      <th>level</th>\n",
       "      <th>transcriptClass</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000456328.2</td>\n",
       "      <td>DDX11L1-002</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000362751.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.5</th>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000450305.2</td>\n",
       "      <td>DDX11L1-001</td>\n",
       "      <td>transcribed_unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000961.2</td>\n",
       "      <td>OTTHUMT00000002844.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.5</th>\n",
       "      <td>WASH7P</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000488147.1</td>\n",
       "      <td>WASH7P-001</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000958.1</td>\n",
       "      <td>OTTHUMT00000002839.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>pseudo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278267.1</th>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000619216.1</td>\n",
       "      <td>MIR6859-1-201</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.3</th>\n",
       "      <td>RP11-34P13.3</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000473358.1</td>\n",
       "      <td>RP11-34P13.3-001</td>\n",
       "      <td>lincRNA</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000000959.2</td>\n",
       "      <td>OTTHUMT00000002840.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       geneName                            geneType  \\\n",
       "geneId                                                                \n",
       "ENSG00000223972.5       DDX11L1  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000223972.5       DDX11L1  transcribed_unprocessed_pseudogene   \n",
       "ENSG00000227232.5        WASH7P              unprocessed_pseudogene   \n",
       "ENSG00000278267.1     MIR6859-1                               miRNA   \n",
       "ENSG00000243485.3  RP11-34P13.3                             lincRNA   \n",
       "\n",
       "                  geneStatus       transcriptId    transcriptName  \\\n",
       "geneId                                                              \n",
       "ENSG00000223972.5      KNOWN  ENST00000456328.2       DDX11L1-002   \n",
       "ENSG00000223972.5      KNOWN  ENST00000450305.2       DDX11L1-001   \n",
       "ENSG00000227232.5      KNOWN  ENST00000488147.1        WASH7P-001   \n",
       "ENSG00000278267.1      KNOWN  ENST00000619216.1     MIR6859-1-201   \n",
       "ENSG00000243485.3      KNOWN  ENST00000473358.1  RP11-34P13.3-001   \n",
       "\n",
       "                                       transcriptType transcriptStatus  \\\n",
       "geneId                                                                   \n",
       "ENSG00000223972.5                processed_transcript            KNOWN   \n",
       "ENSG00000223972.5  transcribed_unprocessed_pseudogene            KNOWN   \n",
       "ENSG00000227232.5              unprocessed_pseudogene            KNOWN   \n",
       "ENSG00000278267.1                               miRNA            KNOWN   \n",
       "ENSG00000243485.3                             lincRNA            KNOWN   \n",
       "\n",
       "                           havanaGeneId    havanaTranscriptId ccdsId  level  \\\n",
       "geneId                                                                        \n",
       "ENSG00000223972.5  OTTHUMG00000000961.2  OTTHUMT00000362751.1    NaN      2   \n",
       "ENSG00000223972.5  OTTHUMG00000000961.2  OTTHUMT00000002844.2    NaN      2   \n",
       "ENSG00000227232.5  OTTHUMG00000000958.1  OTTHUMT00000002839.1    NaN      2   \n",
       "ENSG00000278267.1                   NaN                   NaN    NaN      3   \n",
       "ENSG00000243485.3  OTTHUMG00000000959.2  OTTHUMT00000002840.1    NaN      2   \n",
       "\n",
       "                  transcriptClass  \n",
       "geneId                             \n",
       "ENSG00000223972.5          pseudo  \n",
       "ENSG00000223972.5          pseudo  \n",
       "ENSG00000227232.5          pseudo  \n",
       "ENSG00000278267.1       nonCoding  \n",
       "ENSG00000243485.3       nonCoding  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Feature labels are ensemble ids, convert to hugo gene names for use in interpreting\n",
    "hidden layers in any trained models as they are better known to most bioinformaticians \n",
    "and clinicians. We're using an assembled table from John Vivian @ UCSC here. Another\n",
    "option would be ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt\n",
    "\"\"\"\n",
    "ensemble_to_hugo = pd.read_table(\n",
    "    \"https://github.com/jvivian/docker_tools/blob/master/gencode_hugo_mapping/attrs.tsv?raw=true\",\n",
    "    index_col=0)\n",
    "ensemble_to_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "hugo = ensemble_to_hugo[~ensemble_to_hugo.index.duplicated(keep='first')].reindex(expression.index.values)[\"geneName\"].fillna(\"\")\n",
    "# Make sure we end up with the order of features being identical as some ensemble id's\n",
    "# map to the same hugo gene id\n",
    "assert(expression.index.equals(hugo.index))\n",
    "# Change index from Ensembl to Hugo\n",
    "expression_hugo = expression\n",
    "expression_hugo.index = hugo.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "expression_hugo_mean = expression_hugo.groupby(expression.index).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoU2-30</th>\n",
       "      <th>snoU2_19</th>\n",
       "      <th>snoU83B</th>\n",
       "      <th>snoZ196</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>uc_338</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.4595</td>\n",
       "      <td>0.9343</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>7.5126</td>\n",
       "      <td>0.8164</td>\n",
       "      <td>-2.1140</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-7.911678</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>1.1512</td>\n",
       "      <td>-1.2828</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>6.0777</td>\n",
       "      <td>-2.3147</td>\n",
       "      <td>0.5568</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.490334</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>5.2411</td>\n",
       "      <td>0.8488</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>10.0319</td>\n",
       "      <td>0.1257</td>\n",
       "      <td>-1.1172</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.653133</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>5.4758</td>\n",
       "      <td>2.6325</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>9.7572</td>\n",
       "      <td>1.7702</td>\n",
       "      <td>-1.8836</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.231703</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.231167</td>\n",
       "      <td>4.5534</td>\n",
       "      <td>1.3051</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>7.7931</td>\n",
       "      <td>-0.0725</td>\n",
       "      <td>-2.2447</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.019897</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  5S_rRNA  5_8S_rRNA       7SK    A1BG  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7 -9.9658  -9.9658    -9.9658 -9.965800  4.4595   \n",
       "GTEX-1117F-0426-SM-5EGHI -9.9658  -9.9658    -9.9658 -9.965800  1.1512   \n",
       "GTEX-1117F-0526-SM-5EGHJ -9.9658  -9.9658    -9.9658 -9.965800  5.2411   \n",
       "GTEX-1117F-0626-SM-5N9CS -9.9658  -9.9658    -9.9658 -9.965800  5.4758   \n",
       "GTEX-1117F-0726-SM-5GIEN -9.9658  -9.9658    -9.9658 -5.231167  4.5534   \n",
       "\n",
       "                          A1BG-AS1    A1CF      A2M  A2M-AS1   A2ML1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7    0.9343 -5.0116   7.5126   0.8164 -2.1140   \n",
       "GTEX-1117F-0426-SM-5EGHI   -1.2828 -6.5064   6.0777  -2.3147  0.5568   \n",
       "GTEX-1117F-0526-SM-5EGHJ    0.8488 -6.5064  10.0319   0.1257 -1.1172   \n",
       "GTEX-1117F-0626-SM-5N9CS    2.6325 -9.9658   9.7572   1.7702 -1.8836   \n",
       "GTEX-1117F-0726-SM-5GIEN    1.3051 -9.9658   7.7931  -0.0725 -2.2447   \n",
       "\n",
       "                             ...      snoU2-30  snoU2_19  snoU83B  snoZ196  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0426-SM-5EGHI     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0526-SM-5EGHJ     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0626-SM-5N9CS     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-1117F-0726-SM-5GIEN     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "\n",
       "                          snoZ278  snoZ40   snoZ6  snosnR66    uc_338  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658 -9.9658 -9.9658   -9.9658 -7.911678   \n",
       "GTEX-1117F-0426-SM-5EGHI  -9.9658 -9.9658 -9.9658   -9.9658 -8.490334   \n",
       "GTEX-1117F-0526-SM-5EGHJ  -9.9658 -9.9658 -9.9658   -9.9658 -8.653133   \n",
       "GTEX-1117F-0626-SM-5N9CS  -9.9658 -9.9658 -9.9658   -9.9658 -9.231703   \n",
       "GTEX-1117F-0726-SM-5GIEN  -9.9658 -9.9658 -9.9658   -9.9658 -9.019897   \n",
       "\n",
       "                          yR211F11.2  \n",
       "GTEX-1117F-0226-SM-5GZZ7     -9.9658  \n",
       "GTEX-1117F-0426-SM-5EGHI     -9.9658  \n",
       "GTEX-1117F-0526-SM-5EGHJ     -9.9658  \n",
       "GTEX-1117F-0626-SM-5N9CS     -9.9658  \n",
       "GTEX-1117F-0726-SM-5GIEN     -9.9658  \n",
       "\n",
       "[5 rows x 58582 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace duplicate Hugo values with their mean and invert so matrix is in machine learning rows = samples\n",
    "X = expression_hugo_mean.T\n",
    "X.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Ingest Expression and Gene Sets\n",
    "\n",
    "Load the expression and clinical as wrangled in ingest.ipynb as well as gene sets downloaded from msigdb. Wrangle so that only genes present in each are present and ensure the order of features is consistent."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Y: samples=19131 labels=6\n",
      "CPU times: user 71.3 ms, sys: 9.3 ms, total: 80.6 ms\n",
      "Wall time: 78.7 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Load samples and labels\n",
    "# X = pd.read_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\")\n",
    "# print(\"X: samples={} genes={}\".format(*X.shape))\n",
    "\n",
    "Y = pd.read_table(\"data/TcgaTargetGTEX_phenotype.txt.gz\", compression=\"gzip\", header=0,\n",
    "                      names=[\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "                      sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\")\n",
    "print(\"Y: samples={} labels={}\".format(*Y.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19126 samples with non-null labels\n"
     ]
    }
   ],
   "source": [
    "# Sort so indexes match\n",
    "X.sort_index(axis=0, inplace=True)\n",
    "X.sort_index(axis=1, inplace=True)\n",
    "Y.sort_index(axis=0, inplace=True)\n",
    "\n",
    "# Remove rows where the class is null or the sample is missing\n",
    "intersection = X.index.intersection(Y[pd.notnull(Y[\"primary_site\"])].index)\n",
    "X = X[X.index.isin(intersection)]\n",
    "Y = Y[Y.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(X.index.equals(Y.index))\n",
    "\n",
    "print(intersection.shape[0], \"samples with non-null labels\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "      <th>tumor_normal_value</th>\n",
       "      <th>primary_site_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>18968</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126</td>\n",
       "      <td>19126.000000</td>\n",
       "      <td>19126.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unique</th>\n",
       "      <td>93</td>\n",
       "      <td>93</td>\n",
       "      <td>46</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>top</th>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Breast Invasive Carcinoma</td>\n",
       "      <td>Brain</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Male</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>freq</th>\n",
       "      <td>1212</td>\n",
       "      <td>1212</td>\n",
       "      <td>1846</td>\n",
       "      <td>9185</td>\n",
       "      <td>10453</td>\n",
       "      <td>10534</td>\n",
       "      <td>10530</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.550559</td>\n",
       "      <td>20.651992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.497450</td>\n",
       "      <td>12.419634</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>19.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>45.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         category                    disease primary_site  \\\n",
       "count                       19126                      19126        19126   \n",
       "unique                         93                         93           46   \n",
       "top     Breast Invasive Carcinoma  Breast Invasive Carcinoma        Brain   \n",
       "freq                         1212                       1212         1846   \n",
       "mean                          NaN                        NaN          NaN   \n",
       "std                           NaN                        NaN          NaN   \n",
       "min                           NaN                        NaN          NaN   \n",
       "50%                           NaN                        NaN          NaN   \n",
       "max                           NaN                        NaN          NaN   \n",
       "\n",
       "          sample_type gender  study tumor_normal  tumor_normal_value  \\\n",
       "count           19126  18968  19126        19126        19126.000000   \n",
       "unique             16      2      3            2                 NaN   \n",
       "top     Primary Tumor   Male   TCGA        Tumor                 NaN   \n",
       "freq             9185  10453  10534        10530                 NaN   \n",
       "mean              NaN    NaN    NaN          NaN            0.550559   \n",
       "std               NaN    NaN    NaN          NaN            0.497450   \n",
       "min               NaN    NaN    NaN          NaN            0.000000   \n",
       "50%               NaN    NaN    NaN          NaN            1.000000   \n",
       "max               NaN    NaN    NaN          NaN            1.000000   \n",
       "\n",
       "        primary_site_value  \n",
       "count         19126.000000  \n",
       "unique                 NaN  \n",
       "top                    NaN  \n",
       "freq                   NaN  \n",
       "mean             20.651992  \n",
       "std              12.419634  \n",
       "min               0.000000  \n",
       "50%              19.000000  \n",
       "max              45.000000  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "Y.loc[:, \"tumor_normal\"] = Y.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)\n",
    "\n",
    "# Convert tumor_normal and primary_site into numerical values for two-hot multi-class training\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "encoder = LabelEncoder()\n",
    "Y[\"tumor_normal_value\"] = pd.Series(encoder.fit_transform(Y[\"tumor_normal\"]), index=Y.index)\n",
    "encoder = LabelEncoder()\n",
    "Y[\"primary_site_value\"] = pd.Series(encoder.fit_transform(Y[\"primary_site\"]), index=Y.index)\n",
    "Y.describe(include=\"all\", percentiles=[])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>5S_rRNA</th>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <th>7SK</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2M-AS1</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>...</th>\n",
       "      <th>snoU2-30</th>\n",
       "      <th>snoU2_19</th>\n",
       "      <th>snoU83B</th>\n",
       "      <th>snoZ196</th>\n",
       "      <th>snoZ278</th>\n",
       "      <th>snoZ40</th>\n",
       "      <th>snoZ6</th>\n",
       "      <th>snosnR66</th>\n",
       "      <th>uc_338</th>\n",
       "      <th>yR211F11.2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.4595</td>\n",
       "      <td>0.9343</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>7.5126</td>\n",
       "      <td>0.8164</td>\n",
       "      <td>-2.1140</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-7.911678</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-145LS-1226-SM-5Q5D9</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.601650</td>\n",
       "      <td>3.3350</td>\n",
       "      <td>1.6281</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>11.3547</td>\n",
       "      <td>1.0847</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.314128</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-X638-0126-SM-47JZ8</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>4.2609</td>\n",
       "      <td>1.6875</td>\n",
       "      <td>-4.0350</td>\n",
       "      <td>6.7694</td>\n",
       "      <td>0.7146</td>\n",
       "      <td>5.0845</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.334375</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-24-1422-01</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.618767</td>\n",
       "      <td>4.1962</td>\n",
       "      <td>2.3788</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>6.4925</td>\n",
       "      <td>0.6425</td>\n",
       "      <td>-0.6643</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-8.940686</td>\n",
       "      <td>-4.0350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BF-AAP0-06</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>6.6121</td>\n",
       "      <td>3.7919</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>7.9917</td>\n",
       "      <td>0.1648</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.397798</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 58582 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  5S_rRNA  5_8S_rRNA       7SK    A1BG  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7 -9.9658  -9.9658    -9.9658 -9.965800  4.4595   \n",
       "GTEX-145LS-1226-SM-5Q5D9 -9.9658  -9.9658    -9.9658 -8.601650  3.3350   \n",
       "GTEX-X638-0126-SM-47JZ8  -9.9658  -9.9658    -9.9658 -9.965800  4.2609   \n",
       "TCGA-24-1422-01          -9.9658  -9.9658    -9.9658 -8.618767  4.1962   \n",
       "TCGA-BF-AAP0-06          -9.9658  -9.9658    -9.9658 -9.965800  6.6121   \n",
       "\n",
       "                          A1BG-AS1    A1CF      A2M  A2M-AS1   A2ML1  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7    0.9343 -5.0116   7.5126   0.8164 -2.1140   \n",
       "GTEX-145LS-1226-SM-5Q5D9    1.6281 -6.5064  11.3547   1.0847 -4.6082   \n",
       "GTEX-X638-0126-SM-47JZ8     1.6875 -4.0350   6.7694   0.7146  5.0845   \n",
       "TCGA-24-1422-01             2.3788 -5.5735   6.4925   0.6425 -0.6643   \n",
       "TCGA-BF-AAP0-06             3.7919 -9.9658   7.9917   0.1648 -4.6082   \n",
       "\n",
       "                             ...      snoU2-30  snoU2_19  snoU83B  snoZ196  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-145LS-1226-SM-5Q5D9     ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "GTEX-X638-0126-SM-47JZ8      ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "TCGA-24-1422-01              ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "TCGA-BF-AAP0-06              ...       -9.9658   -9.9658  -9.9658  -9.9658   \n",
       "\n",
       "                          snoZ278  snoZ40   snoZ6  snosnR66    uc_338  \\\n",
       "GTEX-1117F-0226-SM-5GZZ7  -9.9658 -9.9658 -9.9658   -9.9658 -7.911678   \n",
       "GTEX-145LS-1226-SM-5Q5D9  -9.9658 -9.9658 -9.9658   -9.9658 -8.314128   \n",
       "GTEX-X638-0126-SM-47JZ8   -9.9658 -9.9658 -9.9658   -9.9658 -9.334375   \n",
       "TCGA-24-1422-01           -9.9658 -9.9658 -9.9658   -9.9658 -8.940686   \n",
       "TCGA-BF-AAP0-06           -9.9658 -9.9658 -9.9658   -9.9658 -9.397798   \n",
       "\n",
       "                          yR211F11.2  \n",
       "GTEX-1117F-0226-SM-5GZZ7     -9.9658  \n",
       "GTEX-145LS-1226-SM-5Q5D9     -9.9658  \n",
       "GTEX-X638-0126-SM-47JZ8      -9.9658  \n",
       "TCGA-24-1422-01              -4.0350  \n",
       "TCGA-BF-AAP0-06              -9.9658  \n",
       "\n",
       "[5 rows x 58582 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X[0:20000:3000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>category</th>\n",
       "      <th>disease</th>\n",
       "      <th>primary_site</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>study</th>\n",
       "      <th>tumor_normal</th>\n",
       "      <th>tumor_normal_value</th>\n",
       "      <th>primary_site_value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose - Subcutaneous</td>\n",
       "      <td>Adipose Tissue</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-145LS-1226-SM-5Q5D9</th>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Lung</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTEX-X638-0126-SM-47JZ8</th>\n",
       "      <td>Skin - Sun Exposed (Lower Leg)</td>\n",
       "      <td>Skin - Sun Exposed (Lower Leg)</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Normal Tissue</td>\n",
       "      <td>Female</td>\n",
       "      <td>GTEX</td>\n",
       "      <td>Normal</td>\n",
       "      <td>0</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-24-1422-01</th>\n",
       "      <td>Ovarian Serous Cystadenocarcinoma</td>\n",
       "      <td>Ovarian Serous Cystadenocarcinoma</td>\n",
       "      <td>Ovary</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-BF-AAP0-06</th>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin Cutaneous Melanoma</td>\n",
       "      <td>Skin</td>\n",
       "      <td>Metastatic</td>\n",
       "      <td>Female</td>\n",
       "      <td>TCGA</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   category  \\\n",
       "id                                                            \n",
       "GTEX-1117F-0226-SM-5GZZ7             Adipose - Subcutaneous   \n",
       "GTEX-145LS-1226-SM-5Q5D9                               Lung   \n",
       "GTEX-X638-0126-SM-47JZ8      Skin - Sun Exposed (Lower Leg)   \n",
       "TCGA-24-1422-01           Ovarian Serous Cystadenocarcinoma   \n",
       "TCGA-BF-AAP0-06                     Skin Cutaneous Melanoma   \n",
       "\n",
       "                                                    disease    primary_site  \\\n",
       "id                                                                            \n",
       "GTEX-1117F-0226-SM-5GZZ7             Adipose - Subcutaneous  Adipose Tissue   \n",
       "GTEX-145LS-1226-SM-5Q5D9                               Lung            Lung   \n",
       "GTEX-X638-0126-SM-47JZ8      Skin - Sun Exposed (Lower Leg)            Skin   \n",
       "TCGA-24-1422-01           Ovarian Serous Cystadenocarcinoma           Ovary   \n",
       "TCGA-BF-AAP0-06                     Skin Cutaneous Melanoma            Skin   \n",
       "\n",
       "                            sample_type  gender study tumor_normal  \\\n",
       "id                                                                   \n",
       "GTEX-1117F-0226-SM-5GZZ7  Normal Tissue  Female  GTEX       Normal   \n",
       "GTEX-145LS-1226-SM-5Q5D9  Normal Tissue  Female  GTEX       Normal   \n",
       "GTEX-X638-0126-SM-47JZ8   Normal Tissue  Female  GTEX       Normal   \n",
       "TCGA-24-1422-01           Primary Tumor  Female  TCGA        Tumor   \n",
       "TCGA-BF-AAP0-06              Metastatic  Female  TCGA        Tumor   \n",
       "\n",
       "                          tumor_normal_value  primary_site_value  \n",
       "id                                                                \n",
       "GTEX-1117F-0226-SM-5GZZ7                   0                   0  \n",
       "GTEX-145LS-1226-SM-5Q5D9                   0                  22  \n",
       "GTEX-X638-0126-SM-47JZ8                    0                  33  \n",
       "TCGA-24-1422-01                            1                  26  \n",
       "TCGA-BF-AAP0-06                            1                  33  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Y[0:20000:3000].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded 186 gene sets\n",
      "Subsetting to 5172 genes\n",
      "X_pruned shape (19126, 5172)\n"
     ]
    }
   ],
   "source": [
    "# Load gene sets from downloaded MSigDB gmt file (KEGG to for now as its experimental vs. computational)\n",
    "with open(\"data/c2.cp.kegg.v6.1.symbols.gmt\") as f:\n",
    "    gene_sets = { line.strip().split(\"\\t\")[0]: line.strip().split(\"\\t\")[2:] for line in f.readlines()}\n",
    "print(\"Loaded {} gene sets\".format(len(gene_sets)))\n",
    "\n",
    "# Drop any genes that are not in X - sort so that the order is the same as X_pruned.columns\n",
    "gene_sets = {name: sorted([gene for gene in genes if gene in X.columns.values]) \n",
    "             for name, genes in gene_sets.items()}\n",
    "\n",
    "# Find the union of all gene's in the gene sets in order to filter our input rows\n",
    "all_gene_set_genes = sorted(list(set().union(*[gene_set for gene_set in gene_sets.values()])))\n",
    "print(\"Subsetting to {} genes\".format(len(all_gene_set_genes)))\n",
    "\n",
    "# Prune X to only include genes in the gene sets\n",
    "X_pruned = X.drop(labels=(set(X.columns) - set(all_gene_set_genes)), axis=1, errors=\"ignore\")\n",
    "assert X_pruned[\"TP53\"][\"TCGA-ZP-A9D4-01\"] == X[\"TP53\"][\"TCGA-ZP-A9D4-01\"]\n",
    "print(\"X_pruned shape\", X_pruned.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make sure the genes are the same and in the same order\n",
    "assert len(all_gene_set_genes) == len(X_pruned.columns.values)\n",
    "assert list(X_pruned.columns.values) == all_gene_set_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Wrangle Labels, Train and Test Sets\n",
    "\n",
    "Add one hot labels for tumor/normal and primary site and then build training and test sets stratifying on primary site."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(19126, 47)\n",
      "[[0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]\n",
      " [1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.\n",
      "  0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]]\n"
     ]
    }
   ],
   "source": [
    "# Create a multi-label output with tumor/normal + primary site one-hots\n",
    "from keras.utils import np_utils\n",
    "primary_site_one_hot = np_utils.to_categorical(Y[\"primary_site_value\"])\n",
    "tumor_normal_one_hot = Y[\"tumor_normal_value\"]\n",
    "Y_multiclass_onehot = np.append(\n",
    "    tumor_normal_one_hot.values.reshape(tumor_normal_one_hot.shape[0],-1), \n",
    "    primary_site_one_hot, axis=1)\n",
    "print(Y_multiclass_onehot.shape)\n",
    "print(Y_multiclass_onehot[0:20000:5000])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(15300, 5172) (3826, 5172)\n"
     ]
    }
   ],
   "source": [
    "# Split into stratified training and test sets based on classes (i.e. tissue type) so that we have equal\n",
    "# proportions of each tissue type in the train and test sets\n",
    "from sklearn.model_selection import StratifiedShuffleSplit\n",
    "split = StratifiedShuffleSplit(n_splits=1, test_size=0.2, random_state=42)\n",
    "for train_index, test_index in split.split(X_pruned.values, Y[\"tumor_normal_value\"]):\n",
    "    X_train, X_test = X_pruned.values[train_index], X_pruned.values[test_index]\n",
    "    y_train, y_test = Y_multiclass_onehot[train_index], Y_multiclass_onehot[test_index]\n",
    "    classes_train, classes_test = Y[\"primary_site_value\"].values[train_index], Y[\"primary_site_value\"].values[test_index]\n",
    "\n",
    "print(X_train.shape, X_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEICAYAAABWJCMKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xu8VVW99/HPV8QgxVBEQ0BBQ4+XjoTbW9rR8s7jSeto6skb6UF79JWWVmh1vGXpOaVpmj6UJFpKHM3kePQxNC9ZeUHEC5KJSrJlCxsUwfKG/s4fY2yZLNbee+077Pl9v17rtdYcc8w5xxxrzvmbY8y55lJEYGZm5bNOTxfAzMx6hgOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSfWKACDpPEm/6OZl3iTpsHZO+ylJz3Z2mbqapC0kvSGpTzunv1TSKW2cZp6k/fLncyT9rD3Lbmbeb0jaKn++TtJ3O3He10j6TmfNrzDfbt/W1zaSvijpt2tAOUZICknr9nRZmrPWBABJ/yppRt5pGyTdKWmvHirLPwI7Abfl4RMkvZfLtkzSLEmHNDd9RPw+IrbtrvK2haRhkm6RtFjS65KeknQCQES8FBEbRMR7Oe99kk5qw+z/E/iWpPXaU7aI+F5EtLq8WsuV1+WF9pSlYnknSHqwYt6nRMSFHZ13Z1obgkdnBOKI+GVEHNBZZeoqxZObDs5nte2vVmtFAJD0NeBHwPeAzYAtgJ8Ah/ZQkU4Gfhmr/oruTxGxATAQuBaYKmnjygm7+2ygHWfrNwDzgS2BQcBxwMLOKEtENAB/Bj7bGfNrrzX5jMxa5u+uk0XEGv0CPgK8ARzRQp7zgF8Uhv8LeAV4HXgA2KEwbizwDLAceBk4K6dvAtwOLAVeBX4PrNPM8l4A9ioMnwA8WBheHwigDtgHqAe+mct0Q1NaIf884OvAk8DfSAFkM+DOXM67gY1qXL/rgKuBO/K8vk46gK9byPMvwKxm1u0NYHQz40bk9VoXuAh4D3grT3NlzvMPwPRch88CX6iYx7eAn7fwXR4L/BVYkvPOA/ar/J6BfsAvcr6lwKO5zporVwCnAs8BLxbSPlaot2ty2ZcD9wNbVq53oZz3AScB2+VlvZeXt7Qwv+8W8v8bMDfXyzRg88K4AE7JZXsNuApQC9v6zcCvcjlnAjsVxm8O3AI0Ai8CX8npBwHvAO/mcj4BfBp4qjDt3cAjheEHgcNamm8etw4wAXg+fx9TgY0r6u544CVgMfCtZtZtfC7fO7mM/13YP75J2j/eJm1/TctbTtqfP9fC/tiW+t0V+BNpm2oArgTWq2VeQB/gB3kdXyBtb6tsN4X53AC8D7yZ1/UbOX134I95+U8A+1Ss1wt5nV8Evkgz21/Nx9e2HpC7+5U33BXVKrFipygGgC8BA4APkVoOswrjGoBP5c8bAWPy5++TDgB98+tT1TYSVh7cB1fb4PLGeXr+kj5COtivAC7J5elP9QDwEOkANhRYRNqxP5Gn+R1wbo3rdx0pMOxJ2jH7kXaQgwt5bgXObKYu7wb+ABwFbFExbkRxgyYfBCvqZj4wLtfDmLwzFAPU54GZzSx7+7wR/1Net0tz3VULACcD/w18mLTj7QxsWK1chR13OrAx0L+QVgwAywvLvrzwna6y3pXLoOKAU5jfd/Pnz+R6GJPn/WPggYqy3U5qPW5BOsge1MK2/i5wOGk7PYt0MOibv+/HgH8H1gO2Ih0wDmxmP+lHOgBtkr+vV4AFpG2rfx43qIb5nkHafofl9ft/wE0VdffTPM+dSAfx7ZpZvw/qrWL/mAUML3x3R5CC0jrAkaSTnSHVvo821u/OpIPwurnsc4AzapkXKTD8OZdzY+BemgkAhfXarzA8lBRAx+b12j8PDybtW8uAbXPeIeT9qnJ92/JaG7qABgGLI2JFrRNExKSIWB4Rb5M2+p0kfSSPfhfYXtKGEfFaRMwspA8hnfW9G6mfvtqDkgbm9+UV6btLWkraiY4mnZG8nse9TzqAvx0RbzZT7B9HxMKIeJnU+ng4Ih7P63ArKRjUsn4At0XEHyLi/Yh4C5gMHAOQu6UOBG5sphxH5OV/B3gxX8/YpZm8lQ4B5kXEzyNiRa7bW0gHqybLWVmHlQ4Hbo+IB/K6fYdUd9W8S9o2PhYR70XEYxGxrJXyfT8iXm3hO/ifwrK/BewhaXgr86zFF4FJETEzz/vsPO8RhTwXR8TSiHiJdOAY3cL8HouImyPiXVKQ7Ec6aO1COjG5ICLeiXR946ekYL6avG3MIAW9OtIZ9oOkk4fdgeciYkkN8z2ZdFZfX9gmD6/orjk/It6MiCdIZ7Y7tVZpFa6IiPlN311E/FdELMjb+K9IZ+S7tjB9TfWbt6OH8vY7jxTM9q5xXl8AfpTL+SrppLItjgHuiIg78npNJ30/Y/P494EdJfWPiIaImN3G+a9mbQgAS4BNau37k9RH0sWSnpe0jBRlIZ3lQOr+GAv8VdL9kvbI6f9JaqL/VtILkiY0s4il+X1ARfpDETEwIjaJiN0j4u7CuMa8s7Wk2M/+ZpXhDWpcP0hn4UW/AP5Z0gakjfT3kfrjV5OD4oSI2IHUIpkF/EaSWik/pOsGu0la2vQiHfw+WsgzgJV1WGnzYtkj4m+k77+aG4C7gCmSFkj6D0l9WylfZb00Oz4i3iB112zeyjS12JzUrVWc9xLSGV+TVwqf/07+vmso5/ukLsbNSfW/eUX9n0P6HptzP6lF+k/5832kA97eeZga5rslcGth3BxSl0RxuW1ZvxbXGUDScfnkpGmZO7LqPlCppuVL2kbS7ZJeyfvX96rMt7l5rbL9UvjOa7QlcERFPe9Fatn8jdTSOQVokPQ/kv6hjfNfzdoQAP5E6uOq9ZbLfyVdHN6P1AUzIqcLICIejYhDgU2B35D6K8ln1GdGxFbAPwNfk7Rv5czzF/E8sE0b1qFaS6K9Wly/asvLrYo/AZ8j9bHfUMuCImIxqU9zc1KTdrUsFcPzgftzIGx6bRARXy7k2Y50BlhNA6n5DICkD5PO8quV7d2IOD8itgc+SWp9HNdMuZorb6XisjcgrfMCUvcCpO6mJsWg1tp8F5B27qZ5r09ar5dbma6Wcq5D6npZQKr/Fyvqf0BENJ1BVitnZQC4n9UDQGvznU/qYiyO75e3u7Zq9buTtCWpBXIaMCgiBgJPs+o+0F5Xk7pxRkXEhqRAV+t8V9l+SV1ELam2/9xQUY/rR8TFABFxV0TsT+qp+DOpDqrNp2ZrfADI3Sj/Dlwl6TBJH5bUV9LBkv6jyiQDSH2MS0g77PeaRkhaL98j/JHcfF5GOlNB0iGSPpbPdJvS32umWHewerOwuzS7fq24HvgG8HFSl1JVki6RtKOkdSUNAL4MzM1dAZUWkvqDm9wObCPp2Pwd9ZW0i6TtCnn2Jl3cruZm4BBJe+VbRS+gmW1U0qclfTzf5bSM1CXU9H1VlqtWYwvLvpDUDTc/IhpJB+tjcgvsS8DWhekWAsNauL31RmCcpNGSPkT6zh7OXQztsbOkz+dW8Rmk7eEh4BFgmaRvSuqfy7pjoQtvITAiB40mfwS2JXWfPJK7FbYEdiPdYEAN870GuCgfmJE0WFJ779Cr5btrug7XmJc3jtQC6AwDSNvTG/kM+8ut5C+aCnxF6VbqjUgXqltSua5NLfUDcx33k7RPnt9mkj6bTx7eJl0rK27vLW1/zVrjAwBARFwKfA34NulLn0+K/r+pkv16UtPrZdLFz4cqxh8LzMvNu1PIfePAKNIF0DdIZ8s/iYj7minSROCLNXaLdLbW1q85t5Kb6rkV05wP57xLSRf6tqT52zYvJ/X1vibpiohYDhxA6hteQGoqN138RtIQ0oXeat8b+eBzKumA2UC6y6K+mWV/lBQwlpG6HO4n7UCrlauFda10I3AuqetnZ1L3VZN/I91RtQTYgXTgbPI7YDbwiqTFVdbrHtL1jFvyem1NM/3yNbqN1B3wGml7/nxuEb1Har2OJl0YXgz8jNRShHT3GMASSTNz2f5GuuFgdkS8k8f/CfhrRCzKeVqb7+WkO5t+K2k5aZvcrZ3rdi3pGt1SSc1tJ88AP8zlXEg6qflDO5dX6SxSK3s56Qz7V22Y9qekbsknSHX661byfx/4dl7XsyJiPql1fw4rj3NfJx2n1wHOJO1Xr5JOpP5vnk+L219Lmm5fsjaSdCMwNSKqbqRrIknPAydXXJ/ozuX/EHg+In7SE8s3s1U5AJSEpH8hnY1vky8cmlnJ+Vd1JSDpPlLXy7E++JtZE7cAzMxKaq24CGxmZp2v1S4gSf1It4N9KOe/OSLOlTQSmEK6V3omqXvhnXyb2/WkuyiWAEc23e4m6WzgRNLtS1+JiLtaWvYmm2wSI0aMaOeqmZmV02OPPbY4Iga3lq+WawBvA5+JiDfyLy0flHQn6bbMyyJiiqRrSAf2q/P7axHxMUlHkS48Hilpe9KtbzuQflh0t6Rt8i1mVY0YMYIZM2bUUEQzM2siqaZfIbfaBRTJG3mw6UFpQXrA1c05fTIrf6l7aB4mj9833y9/KDAl0vNwXiQ9dqGlZ3eYmVkXqukaQP5V2izSUyqnkx6FsDRWPqCtnpXPNRlKfh5GHv866WfvH6RXmcbMzLpZTQEg0tMWR5OeObIr6Xkuq2XL79V+HRstpK9C0nilf/6a0djYWEvxzMysHdr0O4CIWJrvKd8dGChp3XyW3/QwKkhn9sOB+vysko+QfrrclN6kOE1xGRNJj1qgrq7O96iaWc3effdd6uvreeut1h6+2zv069ePYcOG0bdvaw/Cra6Wu4AGA+/mg39/0lMoLyE9B/tw0p1Ax5P/H5f0TJDjSc/pOBz4XUSEpGnAjZIuJV0EHkV6yJSZWaeor69nwIABjBgxgp55VFf3iQiWLFlCfX09I0eObNc8amkBDAEm56curkN6/s3tkp4hPYv9u8DjpIc4kd9vkNT093dH5cLOljSV9ACzFcCpLd0BZGbWVm+99VYpDv4Akhg0aBAd6SpvNQBExJMU/o2qkP4CVe7iyX98ckQz87qI9J+tZmZdogwH/yYdXVf/EtjMrKT8MDgz67Uum/6XTp3fV/dv+Y8AlyxZwr77pj8SfOWVV+jTpw+DB6cf5D7yyCOst17r/9kybtw4JkyYwLbbbtvxArfCAcDWap29g7dFawcDK59BgwYxa9YsAM477zw22GADzjrrrFXyRAQRwTrrVO+A+fnPf97l5WziLiAzsy42d+5cdtxxR0455RTGjBlDQ0MD48ePp66ujh122IELLrjgg7x77bUXs2bNYsWKFQwcOJAJEyaw0047sccee7Bo0aJOLZcDgJlZN3jmmWc48cQTefzxxxk6dCgXX3wxM2bM4IknnmD69Ok888wzq03z+uuvs/fee/PEE0+wxx57MGnSpE4tkwOAmVk32Hrrrdlll10+GL7pppsYM2YMY8aMYc6cOVUDQP/+/Tn44IMB2HnnnZk3b16nlsnXAMzMusH666//wefnnnuOyy+/nEceeYSBAwdyzDHHVP31cvGicZ8+fVixYsVqeTrCLQAzs262bNkyBgwYwIYbbkhDQwN33dXiX6N0GbcAzKzXWlPv1BozZgzbb789O+64I1tttRV77rlnj5Rjjf5P4Lq6uvAfwlhLfBuoFc2ZM4fttqv2sOLeq9o6S3osIupam9ZdQGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlL+HYCZ9V73fr9z5/fps1sc3RmPgwaYNGkSY8eO5aMf/WjHytsKBwAzs05Sy+OgazFp0iTGjBnjAGBm1htMnjyZq666infeeYdPfvKTXHnllbz//vuMGzeOWbNmERGMHz+ezTbbjFmzZnHkkUfSv3//NrUc2soBwMysiz399NPceuut/PGPf2Tddddl/PjxTJkyha233prFixfz1FNPAbB06VIGDhzIj3/8Y6688kpGjx7dpeVyADAz62J33303jz76KHV16ekMb775JsOHD+fAAw/k2Wef5fTTT2fs2LEccMAB3VouBwAzsy4WEXzpS1/iwgsvXG3ck08+yZ133skVV1zBLbfcwsSJE7utXL4N1Mysi+23335MnTqVxYsXA+luoZdeeonGxkYigiOOOILzzz+fmTNnAjBgwACWL1/e5eVyC8DMeq9WbtvsLh//+Mc599xz2W+//Xj//ffp27cv11xzDX369OHEE08kIpDEJZdcAsC4ceM46aSTuvwisB8HbWs1Pw7aivw46MSPgzYzsxY5AJiZlVSrAUDScEn3Spojabak03P6eZJeljQrv8YWpjlb0lxJz0o6sJB+UE6bK2lC16ySmZXZmtyt3dk6uq61XAReAZwZETMlDQAekzQ9j7ssIn5QzCxpe+AoYAdgc+BuSU2dpVcB+wP1wKOSpkXEMx1aAzOzrF+/fixZsoRBgwYhqaeL06UigiVLltCvX792z6PVABARDUBD/rxc0hxgaAuTHApMiYi3gRclzQV2zePmRsQLAJKm5LwOAGbWKYYNG0Z9fT2NjY09XZRu0a9fP4YNG9bu6dt0G6ikEcAngIeBPYHTJB0HzCC1El4jBYeHCpPVszJgzK9I363KMsYD4wG22GKLthTPzEqub9++jBw5sqeLsdao+SKwpA2AW4AzImIZcDWwNTCa1EL4YVPWKpNHC+mrJkRMjIi6iKhreoyqmZl1vppaAJL6kg7+v4yIXwNExMLC+J8Ct+fBemB4YfJhwIL8ubl0MzPrZrXcBSTgWmBORFxaSB9SyPY54On8eRpwlKQPSRoJjAIeAR4FRkkaKWk90oXiaZ2zGmZm1la1tAD2BI4FnpI0K6edAxwtaTSpG2cecDJARMyWNJV0cXcFcGpEvAcg6TTgLqAPMCkiZnfiupiZWRvUchfQg1Tvv7+jhWkuAi6qkn5HS9OZmVn38S+BzcxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspPyXkF3A/1JlZmsDtwDMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKRaDQCShku6V9IcSbMlnZ7TN5Y0XdJz+X2jnC5JV0iaK+lJSWMK8zo+539O0vFdt1pmZtaadWvIswI4MyJmShoAPCZpOnACcE9EXCxpAjAB+CZwMDAqv3YDrgZ2k7QxcC5QB0Sez7SIeK2zV6rMLpv+lx5Z7lf336ZHlmtm7ddqCyAiGiJiZv68HJgDDAUOBSbnbJOBw/LnQ4HrI3kIGChpCHAgMD0iXs0H/enAQZ26NmZmVrM2XQOQNAL4BPAwsFlENEAKEsCmOdtQYH5hsvqc1lx65TLGS5ohaUZjY2NbimdmZm1QcwCQtAFwC3BGRCxrKWuVtGghfdWEiIkRURcRdYMHD661eGZm1kY1BQBJfUkH/19GxK9z8sLctUN+X5TT64HhhcmHAQtaSDczsx5Qy11AAq4F5kTEpYVR04CmO3mOB24rpB+X7wbaHXg9dxHdBRwgaaN8x9ABOc3MzHpALXcB7QkcCzwlaVZOOwe4GJgq6UTgJeCIPO4OYCwwF/g7MA4gIl6VdCHwaM53QUS82ilrYWZmbdZqAIiIB6nefw+wb5X8AZzazLwmAZPaUkAzM+sa/iWwmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUk5AJiZlZQDgJlZSTkAmJmVlAOAmVlJOQCYmZWUA4CZWUm1GgAkTZK0SNLThbTzJL0saVZ+jS2MO1vSXEnPSjqwkH5QTpsraULnr4qZmbVFLS2A64CDqqRfFhGj8+sOAEnbA0cBO+RpfiKpj6Q+wFXAwcD2wNE5r5mZ9ZB1W8sQEQ9IGlHj/A4FpkTE28CLkuYCu+ZxcyPiBQBJU3LeZ9pcYjMz6xQduQZwmqQncxfRRjltKDC/kKc+pzWXvhpJ4yXNkDSjsbGxA8UzM7OWtDcAXA1sDYwGGoAf5nRVyRstpK+eGDExIuoiom7w4MHtLJ6ZmbWm1S6gaiJiYdNnST8Fbs+D9cDwQtZhwIL8ubl0MzPrAe1qAUgaUhj8HNB0h9A04ChJH5I0EhgFPAI8CoySNFLSeqQLxdPaX2wzM+uoVlsAkm4C9gE2kVQPnAvsI2k0qRtnHnAyQETMljSVdHF3BXBqRLyX53MacBfQB5gUEbM7fW3MzKxmtdwFdHSV5GtbyH8RcFGV9DuAO9pUOjMz6zL+JbCZWUk5AJiZlZQDgJlZSbXrNlAzK6fLpv+lR5b71f236ZHl9nYOAGZmLejNQc9dQGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5V8CW6foqV9Lmln7uQVgZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZS/h2A2VrGv7mwzuIWgJlZSTkAmJmVlAOAmVlJtXoNQNIk4BBgUUTsmNM2Bn4FjADmAV+IiNckCbgcGAv8HTghImbmaY4Hvp1n+92ImNy5q2JmvZWve3SNWloA1wEHVaRNAO6JiFHAPXkY4GBgVH6NB66GDwLGucBuwK7AuZI26mjhzcys/VoNABHxAPBqRfKhQNMZ/GTgsEL69ZE8BAyUNAQ4EJgeEa9GxGvAdFYPKmZm1o3aew1gs4hoAMjvm+b0ocD8Qr76nNZc+mokjZc0Q9KMxsbGdhbPzMxa09kXgVUlLVpIXz0xYmJE1EVE3eDBgzu1cGZmtlJ7fwi2UNKQiGjIXTyLcno9MLyQbxiwIKfvU5F+XzuXbbZG8IVJW9u1NwBMA44HLs7vtxXST5M0hXTB9/UcJO4Cvle48HsAcHb7i10b76BmZs2r5TbQm0hn75tIqifdzXMxMFXSicBLwBE5+x2kW0Dnkm4DHQcQEa9KuhB4NOe7ICIqLyybmVk3ajUARMTRzYzat0reAE5tZj6TgEltKp2ZmXUZ/xLYzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSqq9fwlptkbY/aWJPbbsh7YY32PLNusMbgGYmZWUA4CZWUk5AJiZlZQDgJlZSfkicBfwhUkzWxu4BWBmVlIOAGZmJeUAYGZWUg4AZmYl5QBgZlZSHQoAkuZJekrSLEkzctrGkqZLei6/b5TTJekKSXMlPSlpTGesgJmZtU9n3Ab66YhYXBieANwTERdLmpCHvwkcDIzKr92Aq/O7daKeugXVt5+arX26ogvoUGBy/jwZOKyQfn0kDwEDJQ3pguWbmVkNOhoAAvitpMckNZ0CbhYRDQD5fdOcPhSYX5i2PqetQtJ4STMkzWhsbOxg8czMrDkd7QLaMyIWSNoUmC7pzy3kVZW0WC0hYiIwEaCurm618WZm1jk61AKIiAX5fRFwK7ArsLCpaye/L8rZ64HhhcmHAQs6snwzM2u/dgcASetLGtD0GTgAeBqYBhyfsx0P3JY/TwOOy3cD7Q683tRVZGZm3a8jXUCbAbdKaprPjRHx/yU9CkyVdCLwEnBEzn8HMBaYC/wdGNeBZZuZWQe1OwBExAvATlXSlwD7VkkP4NT2Ls/MzDqXfwlsZlZSDgBmZiXlAGBmVlIOAGZmJeUAYGZWUg4AZmYl5T+FN7Oa+WmzvYtbAGZmJeUWgJlZC3qq1QM/6PIlOACYrWV67oBkvY0DgHUKH5SsK3n76hq+BmBmVlJuAZi1k89KbW3XqwOAd1Azs+a5C8jMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKQcAMzMSsoBwMyspBwAzMxKygHAzKykHADMzErKAcDMrKS6PQBIOkjSs5LmSprQ3cs3M7OkWwOApD7AVcDBwPbA0ZK2784ymJlZ0t0tgF2BuRHxQkS8A0wBDu3mMpiZGd3/hzBDgfmF4Xpgt2IGSeOB8XnwDUnPdmB5mwCLOzB9b+K6WJXrY1Wuj5XWjLo46YcdmXrLWjJ1dwBQlbRYZSBiItApf+UlaUZE1HXGvNZ2rotVuT5W5fpYqUx10d1dQPXA8MLwMGBBN5fBzMzo/gDwKDBK0khJ6wFHAdO6uQxmZkY3dwFFxApJpwF3AX2ASRExuwsX6X+FX8l1sSrXx6pcHyuVpi4UEa3nMjOzXse/BDYzKykHADOzkuqVAaDsj5uQNEnSIklPF9I2ljRd0nP5faOeLGN3kTRc0r2S5kiaLen0nF7W+ugn6RFJT+T6OD+nj5T0cK6PX+WbNEpBUh9Jj0u6PQ+Xpi56XQDw4yYAuA44qCJtAnBPRIwC7snDZbACODMitgN2B07N20NZ6+Nt4DMRsRMwGjhI0u7AJcBluT5eA07swTJ2t9OBOYXh0tRFrwsA+HETRMQDwKsVyYcCk/PnycBh3VqoHhIRDRExM39eTtrRh1Le+oiIeCMP9s2vAD4D3JzTS1MfkoYB/wf4WR4WJaqL3hgAqj1uYmgPlWVNsllENEA6KAKb9nB5up2kEcAngIcpcX3kLo9ZwCJgOvA8sDQiVuQsZdpnfgR8A3g/Dw+iRHXRGwNAq4+bsPKRtAFwC3BGRCzr6fL0pIh4LyJGk36JvyuwXbVs3Vuq7ifpEGBRRDxWTK6StdfWRXc/C6g7+HET1S2UNCQiGiQNIZ39lYKkvqSD/y8j4tc5ubT10SQilkq6j3RtZKCkdfOZb1n2mT2Bz0oaC/QDNiS1CEpTF72xBeDHTVQ3DTg+fz4euK0Hy9Jtcp/utcCciLi0MKqs9TFY0sD8uT+wH+m6yL3A4TlbKeojIs6OiGERMYJ0nPhdRHyREtVFr/wlcI7oP2Ll4yYu6uEidStJNwH7kB5ruxBlGl/tAAAAgUlEQVQ4F/gNMBXYAngJOCIiKi8U9zqS9gJ+DzzFyn7ec0jXAcpYH/9IurDZh3QCODUiLpC0FemGiY2Bx4FjIuLtnitp95K0D3BWRBxSprrolQHAzMxa1xu7gMzMrAYOAGZmJeUAYGZWUg4AZmYl5QBgZlZSDgBmZiXlAGBmVlL/C1Dl1pwt/VgfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fca10bda6d8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pyplot.hist(classes_train, alpha=0.5, label='Train')\n",
    "pyplot.hist(classes_test, alpha=0.5, label='Test')\n",
    "pyplot.legend(loc='upper right')\n",
    "pyplot.title(\"Class (Primary Site) distribution between train and test\")\n",
    "pyplot.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYAAAAEICAYAAABWJCMKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAIABJREFUeJzt3X2UHFWd//H3hySQEAJ5GhAyWRMPEUFXEOYHyXIEJBAC7hL8CRqOSmDDZo9GXfAxqLtRngSf4nJWYbMQDahgRJH8WBVjICJqAoNAIMSQgJiMCckweQCBYEK+vz/qDnSGmenqmZ4eJvV5ndOnq27dunVvdXV9u25VVykiMDOz4tmrtytgZma9wwHAzKygHADMzArKAcDMrKAcAMzMCsoBwMysoBwArCKSvifpiznzHiopSsZ/KekDVarHSZJWlIw3STqpGmWn8lZJeme1yisp915J51e73D2JpOslfe51UI8LJS3p7Xr0pD0+AEj6a8lrl6QXS8arsjOqFkn9JYWkhySpJP0qSdf3Zt2qISImRcT3O8tTsg7GlClrSUS8tRr1ai+oRcRhEfGbapRfLX0heFQjEEfEhRFxZZWq1CPa/rjpZlm5f1RV2x4fACJiv9YXsBb4p5K0TndGPUlS/04mjwbO6eFl9Fl7aruKwJ/d68seHwDKaRt9JZ0i6amS8SZJn5L0aDpqmCvpIEl3Sno2dWsMLcl/lqQVkrZKukvSYW3K+rSkR4AXOqnWV4BLJfXroM4VLaOSNkjaS9Ktkp5O5S+RdHjOddlP0hxJLZKeACa3mf7KL1hJb5Z0j6Rtkp6R9IOU7Z70viLV9b2tn4mkz0l6Gviftp9TcpyklZK2SLpB0j5pWbsdypceZUj6CPB+4HNpebeVrLOT0vBASddI2iDpL5K+IWnvNK21bp+R1CxpvaTzyqyqcZIaU9tvkzSspG7HS1qa1v1Dkk5I6VcDE4DrUj2/KekKSXPS9H0kbZd0ZRrfL40f0Fm5adpQSd9J7WuSdKmkvUrW3a/T57pV0pOSJnXw+d8MHAL8PNXxE0q/lCVdIGkt8Mty25hKvpOVrt9U35WSnpP0hKQLS6Z1WpakOkl3pO/EUmBsJ5/hPWme1t6E/1Oy/D+mbfDnkkan9L3SNrQpfe7LJR3R0fZXMxFRmBfwFHBKm7TvAV8sGT8FeKpkvAn4HXAgUA+0AI3AkcBA4NfA51Pew4G/AicDA4DPAY8DA0rKeiCVM6id+vUHgmzDewg4P6VfBVzf1WVU2Ia9gPOBIWnafwGNHa2vNvX/KLAiLWME2ZckSqbfW9KmHwGfTcsbCBzfZh2MafOZ7ASuBPYGBnXwOS1Pyx4JLG2tJ3AhsKSd9Tymozal8k5Kw1em9VeX1uEyYHabus1On8eZwPPA/h2so3uBdcARwGDgp8B307TR6bM5La2XycAzwIi26y+NTwIeTMMnAE8Avy2Z9kDOcu8Avg3sC7yBbPuZXrLudgD/DPQDPgas6+Q79sp6S+OHpnX9nVT+ICrYxrqwfv8JeBMgsu/Ii8Db85QF3ArcnOr5dmBD6XbTZjmHUrJtp7SzgVXAYWTb2BeB36Rp7wbuAw5I7T8CeEO571RPvwp/BJDTf0bEpohoIvsS/j4iHo6I7WRf4HekfFOBhRFxV0TsINtx7w8c16aspoh4sZPlBfAfwGxJA9pM6+oycrUhInZFxHcj4rk07YvAMZIG51hP7wPmpGW3pLp1ZAcwBjg4IrZHxG/LlL2T7Evyt07W3TVp2c+Q7bTPzVHnPD6Qlt0cEZuAS4EPlUzfDlweETsiYiHwEvDmTsqbHxGPRcTzZJ/zVEkCziP7bO9Mn8MvgIdpcyRV4rfAEcqO3k4A5gJjJe0LnEgW2OmsXEmjgInAxRHxQkQ8DXyTbDtr9UREzIuIl4H5QL2kkeVX225mp/Jf7MI2lnv9RsT/i4gnI3MXsBh4Z7my0vfsLODfUz2XAzdV2MZ/Ba6MiFURsRO4HDg2reMdZN/Tt6R6PpbWda9yAMhnY8nwi+2M75eGDwH+3DohInaR/SIaVZJ/XZ4Fpo1zI9kvsFJdXUauNijrxvlKOtR/FliT8uT5wh/SZtl/7igj8EmyX2GNkh6RNK1M2Rsj4m9l8rRd9iFl8ud1MLu35c/svr6fSTvHVi/w6jbRnrb13AcYDrwRODd1i2yVtBUYTwftSAHkQbKd/wnAErIjnwnsHgA6K/eNafkbS6Z9CzioZFGlO6rWrsvO2tdpm7uwjeVev5L+UdIySZtTWya1Kbejsg4iO8LJu/22543At0rW4zPALqA+In4JXAdcS7aur5M0pMLyq84BIDsE3Ldk/A3dKGs92UYAZP1+ZF0SfynJU8mVA19Ir0E9uIy2zgPOIDt8PoDsUBeyQ+pyNpB1N7T6u44yRsSGyK72OBiYCcyVNJaO656nTW2XvT4Nl/uMy5W9gZJ1nsr+Swd582hbz5eAzWQ7n+9ExNCS1+CI+Gon9fw1WdfG35N13fwaOB04Bmi9iqmzcteR7QSHl0zbPyLe3sW2tbsuI/V1JN3ZxjokaRBZN86XgYMiYijwy5zlbiTbWefafmm/nevIus5K1/OgiFgGEBHfjIijgbeRdQF9opOyasIBIOtrf7ekYZIOBj7ejbIWAGcqu0Z9APBp4DmyPuOKRcSvyPr3P9hTy2jHELIdUgvZTvOKCuZdAFwkaZSkEWR9/O2S9L50aAywlexL8HL6ddZC1o9bqY+WLPsS4Icp/WHg7ZL+Pu0kZreZb2OZ5d0M/IekkZLqgH8n67ftqvMkvSV1eXwJWJB2kDcB75F0avqVPFDSuyS1HgG0V89fk/WnL0/dDkuAGcDjEbE55emw3IhYl8r4mqT908nKQ1VykrhC5dYldG8b68w+ZOeImoGXJf0jWfdWWak79afAlyQNkvQ2du/ma2sTEJJK23od8PnWE9rKTq6fnYaPTa/+ZD9I/ga0HonkWWc9wgEAvgusJDvc+wVwS1cLiogVwDSyw7xmsr7bM9PG1VWfJ+se6MlllPoO2S/n9WQndH9XwbzXkvW5PgLcT/ZrrCPHAfdLeh74CTAzItamabOBH6RD6f9bwfJvBn5FdjJ0Fdl5ACLisTS8JKXf02a+64Ej05Ub7dX5S2RB5BGyE83LyH5ldtVNZAFkA1m3w0Wpnk8B7yELMM1kly1/kle/p9/k1a6cb6S0e8l2oq1tWk7W3/xKG3OU+0GyE9KPAVvITtB39Uj4SrKd6FZJF3WQpzvbWIciYitwMXAb2RHV2WQnuPP6MDCMbId8Q6pnR8t6jmwbWJba2hARPwK+AfwodW0tJzvxDjA0lbmV7GKUDcCcNK3c9tdjtPuRmZmZFYWPAMzMCsoBwMysoBwAzMwKygHAzKygXtc3Zho5cmSMGTOmt6thZtanPPDAA89ERF25fK/rADBmzBgaGxt7uxpmZn2KpFz/YnYXkJlZQTkAmJkVlAOAmVlBva7PAZiZVWLHjh00NTWxffv23q5KTQwcOJD6+noGDGh71/h8HADMbI/R1NTEkCFDGDNmDFK3bi76uhcRtLS00NTUxNixnT28rGPuAjKzPcb27dsZMWLEHr/zB5DEiBEjunW04wBgZnuUIuz8W3W3rQ4AZmYF5XMAZrbHmrPo8aqWd/GpnT3qGVpaWpg4MXsGzdNPP02/fv2oq8v+kHvfffex9957l13GBRdcwKxZszjssMO6X+EyHADMzDrw/Es7dxvf+GyZ/vYBg7nznqUAfPXLlzN48GA+8vGLAdiyfRds305EEBHstVf7HTBX/ee13a94Tu4CMjPrYX964glOHH8Mn7noY5z6zglsfHoDn/r4TCadeDwnHHc0X7/6ylfynnnayTy6/GF27tzJ0KFDmTVrFkceeSQTJkxg06ZNVa2XA4CZWQ08/seVnPuhafzq3qUcfMgoPv/Fy/jlr3/LXb+9j3vuXsyqP658zTzbtm3jxBNP5OGHH2bChAnMmzevqnVyADAzq4ExY9/EO45peGX8th8v4NR3TuDUEyawetUqHm8nAAwaNIjTTz8dgGOOOYannnqqqnXyOQAzsxrYd/DgV4affGIN11/7LX5+1284YOhQZv7LBbz00kuvmaf0pHG/fv3YuXPna/J0h48AzMxq7Llnn2W//YYwZP/92fj0Bu5e/KteqYePAMxsj1Xuss1yyl7100VvP+odvPktb+HE8cfwxjFjOXb8+B5ZTjmKiPKZpIuBC4EAHgEuAA4GbgGGA38APhQRf5O0D3AjcAzQArw/Ip5K5VwCTAdeBj4eEXd2ttyGhobwA2HMLK+VK1dy+OGHV628ngoAeRy0/8Bc+dprs6QHIqKhg1leUbYLSNIo4ONAQ0S8DegHTAWuBuZExDhgC9mOnfS+JSIOBeakfEg6Is33VmAy8G1J/XK10MzMqi7vOYD+wCBJ/YF9gQ3AycCtafp84Kw0PCWNk6ZPVHbDiinALRHxUkT8CVgDHNv9JpiZWVeUDQAR8Rfga8Bash3/NuABYGtEtJ6SbgJGpeFRwLo0786Uf0RpejvzvELSDEmNkhqbm5u70iYzM8shTxfQMLJf72OBQ4DBwOntZG09mdDe7emik/TdEyLmRkRDRDS03kPDzMyqL08X0CnAnyKiOSJ2AD8B/gEYmrqEAOqB9Wm4CRgNkKYfAGwuTW9nHjMzq7E8AWAtMF7SvqkvfyLwGHA3cHbKMw24PQ0vTOOk6XdFdqnRQmCqpH0kjQXGAfdVpxlmZlapsv8DiIhlkm4lu9RzJ/AgMBf4X+AWSZentBvSLDcAN0laQ/bLf2oqZ4WkBWTBYycwMyJernJ7zMxedfeXuzX74DZ3A33+Hz7daf7Nm1s458wzANi0cSP9+u3FiJFZV/bP7/pNrttBA/zgpvmc+94pvOENb+hCrfPL9UewiJgNzG6T/CTtXMUTEduBczoo5wrgigrraGbWJwwfPoLF9y4DXns76Erc8r35vOv4414fAcDMzLrnhz/4Ht/5n+vYsWMHDceO58tfm8OuXbu46CMzePSR5UQEHzr/nxlZdyCPPrKc97///QwaNCj3g2S6wgHAzKyHrXxsBT+/43buWLSE/v3786mPz+SnP17AmLFvoqWlhSW/z+54sG3rVg4YOpR5c6/lv6/9NkcddVSP1ssBwMysh/1myV089IcHOO2k4wHY/uKLHFJfz0kTT+WJ1Y/zhc9+komnTuakiafUtF4OAGZmPSwiOPeD0/jsF9qeSoW7f3c/ixfdyfX//W3+d+FP+do136pZvXw7aDOzHnbCSSez8LYf09LyDJBdLdS0bi3PPNNMRHDme97Lpy/5AssffgiAwfsN4bnnnuvxevkIwMz2XO+6pFuzP1+lu4Ee/ta38clZn+N9U97Nrl27GNB/AFfPuYZ+/frxiY9+mCAQ4gtfuhyAqR/4EBdeeGGPnwTOdTvo3uLbQZtZJXw76EzVbgdtZmZ7JgcAM7OCcgAwsz3K67lbu9q621YHADPbYwwcOJCWlpZCBIGIoKWlhYED850raI+vAjKzPUZ9fT1NTU1U62FSz764oyrldMXmQQPK5hk4cCD19fVdXoYDgJntMQYMGMDYsWOrVt6cRY9XraxKXXzqm3t8Ge4CMjMrKAcAM7OCcgAwMyuoPA+FP0zSQyWvZyVdJGm4pEWSVqf3YSm/JF0jaY2k5ZKOLilrWsq/WtK0jpdqZmY9rWwAiIhVEXFURBwFHAO8ANwGzAIWR8Q4YHEaBzid7Hm/44AZwLUAkoaTPVXsOLInic1uDRpmZlZ7lXYBTQSeiIg/A1OA+Sl9PnBWGp4C3BiZpcBQSQcDpwGLImJzRGwBFgGTu90CMzPrkkoDwFTg5jR8UERsAEjvB6b0UcC6knmaUlpH6buRNENSo6TGal3La2Zmr5U7AEjaGzgT+FG5rO2kRSfpuydEzI2IhohoqKury1s9MzOrUCVHAKcDf4iIjWl8Y+raIb1vSulNwOiS+eqB9Z2km5lZL6jkn8Dn8mr3D8BCYBpwVXq/vST9o5JuITvhuy0iNki6E7iy5MTvJKB7T2soo7f+xVeLf/CZmXVXrgAgaV/gVOBfS5KvAhZImg6sBc5J6T8DzgDWkF0xdAFARGyWdBlwf8p3aURs7nYLzMysS3IFgIh4ARjRJq2F7KqgtnkDmNlBOfOAeZVX08zMqs3/BDYzKygHADOzgnIAMDMrKAcAM7OCcgAwMysoBwAzs4JyADAzKygHADOzgnIAMDMrKAcAM7OCcgAwMysoBwAzs4JyADAzKygHADOzgnIAMDMrqFwBQNJQSbdK+qOklZImSBouaZGk1el9WMorSddIWiNpuaSjS8qZlvKvljStpxplZmbl5T0C+E/gFxHxFuBIYCUwC1gcEeOAxWkcsmcHj0uvGcC1AJKGA7PJHhN5LDC75PGQZmZWY2UDgKT9gROAGwAi4m8RsRWYAsxP2eYDZ6XhKcCNkVkKDE0PjT8NWBQRmyNiC7AImFzV1piZWW55jgDeBDQD35H0oKTrJQ0GDoqIDQDp/cCUfxSwrmT+ppTWUfpuJM2Q1Cipsbm5ueIGmZlZPnkCQH/gaODaiHgH8Dyvdve0R+2kRSfpuydEzI2IhohoqKury1E9MzPrijwBoAloiohlafxWsoCwMXXtkN43leQfXTJ/PbC+k3QzM+sFZQNARDwNrJN0WEqaCDwGLARar+SZBtyehhcC56WrgcYD21IX0Z3AJEnD0snfSSnNzMx6Qf+c+T4GfF/S3sCTwAVkwWOBpOnAWuCclPdnwBnAGuCFlJeI2CzpMuD+lO/SiNhclVaYmVnFcgWAiHgIaGhn0sR28gYws4Ny5gHzKqmgmZn1DP8T2MysoBwAzMwKygHAzKygHADMzArKAcDMrKAcAMzMCsoBwMysoBwAzMwKygHAzKygHADMzArKAcDMrKAcAMzMCsoBwMysoBwAzMwKygHAzKygHADMzAoqVwCQ9JSkRyQ9JKkxpQ2XtEjS6vQ+LKVL0jWS1khaLunoknKmpfyrJU3raHlmZtbzKjkCeFdEHBURrU8GmwUsjohxwOI0DnA6MC69ZgDXQhYwgNnAccCxwOzWoGFmZrXXnS6gKcD8NDwfOKsk/cbILAWGSjoYOA1YFBGbI2ILsAiY3I3lm5lZN+QNAAH8UtIDkmaktIMiYgNAej8wpY8C1pXM25TSOkrfjaQZkholNTY3N+dviZmZVSTXQ+GB4yNivaQDgUWS/thJXrWTFp2k754QMReYC9DQ0PCa6WZmVh25jgAiYn163wTcRtaHvzF17ZDeN6XsTcDoktnrgfWdpJuZWS8oGwAkDZY0pHUYmAQ8CiwEWq/kmQbcnoYXAuelq4HGA9tSF9GdwCRJw9LJ30kpzczMekGeLqCDgNskteb/QUT8QtL9wAJJ04G1wDkp/8+AM4A1wAvABQARsVnSZcD9Kd+lEbG5ai0xM7OKlA0AEfEkcGQ76S3AxHbSA5jZQVnzgHmVV9PMzKrN/wQ2MysoBwAzs4JyADAzKygHADOzgnIAMDMrKAcAM7OCcgAwMysoBwAzs4JyADAzKygHADOzgnIAMDMrKAcAM7OCcgAwMysoBwAzs4JyADAzK6jcAUBSP0kPSrojjY+VtEzSakk/lLR3St8nja9J08eUlHFJSl8l6bRqN8bMzPKr5Ajg34CVJeNXA3MiYhywBZie0qcDWyLiUGBOyoekI4CpwFuBycC3JfXrXvXNzKyrcgUASfXAu4Hr07iAk4FbU5b5wFlpeEoaJ02fmPJPAW6JiJci4k9kj4w8thqNMDOzyuU9Avgm8BlgVxofAWyNiJ1pvAkYlYZHAesA0vRtKf8r6e3M8wpJMyQ1Smpsbm6uoClmZlaJsgFA0j8CmyLigdLkdrJGmWmdzfNqQsTciGiIiIa6urpy1TMzsy4q+1B44HjgTElnAAOB/cmOCIZK6p9+5dcD61P+JmA00CSpP3AAsLkkvVXpPGZmVmNljwAi4pKIqI+IMWQnce+KiA8AdwNnp2zTgNvT8MI0Tpp+V0RESp+arhIaC4wD7qtaS8zMrCJ5jgA68lngFkmXAw8CN6T0G4CbJK0h++U/FSAiVkhaADwG7ARmRsTL3Vi+mZl1Q0UBICKWAEvS8JO0cxVPRGwHzulg/iuAKyqtpJmZVZ//CWxmVlAOAGZmBeUAYGZWUA4AZmYF5QBgZlZQDgBmZgXlAGBmVlAOAGZmBeUAYGZWUA4AZmYF5QBgZlZQDgBmZgXlAGBmVlAOAGZmBeUAYGZWUA4AZmYFleeh8AMl3SfpYUkrJH0ppY+VtEzSakk/lLR3St8nja9J08eUlHVJSl8l6bSeapSZmZWX5wjgJeDkiDgSOAqYLGk8cDUwJyLGAVuA6Sn/dGBLRBwKzEn5kHQE2eMh3wpMBr4tqV81G2NmZvnleSh8RMRf0+iA9ArgZODWlD4fOCsNT0njpOkTJSml3xIRL0XEn4A1tPNISTMzq41c5wAk9ZP0ELAJWAQ8AWyNiJ0pSxMwKg2PAtYBpOnbgBGl6e3MU7qsGZIaJTU2NzdX3iIzM8slVwCIiJcj4iignuxX++HtZUvv6mBaR+ltlzU3IhoioqGuri5P9czMrAsqugooIrYCS4DxwFBJ/dOkemB9Gm4CRgOk6QcAm0vT25nHzMxqLM9VQHWShqbhQcApwErgbuDslG0acHsaXpjGSdPviohI6VPTVUJjgXHAfdVqiJmZVaZ/+SwcDMxPV+zsBSyIiDskPQbcIuly4EHghpT/BuAmSWvIfvlPBYiIFZIWAI8BO4GZEfFydZtjZmZ5lQ0AEbEceEc76U/SzlU8EbEdOKeDsq4Arqi8mmZmVm3+J7CZWUE5AJiZFZQDgJlZQTkAmJkVlAOAmVlBOQCYmRWUA4CZWUE5AJiZFZQDgJlZQTkAmJkVlAOAmVlBOQCYmRWUA4CZWUE5AJiZFZQDgJlZQeV5IthoSXdLWilphaR/S+nDJS2StDq9D0vpknSNpDWSlks6uqSsaSn/aknTOlqmmZn1vDxPBNsJfDIi/iBpCPCApEXA+cDiiLhK0ixgFvBZ4HSyxz2OA44DrgWOkzQcmA00kD0M/gFJCyNiS7UbZWZWDePXzu3FpX+tx5dQ9gggIjZExB/S8HNkzwMeBUwB5qds84Gz0vAU4MbILCV7ePzBwGnAoojYnHb6i4DJVW2NmZnlVtE5AEljyB4PuQw4KCI2QBYkgANTtlHAupLZmlJaR+lmZtYLcgcASfsBPwYuiohnO8vaTlp0kt52OTMkNUpqbG5uzls9MzOrUK4AIGkA2c7/+xHxk5S8MXXtkN43pfQmYHTJ7PXA+k7SdxMRcyOiISIa6urqKmmLmZlVIM9VQAJuAFZGxDdKJi0EWq/kmQbcXpJ+XroaaDywLXUR3QlMkjQsXTE0KaWZmVkvyHMV0PHAh4BHJD2U0j4HXAUskDQdWAuck6b9DDgDWAO8AFwAEBGbJV0G3J/yXRoRm6vSCjMzq1jZABAR99J+/z3AxHbyBzCzg7LmAfMqqaCZmfUM/xPYzKygHADMzAoqzzmAPqv3/sXX8//gMzPrLh8BmJkVlAOAmVlBOQCYmRWUA4CZWUE5AJiZFZQDgJlZQTkAmJkVlAOAmVlBOQCYmRWUA4CZWUE5AJiZFZQDgJlZQTkAmJkVVJ5HQs6TtEnSoyVpwyUtkrQ6vQ9L6ZJ0jaQ1kpZLOrpknmkp/2pJ09pblpmZ1U6eI4DvApPbpM0CFkfEOGBxGgc4HRiXXjOAayELGMBs4DjgWGB2a9AwM7PeUTYARMQ9QNtn904B5qfh+cBZJek3RmYpMFTSwcBpwKKI2BwRW4BFvDaomJlZDXX1HMBBEbEBIL0fmNJHAetK8jWltI7SX0PSDEmNkhqbm5u7WD0zMyun2ieB23t4fHSS/trEiLkR0RARDXV1dVWtnJmZvaqrAWBj6tohvW9K6U3A6JJ89cD6TtLNzKyXdDUALARar+SZBtxekn5euhpoPLAtdRHdCUySNCyd/J2U0szMrJeUfSi8pJuBk4CRkprIrua5ClggaTqwFjgnZf8ZcAawBngBuAAgIjZLugy4P+W7NCLanlg2M7MaKhsAIuLcDiZNbCdvADM7KGceMK+i2pmZWY/xP4HNzArKAcDMrKAcAMzMCsoBwMysoBwAzMwKygHAzKygHADMzArKAcDMrKAcAMzMCsoBwMysoBwAzMwKygHAzKygHADMzArKAcDMrKAcAMzMCsoBwMysoGoeACRNlrRK0hpJs2q9fDMzy9Q0AEjqB3wLOB04AjhX0hG1rIOZmWVqfQRwLLAmIp6MiL8BtwBTalwHMzMjxzOBq2wUsK5kvAk4rjSDpBnAjDT6V0mrurG8kcAz3Zi/ay78es0XmfROe3uX21wMxWvzhV/vTpvfmCdTrQOA2kmL3UYi5gJzq7IwqTEiGqpRVl9QtPaC21wUbnPPqHUXUBMwumS8Hlhf4zqYmRm1DwD3A+MkjZW0NzAVWFjjOpiZGTXuAoqInZI+CtwJ9APmRcSKHlxkVbqS+pCitRfc5qJwm3uAIqJ8LjMz2+P4n8BmZgXlAGBmVlB9PgCUu7WEpH0k/TBNXyZpTO1rWV052vwJSY9JWi5psaRc1wS/nuW9hYiksyWFpD5/yWCeNkt6X/qsV0j6Qa3rWG05tu2/k3S3pAfT9n1Gb9SzWiTNk7RJ0qMdTJeka9L6WC7p6KpWICL67IvsRPITwJuAvYGHgSPa5PkIcF0angr8sLfrXYM2vwvYNw1/uAhtTvmGAPcAS4GG3q53DT7nccCDwLA0fmBv17sGbZ4LfDgNHwE81dv17mabTwCOBh7tYPoZwM/J/kM1HlhWzeX39SOAPLeWmALMT8O3AhMltfeHtL6ibJsj4u6IeCGNLiX7v0VflvcWIpcBXwG217JyPSRPm/8F+FZEbAGIiE01rmO15WlzAPun4QPo4/8jioh7gM2dZJkC3BiZpcBQSQdXa/l9PQC0d2uJUR3liYidwDZgRE1q1zPytLnUdLJfEH1Z2TZLegcwOiLuqGXFelCez/nNwJsl/VbSUkmTa1a7npGnzV8EPiipCfgZ8LHaVK2yXOLJAAABzElEQVTXVPp9r0itbwVRbWVvLZEzT1+Suz2SPgg0ACf2aI16XqdtlrQXMAc4v1YVqoE8n3N/sm6gk8iO8n4j6W0RsbWH69ZT8rT5XOC7EfF1SROAm1Kbd/V89XpFj+6/+voRQJ5bS7ySR1J/ssPGzg65Xu9y3U5D0inA54EzI+KlGtWtp5Rr8xDgbcASSU+R9ZUu7OMngvNu27dHxI6I+BOwiiwg9FV52jwdWAAQEb8HBpLdKG5P1aO3z+nrASDPrSUWAtPS8NnAXZHOrvRRZducukP+m2zn39f7haFMmyNiW0SMjIgxETGG7LzHmRHR2DvVrYo82/ZPyU74I2kkWZfQkzWtZXXlafNaYCKApMPJAkBzTWtZWwuB89LVQOOBbRGxoVqF9+kuoOjg1hKSLgUaI2IhcAPZYeIasl/+U3uvxt2Xs81fBfYDfpTOd6+NiDN7rdLdlLPNe5Scbb4TmCTpMeBl4NMR0dJ7te6enG3+JPA/ki4m6wo5vy//oJN0M1kX3sh0XmM2MAAgIq4jO89xBrAGeAG4oKrL78PrzszMuqGvdwGZmVkXOQCYmRWUA4CZWUE5AJiZFZQDgJlZQTkAmJkVlAOAmVlB/X/6TMuoaZKdbAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7fc9fce04160>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pyplot.hist(y_train[:, 0], alpha=0.5, label='Train')\n",
    "pyplot.hist(y_test[:, 0], alpha=0.5, label='Test')\n",
    "pyplot.legend(loc='upper right')\n",
    "pyplot.title(\"Tumor Normal distribution between train and test\")\n",
    "pyplot.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Tumor Normal Binary Classification\n",
    "First train a basic binary tumor vs. normal classifier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "input_1 (InputLayer)         (None, 5172)              0         \n",
      "_________________________________________________________________\n",
      "batch_normalization_1 (Batch (None, 5172)              20688     \n",
      "_________________________________________________________________\n",
      "dense_1 (Dense)              (None, 32)                165536    \n",
      "_________________________________________________________________\n",
      "dropout_1 (Dropout)          (None, 32)                0         \n",
      "_________________________________________________________________\n",
      "dense_2 (Dense)              (None, 8)                 264       \n",
      "_________________________________________________________________\n",
      "dropout_2 (Dropout)          (None, 8)                 0         \n",
      "_________________________________________________________________\n",
      "dense_3 (Dense)              (None, 1)                 9         \n",
      "=================================================================\n",
      "Total params: 186,497\n",
      "Trainable params: 176,153\n",
      "Non-trainable params: 10,344\n",
      "_________________________________________________________________\n",
      "None\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 3s 198us/step - loss: 0.3129 - acc: 0.8811\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 3s 182us/step - loss: 0.1964 - acc: 0.9371\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 3s 196us/step - loss: 0.1870 - acc: 0.9425\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 3s 188us/step - loss: 0.1905 - acc: 0.9477\n",
      "Epoch 00004: early stopping\n",
      "3826/3826 [==============================] - 0s 117us/step\n",
      "['loss', 'acc'] [0.035609934753575254, 0.9924202822791427]\n"
     ]
    }
   ],
   "source": [
    "from keras.models import Model, Sequential\n",
    "from keras.layers import InputLayer, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "classify = [\n",
    "    InputLayer(input_shape=(X_train.shape[1],)),\n",
    "    BatchNormalization(),\n",
    "\n",
    "    Dense(32, activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(8, activity_regularizer=regularizers.l1(1e-5), activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(1, activation='sigmoid')\n",
    "]\n",
    "\n",
    "model = Sequential(classify)\n",
    "print(model.summary())\n",
    "\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\")]\n",
    "model.fit(X_train, y_train[:,0], epochs=epochs, batch_size=batch_size, shuffle=\"batch\", callbacks=callbacks)\n",
    "\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test[:,0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Multi-label Tumor vs. Normal and Primary Site Classification\n",
    "Multi-label classification of tumor/normal + primary site with union of genes from all gene sets as input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_________________________________________________________________\n",
      "Layer (type)                 Output Shape              Param #   \n",
      "=================================================================\n",
      "input_2 (InputLayer)         (None, 5172)              0         \n",
      "_________________________________________________________________\n",
      "batch_normalization_2 (Batch (None, 5172)              20688     \n",
      "_________________________________________________________________\n",
      "dense_4 (Dense)              (None, 32)                165536    \n",
      "_________________________________________________________________\n",
      "dropout_3 (Dropout)          (None, 32)                0         \n",
      "_________________________________________________________________\n",
      "dense_5 (Dense)              (None, 16)                528       \n",
      "_________________________________________________________________\n",
      "dropout_4 (Dropout)          (None, 16)                0         \n",
      "_________________________________________________________________\n",
      "dense_6 (Dense)              (None, 47)                799       \n",
      "=================================================================\n",
      "Total params: 187,551\n",
      "Trainable params: 177,207\n",
      "Non-trainable params: 10,344\n",
      "_________________________________________________________________\n",
      "None\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 3s 229us/step - loss: 0.5485 - acc: 0.7529\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 3s 194us/step - loss: 0.2411 - acc: 0.9433\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 3s 192us/step - loss: 0.1915 - acc: 0.9657\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 3s 194us/step - loss: 0.1720 - acc: 0.9697\n",
      "Epoch 00004: early stopping\n",
      "3826/3826 [==============================] - 0s 123us/step\n",
      "['loss', 'acc'] [0.0895781630676452, 0.9775778575212745]\n",
      "CPU times: user 2min 38s, sys: 7min 29s, total: 10min 7s\n",
      "Wall time: 13.6 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from keras.models import Model, Sequential\n",
    "from keras.layers import InputLayer, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "classify = [\n",
    "    InputLayer(input_shape=(X_train.shape[1],)),\n",
    "    BatchNormalization(),\n",
    "\n",
    "    Dense(32, activation='relu'),\n",
    "    Dropout(0.5),\n",
    "\n",
    "    Dense(16, activity_regularizer=regularizers.l1(1e-5), activation='relu'),\n",
    "    Dropout(0.5),\n",
    "    \n",
    "    Dense(y_train.shape[1], activation='sigmoid')\n",
    "]\n",
    "\n",
    "model = Sequential(classify)\n",
    "print(model.summary())\n",
    "\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\")]\n",
    "model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, shuffle=\"batch\", callbacks=callbacks)\n",
    "\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pathway KEGG_ABC_TRANSPORTERS Gene Indexes: [  11   12   13   14   15   16   17   18   19   20   21   22   23   24\n",
      "   25   26   27   28   29   30   31   32   33   34   35   36   37   38\n",
      "   39   40   41   42   43   44   45   46   47   48   49   50   51  848\n",
      " 4692 4693]\n"
     ]
    }
   ],
   "source": [
    "# Convert list of gene names into list of indexes into the features of X_pruned\n",
    "gene_set_indexes = {name: np.searchsorted(X_pruned.columns.values, genes) for name, genes in gene_sets.items()}\n",
    "print(\"Pathway KEGG_ABC_TRANSPORTERS Gene Indexes:\", gene_set_indexes[\"KEGG_ABC_TRANSPORTERS\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mulit-label Classification w/Per Pathway Sub-Network Input Layer\n",
    "For each pathway build a custom input layer that extracts the expression levels for the genes in the pathway from the full input vector and feeds this into a dense single output hidden neuron. These are then aggregated and fed into a standard set of stacked layers and trained to classify tumor/norml and primary site. The hidden per pathway neurons are named after the pathway an 'may' indicate which pathway lead to a given classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Trainable params:: 16352\n",
      "CPU times: user 1min 34s, sys: 2.02 s, total: 1min 36s\n",
      "Wall time: 1min 36s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from keras.models import Model, Sequential\n",
    "from keras.layers import Input, Lambda, Dense, BatchNormalization, Activation, Dropout\n",
    "from keras.callbacks import EarlyStopping\n",
    "from keras import regularizers\n",
    "from keras.layers.merge import concatenate\n",
    "from keras import backend as K\n",
    "import itertools\n",
    "\n",
    "main_input = Input(shape=(X_train.shape[1],), name=\"main_input\")\n",
    "\n",
    "\"\"\"\n",
    "Build per pathway sub-networks\n",
    "\"\"\"\n",
    "# Set to a small number (~4) for debugging, set to None to build all pathways\n",
    "# max_num_pathways = 4\n",
    "max_num_pathways = None \n",
    "\n",
    "# Extract the features/gene's for each pathway from the aggregate x input vector\n",
    "pathways = [Lambda(lambda x: K.concatenate([x[:, i:i+1] for i in indexes]))(main_input)\n",
    "            for name, indexes in itertools.islice(gene_set_indexes.items(), max_num_pathways)]\n",
    "\n",
    "# Add a dense hidden layer per pathway with width proportional to the number of genes in the pathway\n",
    "# pathways = [Dense(max(2, len(i)//4), activation='relu')(p) for p, i in zip(pathways, gene_set_indexes.values())]\n",
    "\n",
    "# Add a named binary output for each pathway \n",
    "# pathways = [Dense(1, activity_regularizer=regularizers.l1(1e-5), activation='relu', name=name)(p) \n",
    "#             for p, name in zip(pathways, gene_set_indexes.keys())]\n",
    "\n",
    "pathways = [Dropout(0.5)(Dense(1, activity_regularizer=regularizers.l1(1e-5), activation='relu', name=name)(p))\n",
    "            for p, name in zip(pathways, gene_set_indexes.keys())]\n",
    "\n",
    "# Concatenate the binary outputs of each of the sub-networks back into a single vector\n",
    "x = keras.layers.concatenate(pathways)\n",
    "           \n",
    "\"\"\"\n",
    "Add traditional stacked network for final multi-label classification\n",
    "\"\"\"\n",
    "x = Dense(16, activation='relu')(x)\n",
    "x = Dropout(0.5)(x)\n",
    "    \n",
    "x = Dense(8, activation='relu')(x)\n",
    "x = Dropout(0.5)(x)\n",
    "\n",
    "x = Dense(y_train.shape[1], activation='sigmoid')(x)\n",
    "\n",
    "model = Model(inputs=[main_input], outputs=[x])\n",
    "# print(model.summary())  # Too detailed when building full set of pathways\n",
    "print(\"Trainable params::\", np.sum([np.product(v.shape) for v in model.trainable_weights]).value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Compiling and training...\n",
      "Epoch 1/10\n",
      "15300/15300 [==============================] - 37s 2ms/step - loss: 1.0227 - acc: 0.7457\n",
      "Epoch 2/10\n",
      "15300/15300 [==============================] - 17s 1ms/step - loss: 0.5918 - acc: 0.8288\n",
      "Epoch 3/10\n",
      "15300/15300 [==============================] - 18s 1ms/step - loss: 0.3769 - acc: 0.9020\n",
      "Epoch 4/10\n",
      "15300/15300 [==============================] - 17s 1ms/step - loss: 0.2516 - acc: 0.9507\n",
      "Epoch 5/10\n",
      "15300/15300 [==============================] - 17s 1ms/step - loss: 0.2049 - acc: 0.9630\n",
      "Epoch 6/10\n",
      "15300/15300 [==============================] - 16s 1ms/step - loss: 0.1791 - acc: 0.9673\n",
      "Epoch 7/10\n",
      "15300/15300 [==============================] - 16s 1ms/step - loss: 0.1656 - acc: 0.9685\n",
      "Epoch 00007: early stopping\n",
      "3826/3826 [==============================] - 32s 8ms/step\n",
      "['loss', 'acc'] [0.11067773076918383, 0.9691584472000693]\n",
      "CPU times: user 8min 22s, sys: 2min 42s, total: 11min 4s\n",
      "Wall time: 4min 8s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from time import time\n",
    "from keras.callbacks import TensorBoard\n",
    "tensorboard = TensorBoard(log_dir=\"data/logs/{}\".format(time()))\n",
    "\n",
    "epochs=10\n",
    "batch_size=128\n",
    "\n",
    "print(\"Compiling and training...\")\n",
    "model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])\n",
    "model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size, shuffle=\"batch\", \n",
    "          callbacks=[EarlyStopping(monitor='acc', min_delta=0.05, patience=2, verbose=2, mode=\"max\"), tensorboard])\n",
    "print(model.metrics_names, model.evaluate(X_test, y_test))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Treehouse\n",
    "Evaluate the model on samples from Treehouse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Converting expression to dataframe and storing in hdf5 file\n"
     ]
    }
   ],
   "source": [
    "if not os.path.exists(\"data\"):\n",
    "    os.makedirs(\"data\")\n",
    "    \n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\"):\n",
    "    print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "    r = requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\")\n",
    "    r.raise_for_status()\n",
    "    with open(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    expression = pd.read_csv(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \n",
    "                             sep=\"\\t\", index_col=0).dropna().astype(np.float32)\n",
    "    expression.to_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "treehouse_expression = pd.read_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\").sort_index(axis=1)\n",
    "print(\"expression: samples={} genes={}\".format(*expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(60448, 6)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>ENST00000373020.8,ENST00000494424.1,ENST000004...</td>\n",
       "      <td>2131.20</td>\n",
       "      <td>1986.93</td>\n",
       "      <td>155.00</td>\n",
       "      <td>1.95</td>\n",
       "      <td>5.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>ENST00000373031.4,ENST00000485971.1</td>\n",
       "      <td>1339.00</td>\n",
       "      <td>1194.73</td>\n",
       "      <td>2.00</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>ENST00000371582.8,ENST00000371584.8,ENST000003...</td>\n",
       "      <td>963.92</td>\n",
       "      <td>819.66</td>\n",
       "      <td>137.00</td>\n",
       "      <td>4.17</td>\n",
       "      <td>11.11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>ENST00000367770.5,ENST00000367771.10,ENST00000...</td>\n",
       "      <td>3037.97</td>\n",
       "      <td>2893.71</td>\n",
       "      <td>128.01</td>\n",
       "      <td>1.10</td>\n",
       "      <td>2.94</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>ENST00000286031.10,ENST00000359326.8,ENST00000...</td>\n",
       "      <td>1863.66</td>\n",
       "      <td>1719.39</td>\n",
       "      <td>57.99</td>\n",
       "      <td>0.84</td>\n",
       "      <td>2.24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            transcript_id(s)   length  \\\n",
       "gene_name                                                               \n",
       "TSPAN6     ENST00000373020.8,ENST00000494424.1,ENST000004...  2131.20   \n",
       "TNMD                     ENST00000373031.4,ENST00000485971.1  1339.00   \n",
       "DPM1       ENST00000371582.8,ENST00000371584.8,ENST000003...   963.92   \n",
       "SCYL3      ENST00000367770.5,ENST00000367771.10,ENST00000...  3037.97   \n",
       "C1orf112   ENST00000286031.10,ENST00000359326.8,ENST00000...  1863.66   \n",
       "\n",
       "           effective_length  expected_count   TPM   FPKM  \n",
       "gene_name                                                 \n",
       "TSPAN6              1986.93          155.00  1.95   5.18  \n",
       "TNMD                1194.73            2.00  0.04   0.11  \n",
       "DPM1                 819.66          137.00  4.17  11.11  \n",
       "SCYL3               2893.71          128.01  1.10   2.94  \n",
       "C1orf112            1719.39           57.99  0.84   2.24  "
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_id = \"TH01_0717_S01\"\n",
    "\n",
    "x = pd.read_csv(\"/treehouse/archive/downstream/{}/secondary/ucsc_cgl-rnaseq-cgl-pipeline-3.3.4-785eee9/RSEM/Hugo/rsem_genes.hugo.results\"\n",
    "            .format(sample_id), \n",
    "            sep=\"\\t\").set_index(\"gene_name\")\n",
    "print(x.shape)\n",
    "x.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(5172, 6)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_pruned = x.loc[X_pruned.columns.values]\n",
    "x_pruned.sort_index(axis=0, inplace=True)\n",
    "x_pruned = x_pruned[~x_pruned.index.duplicated(keep='first')]\n",
    "x_pruned.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([11809.04,   273.  ,   341.  , ...,   638.  ,    93.82,  1649.  ])"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x_pruned[\"expected_count\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[9.11246002e-01, 6.86533594e-19, 9.50189524e-23, 1.35506175e-25,\n",
       "        2.45716101e-21, 1.34075418e-22, 5.62957246e-18, 3.02469971e-22,\n",
       "        1.50773833e-29, 8.96338351e-11, 1.67384986e-13, 5.20123035e-30,\n",
       "        2.80644934e-25, 6.93729254e-13, 3.63123822e-30, 1.07628665e-15,\n",
       "        1.27661023e-32, 1.96785058e-28, 1.22029674e-20, 1.17232586e-21,\n",
       "        3.36096236e-14, 1.02125776e-21, 5.51603671e-17, 1.46389221e-13,\n",
       "        3.71290126e-37, 1.24981215e-20, 1.04258508e-28, 5.97230865e-20,\n",
       "        1.97737215e-18, 2.46605748e-33, 1.51068103e-28, 1.53874162e-16,\n",
       "        2.96808765e-29, 7.13299103e-20, 2.08419726e-09, 1.67728000e-23,\n",
       "        7.94406518e-24, 2.03599489e-20, 7.95892175e-18, 1.86481548e-21,\n",
       "        5.88688338e-25, 6.56929896e-32, 8.55488294e-25, 3.74134453e-18,\n",
       "        1.34325850e-28, 5.08881470e-37, 1.84078295e-12]], dtype=float32)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.predict(np.array([x_pruned[\"FPKM\"].values]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene_name\ttranscript_id(s)\tlength\teffective_length\texpected_count\tTPM\tFPKM\n",
      "TSPAN6\tENST00000373020.8,ENST00000494424.1,ENST00000496771.5,ENST00000612152.4,ENST00000614008.4\t2131.2\t1986.93\t155.0\t1.95\t5.18\n",
      "TNMD\tENST00000373031.4,ENST00000485971.1\t1339.0\t1194.73\t2.0\t0.04\t0.11\n",
      "DPM1\tENST00000371582.8,ENST00000371584.8,ENST00000371588.9,ENST00000413082.1,ENST00000466152.5,ENST00000494752.1\t963.92\t819.66\t137.0\t4.17\t11.11\n",
      "SCYL3\tENST00000367770.5,ENST00000367771.10,ENST00000367772.8,ENST00000423670.1,ENST00000470238.1\t3037.97\t2893.71\t128.01\t1.1\t2.94\n",
      "C1orf112\tENST00000286031.10,ENST00000359326.8,ENST00000413811.3,ENST00000459772.5,ENST00000466580.6,ENST00000472795.5,ENST00000481744.5,ENST00000496973.5,ENST00000498289.5\t1863.66\t1719.39\t57.99\t0.84\t2.24\n",
      "FGR\tENST00000374003.7,ENST00000374004.5,ENST00000374005.7,ENST00000399173.5,ENST00000457296.5,ENST00000468038.1,ENST00000475472.5\t1825.22\t1680.95\t60.0\t0.89\t2.37\n",
      "CFH\tENST00000359637.2,ENST00000367429.8,ENST00000466229.5,ENST00000470918.1,ENST00000496761.1,ENST00000630130.2\t3437.07\t3292.8\t1137.76\t8.62\t22.96\n",
      "FUCA2\tENST00000002165.10,ENST00000367585.1,ENST00000451668.1\t1798.09\t1653.83\t294.67\t4.45\t11.84\n",
      "GCLC\tENST00000229416.10,ENST00000504353.1,ENST00000504525.1,ENST00000505197.1,ENST00000505294.5,ENST00000509541.5,ENST00000510837.5,ENST00000513939.5,ENST00000514004.5,ENST00000514373.2,ENST00000514933.1,ENST00000515580.1,ENST00000616923.4\t1463.06\t1318.8\t151.0\t2.86\t7.61\n"
     ]
    }
   ],
   "source": [
    "!head /treehouse/archive/downstream/TH01_0717_S01/secondary/ucsc_cgl-rnaseq-cgl-pipeline-3.3.4-785eee9/RSEM/Hugo/rsem_genes.hugo.results"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
