{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ingest Gene Expression and Clinical Data from TCGA+TARGET+GTEX and Treehouse\n",
    "\n",
    "Download gene expression and clinical data from the [UCSC Xena Toil re-compute dataset](https://xenabrowser.net/datapages/?host=https://toil.xenahubs.net) and the [Treehouse Childhood Cancer Initiative](https://xenabrowser.net/datapages/?host=https://treehouse.xenahubs.net), wrangle, and store in an hdf5 file for quick loading machine learning. This dataset comprises gene expression data for twenty thousand tumor and normal samples processed using the exact same genomics pipeline and therefore can be compared to each other. Treehouse contains many of the same samples from TCGA and TARGET as Toil which we can use to verify our conversion. It also includes unique samples (all prefixed with TH or TR) which we can use as a hold-out set.\n",
    "\n",
    "Each of the source data set consists of a float vector, log2(TPM+0.001) in the case of TCGA+TARGET+GTEX or log2(TPM+1.0) in the case of Treehouse normalized, of gene expression for each of ~60k genes. Toil expression is labeled using Ensembl gene ids vs. Treehouse which uses Hugo. Associated with these data is clinical information on each sample such as type (tumor vs. normal), disease, primary site (where the sample came from in the human body) etc... We use this information to label the samples normal/0 vs. tumor/1 as well as to provide additional information for visualization and interpretation of models."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "  from ._conv import register_converters as _register_converters\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import requests\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import h5py\n",
    "\n",
    "if not os.path.exists(\"data\"):\n",
    "    os.makedirs(\"data\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download TGCA+TARGET+GTEX Expression\n",
    "Download expression data files from Xena and save in an hdf5 file. This can take around 30 minutes each between the download and the conversion from tsv into float32 dataframes. We download manually vs. passing read_csv a url directly as the latter times out with this size file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tcga_target_gtex_expression: samples=60498 genes=19260\n",
      "CPU times: user 6.48 s, sys: 5.6 s, total: 12.1 s\n",
      "Wall time: 11.9 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\"):\n",
    "    print(\"Downloading TCGA, TARGET and GTEX expression data from UCSC Xena\")\n",
    "    r = requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGtex_rsem_gene_tpm.gz\", stream=True)\n",
    "    response.raise_for_status()\n",
    "    with open(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(chunk_size=32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    pd.read_csv(\"data/TcgaTargetGtex_rsem_gene_tpm.gz\", sep=\"\\t\", index_col=0) \\\n",
    "        .astype(np.float32).to_hdf(\"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "tcga_target_gtex_expression = pd.read_hdf(\n",
    "    \"data/TcgaTargetGtex_rsem_gene_tpm.hd5\", \"expression\").dropna(axis=\"index\").sort_index(axis=\"columns\")\n",
    "print(\"tcga_target_gtex_expression: samples={} genes={}\".format(*tcga_target_gtex_expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2226-SM-5N9CH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-3026-SM-5GZYU</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ZR-A9CJ-01</th>\n",
       "      <th>TCGA-ZS-A9CD-01</th>\n",
       "      <th>TCGA-ZS-A9CE-01</th>\n",
       "      <th>TCGA-ZS-A9CF-01</th>\n",
       "      <th>TCGA-ZS-A9CF-02</th>\n",
       "      <th>TCGA-ZS-A9CG-01</th>\n",
       "      <th>TCGA-ZT-A8OM-01</th>\n",
       "      <th>TCGA-ZU-A8S4-01</th>\n",
       "      <th>TCGA-ZU-A8S4-11</th>\n",
       "      <th>TCGA-ZX-AA5X-01</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000242268.2</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-1.7809</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000259041.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000270112.3</th>\n",
       "      <td>-4.2934</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>0.3573</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000167578.16</th>\n",
       "      <td>5.1190</td>\n",
       "      <td>4.1277</td>\n",
       "      <td>4.4067</td>\n",
       "      <td>5.6860</td>\n",
       "      <td>4.0357</td>\n",
       "      <td>4.6849</td>\n",
       "      <td>4.5009</td>\n",
       "      <td>5.3954</td>\n",
       "      <td>4.9402</td>\n",
       "      <td>5.4683</td>\n",
       "      <td>...</td>\n",
       "      <td>4.1780</td>\n",
       "      <td>4.5547</td>\n",
       "      <td>3.6737</td>\n",
       "      <td>4.9331</td>\n",
       "      <td>3.6254</td>\n",
       "      <td>3.7646</td>\n",
       "      <td>5.5201</td>\n",
       "      <td>5.4216</td>\n",
       "      <td>3.3647</td>\n",
       "      <td>4.7991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278814.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "sample                                                                   \n",
       "ENSG00000242268.2                    -9.9658                   -9.9658   \n",
       "ENSG00000259041.1                    -9.9658                   -9.9658   \n",
       "ENSG00000270112.3                    -4.2934                    0.0014   \n",
       "ENSG00000167578.16                    5.1190                    4.1277   \n",
       "ENSG00000278814.1                    -9.9658                   -9.9658   \n",
       "\n",
       "                    GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "sample                                                                   \n",
       "ENSG00000242268.2                    -9.9658                   -1.2481   \n",
       "ENSG00000259041.1                    -9.9658                   -9.9658   \n",
       "ENSG00000270112.3                    -9.9658                   -5.5735   \n",
       "ENSG00000167578.16                    4.4067                    5.6860   \n",
       "ENSG00000278814.1                    -9.9658                   -9.9658   \n",
       "\n",
       "                    GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "sample                                                                   \n",
       "ENSG00000242268.2                    -3.8160                   -1.7809   \n",
       "ENSG00000259041.1                    -9.9658                   -9.9658   \n",
       "ENSG00000270112.3                     0.3573                   -9.9658   \n",
       "ENSG00000167578.16                    4.0357                    4.6849   \n",
       "ENSG00000278814.1                    -9.9658                   -9.9658   \n",
       "\n",
       "                    GTEX-1117F-2226-SM-5N9CH  GTEX-1117F-2426-SM-5EGGH  \\\n",
       "sample                                                                   \n",
       "ENSG00000242268.2                    -9.9658                   -9.9658   \n",
       "ENSG00000259041.1                    -9.9658                   -9.9658   \n",
       "ENSG00000270112.3                    -6.5064                   -5.0116   \n",
       "ENSG00000167578.16                    4.5009                    5.3954   \n",
       "ENSG00000278814.1                    -9.9658                   -9.9658   \n",
       "\n",
       "                    GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-3026-SM-5GZYU  \\\n",
       "sample                                                                   \n",
       "ENSG00000242268.2                    -3.6259                   -9.9658   \n",
       "ENSG00000259041.1                    -9.9658                   -9.9658   \n",
       "ENSG00000270112.3                    -9.9658                   -5.0116   \n",
       "ENSG00000167578.16                    4.9402                    5.4683   \n",
       "ENSG00000278814.1                    -9.9658                   -9.9658   \n",
       "\n",
       "                         ...         TCGA-ZR-A9CJ-01  TCGA-ZS-A9CD-01  \\\n",
       "sample                   ...                                            \n",
       "ENSG00000242268.2        ...                 -9.9658          -4.6082   \n",
       "ENSG00000259041.1        ...                 -9.9658          -9.9658   \n",
       "ENSG00000270112.3        ...                 -9.9658          -9.9658   \n",
       "ENSG00000167578.16       ...                  4.1780           4.5547   \n",
       "ENSG00000278814.1        ...                 -9.9658          -9.9658   \n",
       "\n",
       "                    TCGA-ZS-A9CE-01  TCGA-ZS-A9CF-01  TCGA-ZS-A9CF-02  \\\n",
       "sample                                                                  \n",
       "ENSG00000242268.2           -9.9658          -9.9658          -4.6082   \n",
       "ENSG00000259041.1           -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000270112.3           -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000167578.16           3.6737           4.9331           3.6254   \n",
       "ENSG00000278814.1           -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "                    TCGA-ZS-A9CG-01  TCGA-ZT-A8OM-01  TCGA-ZU-A8S4-01  \\\n",
       "sample                                                                  \n",
       "ENSG00000242268.2           -9.9658          -3.6259          -9.9658   \n",
       "ENSG00000259041.1           -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000270112.3           -9.9658          -9.9658          -9.9658   \n",
       "ENSG00000167578.16           3.7646           5.5201           5.4216   \n",
       "ENSG00000278814.1           -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "                    TCGA-ZU-A8S4-11  TCGA-ZX-AA5X-01  \n",
       "sample                                                \n",
       "ENSG00000242268.2           -9.9658          -9.9658  \n",
       "ENSG00000259041.1           -9.9658          -9.9658  \n",
       "ENSG00000270112.3           -9.9658          -6.5064  \n",
       "ENSG00000167578.16           3.3647           4.7991  \n",
       "ENSG00000278814.1           -9.9658          -9.9658  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_expression.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Covert Ensembl to Hugo\n",
    "\n",
    "Toil's expression values are per Ensembl gene id, which have a one or more to one relationship to Hugo gene names so we need to convert back into TPM, average (or add?), and then convert back to log2(tpm+0.001). We're using an assembled table from John Vivian @ UCSC here. Another\n",
    "option would be ftp://ftp.ebi.ac.uk/pub/databases/genenames/new/tsv/hgnc_complete_set.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>geneName</th>\n",
       "      <th>geneType</th>\n",
       "      <th>geneStatus</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>transcriptName</th>\n",
       "      <th>transcriptType</th>\n",
       "      <th>transcriptStatus</th>\n",
       "      <th>havanaGeneId</th>\n",
       "      <th>havanaTranscriptId</th>\n",
       "      <th>ccdsId</th>\n",
       "      <th>level</th>\n",
       "      <th>transcriptClass</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003.14</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000612152.4</td>\n",
       "      <td>TSPAN6-201</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000022002.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CCDS76001.1</td>\n",
       "      <td>3</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005.5</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000373031.4</td>\n",
       "      <td>TNMD-001</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000022001.1</td>\n",
       "      <td>OTTHUMT00000057481.1</td>\n",
       "      <td>CCDS14469.1</td>\n",
       "      <td>2</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419.12</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000371582.8</td>\n",
       "      <td>DPM1-005</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000032742.2</td>\n",
       "      <td>OTTHUMT00000079720.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457.13</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000470238.1</td>\n",
       "      <td>SCYL3-004</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000035941.4</td>\n",
       "      <td>OTTHUMT00000087552.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460.16</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>ENST00000466580.6</td>\n",
       "      <td>C1orf112-008</td>\n",
       "      <td>processed_transcript</td>\n",
       "      <td>KNOWN</td>\n",
       "      <td>OTTHUMG00000035821.7</td>\n",
       "      <td>OTTHUMT00000087524.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>nonCoding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    geneName        geneType geneStatus       transcriptId  \\\n",
       "geneId                                                                       \n",
       "ENSG00000000003.14    TSPAN6  protein_coding      KNOWN  ENST00000612152.4   \n",
       "ENSG00000000005.5       TNMD  protein_coding      KNOWN  ENST00000373031.4   \n",
       "ENSG00000000419.12      DPM1  protein_coding      KNOWN  ENST00000371582.8   \n",
       "ENSG00000000457.13     SCYL3  protein_coding      KNOWN  ENST00000470238.1   \n",
       "ENSG00000000460.16  C1orf112  protein_coding      KNOWN  ENST00000466580.6   \n",
       "\n",
       "                   transcriptName        transcriptType transcriptStatus  \\\n",
       "geneId                                                                     \n",
       "ENSG00000000003.14     TSPAN6-201        protein_coding            KNOWN   \n",
       "ENSG00000000005.5        TNMD-001        protein_coding            KNOWN   \n",
       "ENSG00000000419.12       DPM1-005        protein_coding            KNOWN   \n",
       "ENSG00000000457.13      SCYL3-004  processed_transcript            KNOWN   \n",
       "ENSG00000000460.16   C1orf112-008  processed_transcript            KNOWN   \n",
       "\n",
       "                            havanaGeneId    havanaTranscriptId       ccdsId  \\\n",
       "geneId                                                                        \n",
       "ENSG00000000003.14  OTTHUMG00000022002.1                   NaN  CCDS76001.1   \n",
       "ENSG00000000005.5   OTTHUMG00000022001.1  OTTHUMT00000057481.1  CCDS14469.1   \n",
       "ENSG00000000419.12  OTTHUMG00000032742.2  OTTHUMT00000079720.2          NaN   \n",
       "ENSG00000000457.13  OTTHUMG00000035941.4  OTTHUMT00000087552.1          NaN   \n",
       "ENSG00000000460.16  OTTHUMG00000035821.7  OTTHUMT00000087524.1          NaN   \n",
       "\n",
       "                    level transcriptClass  \n",
       "geneId                                     \n",
       "ENSG00000000003.14      3          coding  \n",
       "ENSG00000000005.5       2          coding  \n",
       "ENSG00000000419.12      2          coding  \n",
       "ENSG00000000457.13      2       nonCoding  \n",
       "ENSG00000000460.16      2       nonCoding  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not os.path.exists(\"data/ensembl_to_hugo.tsv\"):\n",
    "    with open(\"data/ensembl_to_hugo.tsv\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://github.com/jvivian/docker_tools/blob/master/gencode_hugo_mapping/attrs.tsv?raw=true\").content)\n",
    "ensemble_to_hugo = pd.read_table(\"data/ensembl_to_hugo.tsv\",index_col=0).sort_index(axis=\"index\")\n",
    "\n",
    "# Remove duplicates\n",
    "ensemble_to_hugo = ensemble_to_hugo[~ensemble_to_hugo.index.duplicated(keep='first')]\n",
    "ensemble_to_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2226-SM-5N9CH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-3026-SM-5GZYU</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ZR-A9CJ-01</th>\n",
       "      <th>TCGA-ZS-A9CD-01</th>\n",
       "      <th>TCGA-ZS-A9CE-01</th>\n",
       "      <th>TCGA-ZS-A9CF-01</th>\n",
       "      <th>TCGA-ZS-A9CF-02</th>\n",
       "      <th>TCGA-ZS-A9CG-01</th>\n",
       "      <th>TCGA-ZT-A8OM-01</th>\n",
       "      <th>TCGA-ZU-A8S4-01</th>\n",
       "      <th>TCGA-ZU-A8S4-11</th>\n",
       "      <th>TCGA-ZX-AA5X-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RP11-368I23.2</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-1.2481</td>\n",
       "      <td>-3.8160</td>\n",
       "      <td>-1.7809</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-4.6082</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.6259</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-167B3.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RP11-742D12.2</th>\n",
       "      <td>-4.2934</td>\n",
       "      <td>0.0014</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.5735</td>\n",
       "      <td>0.3573</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-5.0116</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-6.5064</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RAB4B</th>\n",
       "      <td>5.1190</td>\n",
       "      <td>4.1277</td>\n",
       "      <td>4.4067</td>\n",
       "      <td>5.6860</td>\n",
       "      <td>4.0357</td>\n",
       "      <td>4.6849</td>\n",
       "      <td>4.5009</td>\n",
       "      <td>5.3954</td>\n",
       "      <td>4.9402</td>\n",
       "      <td>5.4683</td>\n",
       "      <td>...</td>\n",
       "      <td>4.1780</td>\n",
       "      <td>4.5547</td>\n",
       "      <td>3.6737</td>\n",
       "      <td>4.9331</td>\n",
       "      <td>3.6254</td>\n",
       "      <td>3.7646</td>\n",
       "      <td>5.5201</td>\n",
       "      <td>5.4216</td>\n",
       "      <td>3.3647</td>\n",
       "      <td>4.7991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC104071.1</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "RP11-368I23.2                   -9.9658                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -4.2934                    0.0014   \n",
       "RAB4B                            5.1190                    4.1277   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "RP11-368I23.2                   -9.9658                   -1.2481   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -9.9658                   -5.5735   \n",
       "RAB4B                            4.4067                    5.6860   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "RP11-368I23.2                   -3.8160                   -1.7809   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                    0.3573                   -9.9658   \n",
       "RAB4B                            4.0357                    4.6849   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-2226-SM-5N9CH  GTEX-1117F-2426-SM-5EGGH  \\\n",
       "RP11-368I23.2                   -9.9658                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -6.5064                   -5.0116   \n",
       "RAB4B                            4.5009                    5.3954   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "               GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-3026-SM-5GZYU  \\\n",
       "RP11-368I23.2                   -3.6259                   -9.9658   \n",
       "RP11-167B3.1                    -9.9658                   -9.9658   \n",
       "RP11-742D12.2                   -9.9658                   -5.0116   \n",
       "RAB4B                            4.9402                    5.4683   \n",
       "AC104071.1                      -9.9658                   -9.9658   \n",
       "\n",
       "                    ...         TCGA-ZR-A9CJ-01  TCGA-ZS-A9CD-01  \\\n",
       "RP11-368I23.2       ...                 -9.9658          -4.6082   \n",
       "RP11-167B3.1        ...                 -9.9658          -9.9658   \n",
       "RP11-742D12.2       ...                 -9.9658          -9.9658   \n",
       "RAB4B               ...                  4.1780           4.5547   \n",
       "AC104071.1          ...                 -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZS-A9CE-01  TCGA-ZS-A9CF-01  TCGA-ZS-A9CF-02  \\\n",
       "RP11-368I23.2          -9.9658          -9.9658          -4.6082   \n",
       "RP11-167B3.1           -9.9658          -9.9658          -9.9658   \n",
       "RP11-742D12.2          -9.9658          -9.9658          -9.9658   \n",
       "RAB4B                   3.6737           4.9331           3.6254   \n",
       "AC104071.1             -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZS-A9CG-01  TCGA-ZT-A8OM-01  TCGA-ZU-A8S4-01  \\\n",
       "RP11-368I23.2          -9.9658          -3.6259          -9.9658   \n",
       "RP11-167B3.1           -9.9658          -9.9658          -9.9658   \n",
       "RP11-742D12.2          -9.9658          -9.9658          -9.9658   \n",
       "RAB4B                   3.7646           5.5201           5.4216   \n",
       "AC104071.1             -9.9658          -9.9658          -9.9658   \n",
       "\n",
       "               TCGA-ZU-A8S4-11  TCGA-ZX-AA5X-01  \n",
       "RP11-368I23.2          -9.9658          -9.9658  \n",
       "RP11-167B3.1           -9.9658          -9.9658  \n",
       "RP11-742D12.2          -9.9658          -6.5064  \n",
       "RAB4B                   3.3647           4.7991  \n",
       "AC104071.1             -9.9658          -9.9658  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create a new data frame replacing the ensembl based index with hugo dropping any where there is no conversion\n",
    "tcga_target_gtex_expression_hugo = tcga_target_gtex_expression.copy()\n",
    "tcga_target_gtex_expression_hugo.index = ensemble_to_hugo.reindex(tcga_target_gtex_expression.index).geneName.values\n",
    "tcga_target_gtex_expression_hugo = tcga_target_gtex_expression_hugo[tcga_target_gtex_expression_hugo.index.notnull()]\n",
    "tcga_target_gtex_expression_hugo.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GTEX-146FH-1726-SM-5QGQ2    1.000001e+06\n",
       "GTEX-WZTO-2926-SM-3NM9I     9.999974e+05\n",
       "TCGA-AB-2965-03             9.999970e+05\n",
       "dtype: float64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# While we're at it let's verify that the sum of all expression levels for a sample in TPM space sums to 1 million\n",
    "tcga_target_gtex_expression_hugo[[\"GTEX-146FH-1726-SM-5QGQ2\", \"GTEX-WZTO-2926-SM-3NM9I\", \"TCGA-AB-2965-03\"]].apply(np.exp2).apply(lambda x: x - 0.001).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1min 9s, sys: 56.7 s, total: 2min 5s\n",
      "Wall time: 2min\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Multiple Ensemble genes map to the same Hugo name. Each of these values has been normalized via log2(TPM+0.001)\n",
    "# so we convert back into TPM, compute the mean, and re-normalize.\n",
    "tcga_target_gtex_expression_hugo_tpm = tcga_target_gtex_expression_hugo \\\n",
    "    .apply(np.exp2).subtract(0.001).groupby(level=0).aggregate(np.mean).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GTEX-1117F-0226-SM-5GZZ7</th>\n",
       "      <th>GTEX-1117F-0426-SM-5EGHI</th>\n",
       "      <th>GTEX-1117F-0526-SM-5EGHJ</th>\n",
       "      <th>GTEX-1117F-0626-SM-5N9CS</th>\n",
       "      <th>GTEX-1117F-0726-SM-5GIEN</th>\n",
       "      <th>GTEX-1117F-1326-SM-5EGHH</th>\n",
       "      <th>GTEX-1117F-2226-SM-5N9CH</th>\n",
       "      <th>GTEX-1117F-2426-SM-5EGGH</th>\n",
       "      <th>GTEX-1117F-2826-SM-5GZXL</th>\n",
       "      <th>GTEX-1117F-3026-SM-5GZYU</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-ZR-A9CJ-01</th>\n",
       "      <th>TCGA-ZS-A9CD-01</th>\n",
       "      <th>TCGA-ZS-A9CE-01</th>\n",
       "      <th>TCGA-ZS-A9CF-01</th>\n",
       "      <th>TCGA-ZS-A9CF-02</th>\n",
       "      <th>TCGA-ZS-A9CG-01</th>\n",
       "      <th>TCGA-ZT-A8OM-01</th>\n",
       "      <th>TCGA-ZU-A8S4-01</th>\n",
       "      <th>TCGA-ZU-A8S4-11</th>\n",
       "      <th>TCGA-ZX-AA5X-01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.96580</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5_8S_rRNA</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.965800</td>\n",
       "      <td>-9.96580</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7SK</th>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-3.406063</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-2.003874</td>\n",
       "      <td>-2.244654</td>\n",
       "      <td>-3.45799</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "      <td>-9.9658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>4.4595</td>\n",
       "      <td>1.1512</td>\n",
       "      <td>5.2411</td>\n",
       "      <td>5.4758</td>\n",
       "      <td>4.553400</td>\n",
       "      <td>4.6224</td>\n",
       "      <td>4.5028</td>\n",
       "      <td>5.613000</td>\n",
       "      <td>4.447600</td>\n",
       "      <td>4.73020</td>\n",
       "      <td>...</td>\n",
       "      <td>3.6463</td>\n",
       "      <td>11.1518</td>\n",
       "      <td>11.8930</td>\n",
       "      <td>8.9466</td>\n",
       "      <td>7.6347</td>\n",
       "      <td>9.1454</td>\n",
       "      <td>7.6426</td>\n",
       "      <td>3.5994</td>\n",
       "      <td>10.9593</td>\n",
       "      <td>3.5935</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.9343</td>\n",
       "      <td>-1.2828</td>\n",
       "      <td>0.8488</td>\n",
       "      <td>2.6325</td>\n",
       "      <td>1.305100</td>\n",
       "      <td>1.5514</td>\n",
       "      <td>0.8647</td>\n",
       "      <td>3.114500</td>\n",
       "      <td>2.176600</td>\n",
       "      <td>2.45970</td>\n",
       "      <td>...</td>\n",
       "      <td>1.6234</td>\n",
       "      <td>0.5568</td>\n",
       "      <td>0.9038</td>\n",
       "      <td>-0.4719</td>\n",
       "      <td>-1.3183</td>\n",
       "      <td>-1.5522</td>\n",
       "      <td>2.5087</td>\n",
       "      <td>-0.4521</td>\n",
       "      <td>-1.4305</td>\n",
       "      <td>1.8036</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 19260 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           GTEX-1117F-0226-SM-5GZZ7  GTEX-1117F-0426-SM-5EGHI  \\\n",
       "5S_rRNA                     -9.9658                   -9.9658   \n",
       "5_8S_rRNA                   -9.9658                   -9.9658   \n",
       "7SK                         -9.9658                   -9.9658   \n",
       "A1BG                         4.4595                    1.1512   \n",
       "A1BG-AS1                     0.9343                   -1.2828   \n",
       "\n",
       "           GTEX-1117F-0526-SM-5EGHJ  GTEX-1117F-0626-SM-5N9CS  \\\n",
       "5S_rRNA                     -9.9658                   -9.9658   \n",
       "5_8S_rRNA                   -9.9658                   -9.9658   \n",
       "7SK                         -9.9658                   -9.9658   \n",
       "A1BG                         5.2411                    5.4758   \n",
       "A1BG-AS1                     0.8488                    2.6325   \n",
       "\n",
       "           GTEX-1117F-0726-SM-5GIEN  GTEX-1117F-1326-SM-5EGHH  \\\n",
       "5S_rRNA                   -9.965800                   -9.9658   \n",
       "5_8S_rRNA                 -9.965800                   -9.9658   \n",
       "7SK                       -3.406063                   -9.9658   \n",
       "A1BG                       4.553400                    4.6224   \n",
       "A1BG-AS1                   1.305100                    1.5514   \n",
       "\n",
       "           GTEX-1117F-2226-SM-5N9CH  GTEX-1117F-2426-SM-5EGGH  \\\n",
       "5S_rRNA                     -9.9658                 -9.965800   \n",
       "5_8S_rRNA                   -9.9658                 -9.965800   \n",
       "7SK                         -9.9658                 -2.003874   \n",
       "A1BG                         4.5028                  5.613000   \n",
       "A1BG-AS1                     0.8647                  3.114500   \n",
       "\n",
       "           GTEX-1117F-2826-SM-5GZXL  GTEX-1117F-3026-SM-5GZYU  \\\n",
       "5S_rRNA                   -9.965800                  -9.96580   \n",
       "5_8S_rRNA                 -9.965800                  -9.96580   \n",
       "7SK                       -2.244654                  -3.45799   \n",
       "A1BG                       4.447600                   4.73020   \n",
       "A1BG-AS1                   2.176600                   2.45970   \n",
       "\n",
       "                ...         TCGA-ZR-A9CJ-01  TCGA-ZS-A9CD-01  TCGA-ZS-A9CE-01  \\\n",
       "5S_rRNA         ...                 -9.9658          -9.9658          -9.9658   \n",
       "5_8S_rRNA       ...                 -9.9658          -9.9658          -9.9658   \n",
       "7SK             ...                 -9.9658          -9.9658          -9.9658   \n",
       "A1BG            ...                  3.6463          11.1518          11.8930   \n",
       "A1BG-AS1        ...                  1.6234           0.5568           0.9038   \n",
       "\n",
       "           TCGA-ZS-A9CF-01  TCGA-ZS-A9CF-02  TCGA-ZS-A9CG-01  TCGA-ZT-A8OM-01  \\\n",
       "5S_rRNA            -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "5_8S_rRNA          -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "7SK                -9.9658          -9.9658          -9.9658          -9.9658   \n",
       "A1BG                8.9466           7.6347           9.1454           7.6426   \n",
       "A1BG-AS1           -0.4719          -1.3183          -1.5522           2.5087   \n",
       "\n",
       "           TCGA-ZU-A8S4-01  TCGA-ZU-A8S4-11  TCGA-ZX-AA5X-01  \n",
       "5S_rRNA            -9.9658          -9.9658          -9.9658  \n",
       "5_8S_rRNA          -9.9658          -9.9658          -9.9658  \n",
       "7SK                -9.9658          -9.9658          -9.9658  \n",
       "A1BG                3.5994          10.9593           3.5935  \n",
       "A1BG-AS1           -0.4521          -1.4305           1.8036  \n",
       "\n",
       "[5 rows x 19260 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tcga_target_gtex_expression_hugo_tpm.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download Treehouse Expression\n",
    "\n",
    "The Treehouse public compendium is in Hugo log2(tpm+1). We need to download and convert into lot2(tpm+0.001) to match our TCGA+TARGET+GTEXt dataset above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "treehouse_expression: samples=58581 genes=11078\n",
      "CPU times: user 124 ms, sys: 1.43 s, total: 1.55 s\n",
      "Wall time: 1.55 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\"):\n",
    "    print(\"Downloading Treehouse Public Compendium\")\n",
    "    r = requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\",\n",
    "                     stream=True)\n",
    "    r.raise_for_status()\n",
    "    with open(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        for chunk in r.iter_content(chunk_size=32768):\n",
    "            f.write(chunk)\n",
    "\n",
    "if not os.path.exists(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\"):\n",
    "    print(\"Converting expression to dataframe and storing in hdf5 file\")\n",
    "    pd.read_csv(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.tsv.gz\", sep=\"\\t\", index_col=0) \\\n",
    "    .astype(np.float32).to_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \n",
    "                               \"expression\", mode=\"w\", format=\"fixed\")\n",
    "\n",
    "treehouse_expression = pd.read_hdf(\"data/treehouse_public_samples_unique_hugo_log2_tpm_plus_1.2017-09-11.hd5\", \"expression\")\n",
    "print(\"treehouse_expression: samples={} genes={}\".format(*treehouse_expression.shape))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check that we don't have any null/nan at this point\n",
    "assert not tcga_target_gtex_expression_hugo_tpm.isnull().values.any()\n",
    "assert not treehouse_expression.isnull().values.any()\n",
    "\n",
    "# Make sure they have identical hugo gene indexes\n",
    "assert np.array_equal(tcga_target_gtex_expression_hugo_tpm.index, treehouse_expression.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert into log2(tpm+0.001)\n",
    "treehouse_expression_hugo_tpm = treehouse_expression.apply(np.exp2).subtract(1.0).add(0.001).apply(np.log2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## NOTE and REMINDER\n",
    "\n",
    "The current public Treehouse compendium was created by combining expression values that map to the same Hugo gene identify by calculating the mean of their log2(tpm+1) values. As a result those values will not match perfectly with the same samples in the TCGA+TARGET+GTEX dataset. The next public compendium from Treehouse will calculate mean in TPM space. Continue on here but later we need to come back and update this - or calculate things the right way for the TH and TR samples from the raw data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene with maximum delta: Metazoa_SRP 6.039602\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    58581.000000\n",
       "mean         0.001218\n",
       "std          0.060920\n",
       "min         -0.013058\n",
       "25%         -0.000107\n",
       "50%         -0.000016\n",
       "75%         -0.000016\n",
       "max          6.039602\n",
       "Name: TCGA-ZQ-A9CR-01, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to verify the TCGA+TARGET samples in the Treehouse compendium match TPM wise with our conversions above\n",
    "sample_id = \"TCGA-ZQ-A9CR-01\"\n",
    "\n",
    "np.allclose(tcga_target_gtex_expression_hugo_tpm[sample_id], treehouse_expression_hugo_tpm[sample_id], 1, 1)\n",
    "\n",
    "argmax = (tcga_target_gtex_expression_hugo_tpm[sample_id] - treehouse_expression_hugo_tpm[sample_id]).values.argmax()\n",
    "gene = tcga_target_gtex_expression_hugo_tpm.index[argmax]\n",
    "print(\"Gene with maximum delta:\", gene,\n",
    "      tcga_target_gtex_expression_hugo_tpm[sample_id][gene] - treehouse_expression_hugo_tpm[sample_id][gene])\n",
    "\n",
    "(tcga_target_gtex_expression_hugo_tpm[sample_id] - treehouse_expression_hugo_tpm[sample_id]).describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download Labels and Conform to Expression Indexes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the sample labels from Xena ie clinical/phenotype information on each sample\n",
    "if not os.path.exists(\"data/TcgaTargetGTEX_phenotype.txt.gz\"):\n",
    "    with open(\"data/TcgaTargetGTEX_phenotype.txt.gz\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://toil.xenahubs.net/download/TcgaTargetGTEX_phenotype.txt.gz\").content)\n",
    "\n",
    "tcga_target_gtex_labels = pd.read_table(\n",
    "    \"data/TcgaTargetGTEX_phenotype.txt.gz\", compression=\"gzip\", \n",
    "    header=0, names=[\"id\", \"category\", \"disease\", \"primary_site\", \"sample_type\", \"gender\", \"study\"],\n",
    "    sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=\"index\")\n",
    "\n",
    "# Compute and add a tumor/normal column - TCGA and TARGET have some normal samples, GTEX is all normal.\n",
    "tcga_target_gtex_labels[\"tumor_normal\"] = tcga_target_gtex_labels.apply(\n",
    "    lambda row: \"Normal\" if row[\"sample_type\"] in [\"Cell Line\", \"Normal Tissue\", \"Solid Tissue Normal\"]\n",
    "    else \"Tumor\", axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Remove rows where we have no labels or the primary_site label is null\n",
    "intersection = tcga_target_gtex_expression_hugo_tpm.columns.intersection(\n",
    "    tcga_target_gtex_labels[pd.notnull(tcga_target_gtex_labels[\"primary_site\"])].index)\n",
    "tcga_target_gtex_expression_common = tcga_target_gtex_expression_hugo_tpm.loc[:, tcga_target_gtex_expression_hugo_tpm.columns.isin(intersection)]\n",
    "tcga_target_gtex_labels_common = tcga_target_gtex_labels[tcga_target_gtex_labels.index.isin(intersection)]\n",
    "\n",
    "# Make sure the label and example samples are in the same order\n",
    "assert(tcga_target_gtex_expression_common.columns.equals(tcga_target_gtex_labels_common.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_in_years</th>\n",
       "      <th>gender</th>\n",
       "      <th>disease</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAKSWW-03</th>\n",
       "      <td>15.11</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXHJ-09</th>\n",
       "      <td>6.08</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PAMXSP-09</th>\n",
       "      <td>3.35</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-03</th>\n",
       "      <td>6.4</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TARGET-10-PANCVR-04</th>\n",
       "      <td>6.4</td>\n",
       "      <td>male</td>\n",
       "      <td>acute lymphoblastic leukemia</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    age_in_years gender                       disease\n",
       "id                                                                   \n",
       "TARGET-10-PAKSWW-03        15.11   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PAMXHJ-09         6.08   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PAMXSP-09         3.35   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PANCVR-03          6.4   male  acute lymphoblastic leukemia\n",
       "TARGET-10-PANCVR-04          6.4   male  acute lymphoblastic leukemia"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read in the sample labels from Xena ie clinical/phenotype information on each sample\n",
    "if not os.path.exists(\"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\"):\n",
    "    with open(\"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\", \"wb\") as f:\n",
    "        f.write(requests.get(\"https://treehouse.xenahubs.net/download/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\").content)\n",
    "\n",
    "treehouse_labels = pd.read_table(\n",
    "    \"data/treehouse_public_samples_clinical_metadata.2017-09-11.tsv.gz\", compression=\"gzip\", \n",
    "    header=0, names=[\"id\", \"age_in_years\", \"gender\", \"disease\"],\n",
    "    sep=\"\\t\", encoding=\"ISO-8859-1\", index_col=0, dtype=\"str\").sort_index(axis=0)\n",
    "\n",
    "treehouse_labels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# The Treehouse compendium includes TARGET and TCGA Samples\n",
    "treehouse_labels_pruned = treehouse_labels.filter(regex='\\ATH|\\ATHR', axis=\"index\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'acute lymphoblastic leukemia',\n",
       " 'acute myeloid leukemia',\n",
       " 'glioblastoma multiforme',\n",
       " 'lung adenocarcinoma',\n",
       " 'neuroblastoma',\n",
       " 'sarcoma',\n",
       " 'thyroid carcinoma',\n",
       " 'wilms tumor'}"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# See how many disease labels overlap\n",
    "set(tcga_target_gtex_labels.disease.str.lower()).intersection(treehouse_labels_pruned.disease.str.lower())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(58581, 549)\n"
     ]
    }
   ],
   "source": [
    "# Subset treeshouse expression to those not in TCGA+TARGET+GTEX\n",
    "treehouse_expression_hugo_tpm_pruned = treehouse_expression_hugo_tpm.loc[:, treehouse_labels_pruned.index]\n",
    "print(treehouse_expression_hugo_tpm_pruned.shape)\n",
    "assert(tcga_target_gtex_expression_common.index.equals(treehouse_expression_hugo_tpm_pruned.index))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Export\n",
    "Write expression and labels for both datasets out to hdf5 files sorted and clean and in machine learning format of rows = samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.HDFStore(\"data/tcga_target_gtex.h5\", \"w\") as store:\n",
    "    store[\"expression\"] = tcga_target_gtex_expression_common.T.sort_index(axis=\"columns\")\n",
    "    store[\"labels\"] = tcga_target_gtex_labels_common.astype(str)\n",
    "    \n",
    "with pd.HDFStore(\"data/treehouse.h5\", \"w\") as store:\n",
    "    store[\"expression\"] = treehouse_expression_hugo_tpm_pruned.T.sort_index(axis=\"columns\")\n",
    "    store[\"labels\"] = treehouse_labels_pruned.astype(str)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
